---
title: "Neisseriaceae"
author: "J. J. Giacomini"
date: "2023-12-04"
output: html_document
---

# 1. Load packages

```{r, eval=FALSE}

library(ggplot2)
library(dplyr)
library(tidyr)
library(reshape2)
library(scales)
library(vegan)
library(janitor)
library(egg)   
library(ggdendro)
library(forcats)
library(RColorBrewer)
library(grid)
library(ggh4x)

```



# 2.a. Load data: detection,  mean coverage Q2Q3 

```{r, eval=FALSE}

# Load mean coverage Q2Q3 data into R
df_coverage = read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/P_0003_Neisseriaceae-mean_coverage_Q2Q3.txt", header=TRUE, sep = "\t")

# Load detection data into R
df_detect = read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/P_0003_Neisseriaceae-detection.txt", header=TRUE, sep = "\t")

# Load sample metadata into R
sample_metadata <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/HMP_metadata/Final_1297_HMP_metagenomes_metadata.csv", header=TRUE)

# load mapping stats
mapping_stats <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/P_0003_Neisseriaceae-mapping_stats_new.txt", header=TRUE, sep = "\t")

```

# 2.b. Merge mapping data with metagenome metadata

```{r, eval=FALSE}

# Merge the two 
mapping_metadata_merged <- merge(mapping_stats, sample_metadata, by.x = "layers", by.y = "Internal_ID")

# Add variable for percentage reads mapped
mapping_metadata_merged <- mapping_metadata_merged %>% 
  mutate(total_number_reads = (total_pairs_passed*2)) %>% 
  mutate(percent_reads_mapped = 100*(total_reads_mapped/total_number_reads))

```


# 3.a. Organize detection data

### 50% breadth threshold

```{r, eval=FALSE}

# change data frame from wide to long format
melted_df_detect <- melt(df_detect)

melted_df_detect <- melted_df_detect %>% 
  mutate(detected = ifelse(value >= 0.5, "detected", "un-detected"))

melted_df_detect$detected <- as.factor(melted_df_detect$detected)

melted_df_detect_filtered <- melted_df_detect %>% 
  filter(detected == "detected")


# make oral site variable for facet grid
melted_df_detect_2 <- melted_df_detect %>% mutate(site = case_when(grepl('TD_', variable) ~ 'TD', 
                                      grepl('KG_', variable) ~ 'KG', 
                                      grepl('PP_', variable) ~ 'SUPP',
                                      grepl('PB_', variable) ~ 'SUBP',
                                      grepl('SA_', variable) ~ 'SV',
                                      grepl('TH_', variable) ~ 'TH',
                                      grepl('PT_', variable) ~ 'PT',
                                      grepl('HP_', variable) ~ 'HP',
                                      grepl('BM_', variable) ~ 'BM')) 


# specify order for site terms
melted_df_detect_2$site <- factor(melted_df_detect_2$site, levels = c("SUPP", "SUBP", "KG", "BM", "TD", "PT", "HP", "TH", "SV"))

# sort "variable" based on pre-sorted values of "site"
melted_df_detect_2 <- melted_df_detect_2 %>%
  arrange(site, variable) %>%               # sort your dataframe
  mutate(variable = factor(variable, unique(variable))) # reset your factor-column based on that order


# make 1 or 0 binary variable for detection
melted_df_detect_2$detected <- as.factor(melted_df_detect_2$detected)
melted_df_detect_2$detected <- as.factor(melted_df_detect_2$detected)

melted_df_detect_2 <- melted_df_detect_2 %>%  mutate(detection_binary = case_when(grepl("un-detected", detected) ~ 0, 
                                                           grepl("detected", detected) ~ 1))
                                             
melted_df_detect_2$detection_binary <- as.factor(melted_df_detect_2$detection_binary)


# join metadata with to get total reads mapped into data frame
melted_strain_detection = merge(mapping_metadata_merged, melted_df_detect_2, by.x = "layers", by.y = "variable") 



```

# 3.b. (Optional) Reorder genomes based on bray curtis distance and hc

```{r, eval=FALSE}

# subset data frame for clustering species
melted_strain_detection_clustering_df <- melted_strain_detection %>% 
  select(bins, layers, value)

# pivot wider
wide_df <- melted_strain_detection_clustering_df %>% pivot_wider(names_from = layers, values_from = value)

# make tibble into data frame object
wide_df <- as.data.frame(wide_df)

# rename row names to species IDs
wide_df2 <- wide_df[,-1]
rownames(wide_df2) <- wide_df[,1]


wide_matrix <- as.matrix(wide_df2)

# calculate Bray-Curtis distance among samples
dist_df <- vegan::vegdist(wide_matrix, method = "bray")


#Replace na values with 0 using is.na()
dist_df[dist_df=="NaN"]<-0
  
# Hierarchical clustering using Complete Linkage
hc <- hclust(dist_df, method = "complete" )

# create dedrogram object
dhc <- as.dendrogram(hc)

# create rectangular lines object
ddata <- dendro_data(dhc, type = "rectangle")

species_dend_plot <- ggplot(segment(ddata)) + 
  geom_segment(aes(x = x, y = y, xend = xend, yend = yend), linewidth = 1) + 
  coord_flip() + 
  scale_y_reverse(expand = c(0.2, 0)) +
  scale_x_reverse() +
  theme_classic() +
  theme(axis.ticks.y = element_blank(),
        axis.ticks.x=element_blank(),
        axis.text.x = element_blank(),
        axis.text.y =  element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.title.y = element_blank(),
        axis.title.x = element_blank()) +
  theme(panel.spacing.x=unit(0.05, "lines"),
        plot.margin=unit(c(0,0,0,0), "cm"))

# save dend plot
ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/strain_detection_dendrogram.pdf",species_dend_plot, width = 5, height = 7)



# reorder genomes for triples plot based on hclust
genome_order <- hc$labels[hc$order]


melted_strain_detection$bins <- factor(melted_strain_detection$bins, levels = genome_order)
```


# 3.c. Reorder based on pangenome gene freq tree

```{r, eval=FALSE}

library("ape")
library("ggtree")
library("phytools")

# load tree
tree <- read.tree("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/09_PANGENOME/G_0213/gene_cluster_frequencies_custom_newick")

# extract tip labels
tip_labels <- tree$tip.label

# reorder genomes; First, ensure that bins is a factor
melted_strain_detection$bins <- factor(melted_strain_detection$bins, levels = tip_labels)


```

# 3.c.ii Reorder based on ANI tree

```{r, eval=FALSE}

library("ape")
library("ggtree")
library("phytools")

# load tree
ANI_tree <- read.tree("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/09_PANGENOME/G_0213/ANIb-RESULTS/ANIb_percentage_identity.newick")

# extract tip labels
ANI_tip_labels <- ANI_tree$tip.label

# reorder genomes; First, ensure that bins is a factor
melted_strain_detection$bins <- factor(melted_strain_detection$bins, levels = ANI_tip_labels)


```


# 3.d. Save updated detection data & pangenome tree order

```{r, eval=FALSE}

write.csv(melted_strain_detection,"/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Strain_level_updated_detection_data.csv", row.names = FALSE, quote = FALSE)

write.csv(tip_labels,"/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/pangenome_custom_gene_freq_order.csv", row.names = FALSE, quote = FALSE)

```

# 3.e. Load data + pangenome colors

```{r, eval=FALSE}


melted_strain_detection <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Strain_level_updated_detection_data.csv", header = TRUE)

# specify order for site terms
melted_strain_detection$site <- factor(melted_strain_detection$site, levels = c("SUPP", "SUBP", "KG", "BM", "TD", "PT", "HP", "TH", "SV"))


# This file was manually created in Excel. load csv file with genome IDs and colors; theis file was made when the tanglegram was made
bins_and_colors <-read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/pangenome_custom_gene_freq_order.csv", header = TRUE)

# species colors 
species_tick_colors=bins_and_colors$Species_color

# group colors 
group_tick_colors=bins_and_colors$Group_color


# merge with bins_nd_colors
bin_colors <- merge(melted_strain_detection, bins_and_colors, by.x = "bins", by.y = "Pangenome_ID")

# group colors 
group_tick_colors <- bin_colors %>% 
  group_by(bins) %>% 
  select(bins, Group_color) %>% 
  distinct() %>% 
  #mutate(bins = factor(bins, levels = ANI_tip_labels)) %>%
  arrange(bins)



```

# 5. Breadth of Coverage plots

##### 5.a. FIGURE: Binary detection plot - samples ordered by total mapped reads

```{r, eval=FALSE}

dodge <- position_dodge(width=0.9)

total_reads_mapped_detection <- ggplot(data = melted_strain_detection, aes(x=reorder(layers, -total_number_reads), y=total_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 25000000), breaks=seq(0,25000000,by=25000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("No. of reads mapped", "(0 - 25 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x = element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x = unit(0.1, "lines"),
        plot.margin = unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)
#total_reads_mapped_detection

plot_total_reads_detection <- ggplot(data = melted_strain_detection, aes(x=reorder(layers, -total_number_reads), y=total_number_reads)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 200000000), breaks=seq(0,200000000,by=200000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Total reads", "(0 - 200 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.text.y=element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)
#plot_total_reads_detection


plot_percent_reads_mapped_detection <- ggplot(data = melted_strain_detection, aes(x=reorder(layers, -total_number_reads), y=percent_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 50), breaks=seq(0,50,by=50), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Percent reads mapped", "(0 - 50 %)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)
#plot_percent_reads_mapped_detection


plot_strain_level_detection <- ggplot(data = melted_strain_detection, aes(x=reorder(layers,-total_number_reads), y=bins, fill=as.factor(detection_binary))) +
  geom_tile(height=1, width=1) + 
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") + 
  scale_fill_manual(values = c("black", "cyan"),
                    labels = c("Un-detected (Breadth < 50%)", "Detected (Breadth > 50%)"),
                    name = "Genome detection") +
  #scale_y_discrete(limits = rev(levels(melted_strain_detection$bins))) +
  scale_y_discrete(limits = rev(levels(melted_strain_detection$bins))) +
  ylab(NULL) +  
  xlab(NULL) + 
  theme_classic() + 
  theme(text = element_text(size = 12),
        axis.ticks.y = element_line(color = rev(group_tick_colors), size = 1.2),
        axis.ticks.length = unit(1, "cm"),
        axis.text.x=element_blank(),
        axis.text.y =  element_text(size = 2, color = "black"),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "black")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        #panel.border = element_rect(colour = "black", fill = NA),
        strip.text=element_text(size=12)) + 
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,1,1), "cm")) +
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)
#plot_strain_level_detection


Strain_plot_detection_total_reads <- egg::ggarrange(plot_percent_reads_mapped_detection, total_reads_mapped_detection,plot_total_reads_detection, plot_strain_level_detection,
          ncol = 1,
          heights = c(0.15,0.15,0.15,2))

```



Test plot to add pangenome tree

```{r, eval=FALSE}

blank_plot <- ggplot() + theme_void()

test <-  egg::ggarrange(blank_plot, plot_percent_reads_mapped_detection, 
                        blank_plot,total_reads_mapped_detection,
                        blank_plot, plot_total_reads_detection, 
                        tree_3, plot_strain_level_detection,
          ncol = 2,
          nrow = 4,
          heights = c(0.1,0.1,0.1,1),
          widths = c(0.2, 1))

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/test.pdf",test, width = 16, height = 11)

```


Save plot and data:

```{r, eval=FALSE}

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/strain_detection_binary_samples_orderd_by_total_reads.pdf",Strain_plot_detection_total_reads, width = 13, height = 11)

```


##### 5.b. FIGURE: Binary detection plot - samples ordered by HC

```{r, eval=FALSE}

##### Sample order; need to cluster each site separately, then get order, and then concatenate the orders ###

# created vector with 5 characters
columns= c("order")

# pass this vector length to ncol parameter
# and nrow with 0
site_order_df = data.frame(matrix(nrow = 0, ncol = length(columns)))

# assign column names
colnames(site_order_df) = columns


site_list <- list("SUPP", "SUBP", "KG", "BM", "TD", "PT", "TH", "SV") # ignore HP since there is only one sample


for (oral_site in site_list) {
  
  
  oral_site_factor <- trimws(oral_site)
  
  melted_strain_detection$detection_binary <- as.numeric(as.character(melted_strain_detection$detection_binary))
                             
  # filter df by site
  site_df <- melted_strain_detection %>% 
    dplyr::filter(site == oral_site_factor) 
  
  # subset data frame for clustering species
  site_clustering_df <- site_df %>% 
    select(bins, layers, detection_binary)
  
  # pivot wider
  site_wide_df <- site_clustering_df %>% pivot_wider(names_from = layers, values_from = detection_binary)
  
  # make tibble into data frame object
  site_wide_df <- as.data.frame(site_wide_df)
  
  # rename row names to species IDs
  site_wide_df2 <- site_wide_df[,-1]
  rownames(site_wide_df2) <- site_wide_df[,1]
  
  # reformat to matrix
  site_wide_matrix <- as.matrix(site_wide_df2)
  
  # transpose data frame so that we can cluster samples based on euclidian distance and complete linkage
  trans_site_wide_matrix <- t(site_wide_matrix)
  
  # calculate Bray-Curtis distance among samples
  site_dist_df <- vegan::vegdist(trans_site_wide_matrix, method = "jaccard")
  
  #Replace na values with 0 using is.na()
  site_dist_df[site_dist_df=="NaN"]<-0
  
  # Hierarchical clustering using Complete Linkage
  site_hc <- hclust(site_dist_df, method = "complete" )
  
  # create dedrogram object
  site_dhc <- as.dendrogram(site_hc)
  
  # create rectangular lines object
  site_ddata <- dendro_data(site_dhc, type = "rectangle")
  
  site_samples_dend_plot <- ggplot(segment(site_ddata)) + 
    geom_segment(aes(x = x, y = y, xend = xend, yend = yend), linewidth = 1) + 
    coord_flip() + 
    scale_y_reverse(expand = c(0.2, 0)) +
    scale_x_reverse() +
    theme_classic() +
    theme(axis.ticks.y = element_blank(),
          axis.ticks.x=element_blank(),
          axis.text.x = element_blank(),
          axis.text.y =  element_blank(),
          axis.line.x = element_blank(),
          axis.line.y = element_blank(),
          axis.title.y = element_blank(),
          axis.title.x = element_blank()) +
    theme(panel.spacing.x=unit(0.05, "lines"),
          plot.margin=unit(c(0,0,0,0), "cm"))
  
  # save dend plot
  ggsave(paste0("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Dendrograms/Detection_",oral_site,"_samples_dendrogram_all_sites.pdf"),site_samples_dend_plot, width = 5, height = 7)
  
  
  # get list of sample order for site
  
  site_order <- data.frame(order = site_hc$labels[site_hc$order])
  
  site_order_df <- rbind(site_order_df, site_order)
  
}


# add HP sample (HP_HC_HMP_S0195_01) to the end of the order 
#melted_strain_detection$layers[melted_strain_detection$site == "HP"] 
site_order_df <- rbind(site_order_df, "HP_HC_HMP_S0195_01")

# reorder samples based on concatenated hclust order
samples_order <- melted_strain_detection$layers <- factor(melted_strain_detection$layers, levels = site_order_df$order)


#### PLOT #### 

dodge <- position_dodge(width=0.9)

total_reads_mapped_detection <- ggplot(data = melted_strain_detection, aes(x=layers, y=total_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 25000000), breaks=seq(0,25000000,by=25000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("No. of reads mapped", "(0 - 25 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)
#total_reads_mapped_detection

plot_total_reads_detection <- ggplot(data = melted_strain_detection, aes(x=layers, y=total_number_reads)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 200000000), breaks=seq(0,200000000,by=200000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Total reads", "(0 - 200 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.text.y=element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)
#plot_total_reads_detection


plot_percent_reads_mapped_detection <- ggplot(data = melted_strain_detection, aes(x=layers, y=percent_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 50), breaks=seq(0,50,by=50), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Percent reads mapped", "(0 - 50 %)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)
#plot_percent_reads_mapped_detection


# detection plot
plot_strain_level_detection_binary <- ggplot(data = melted_strain_detection, aes(x=layers, y=bins, fill=as.factor(detection_binary))) +
  geom_tile(height=1, width=1) + 
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") + 
  scale_fill_manual(values = c("black", "cyan"),
                    labels = c("Un-detected (Breadth < 50%)", "Detected (Breadth > 50%)"),
                    name = "Genome detection") +
  scale_y_discrete(limits = rev(levels(melted_strain_detection$bins))) +
  ylab(NULL) +  
  xlab(NULL) + 
  theme_classic() + 
  theme(text = element_text(size = 12),
        axis.ticks.y = element_line(color = rev(group_tick_colors), size = 1.2),
        axis.ticks.length = unit(1, "cm"),
        axis.text.x=element_blank(),
        axis.text.y =  element_text(size = 2, color = "black"),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "black")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        #panel.border = element_rect(colour = "black", fill = NA),
        strip.text=element_text(size=12)) + 
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,1,1), "cm")) +
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)
#plot_strain_level_detection_non_binary


Strain_plot_detection_binary_smaples_HC_order <- egg::ggarrange(plot_percent_reads_mapped_detection, plot_total_reads_detection, total_reads_mapped_detection, plot_strain_level_detection_binary,
          ncol = 1,
          heights = c(0.15,0.15,0.15,2))

```


```{r, eval=FALSE}

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/strain_detection_binary_samples_orderd_by_HC.pdf",Strain_plot_detection_binary_smaples_HC_order, width = 13, height = 11)

```


##### 5.c. FIGURE: Non-binary breadth of coverage heatmap V1 

samples ordered by no. of mapped reads

Genomes ordered by Pangenome gene cluster frequencies

```{r, eval=FALSE}

# Order genomes by pangenome tree
melted_strain_detection$bins <- factor(melted_strain_detection$bins, levels = tip_labels)

# Order tick colors
group_tick_colors$bins <- factor(group_tick_colors$bins, levels = tip_labels)

# Reorder the group_tick_colors dataframe to match the order of the bins
group_tick_colors <- group_tick_colors[order(group_tick_colors$bins), ]

# specify order for site terms
melted_strain_detection$site <- factor(melted_strain_detection$site, levels = c("SUPP", "BM", "TD", "SUBP", "KG","PT","TH", "SV","HP"))

Medium_panel_sizes <-c(10,10,10,2,1.5,2,1.75,0.75,0.5)

dodge <- position_dodge(width=0.9)

total_reads_mapped_detection <- ggplot(data = melted_strain_detection, aes(x=reorder(layers, -total_reads_mapped), y=total_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 25000000), breaks=seq(0,25000000,by=25000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("No. of reads mapped", "(0 - 25 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                 TRUE)
#total_reads_mapped_detection

plot_total_reads_detection <- ggplot(data = melted_strain_detection, aes(x=reorder(layers, -total_reads_mapped), y=total_number_reads)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 200000000), breaks=seq(0,200000000,by=200000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Total reads", "(0 - 200 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.text.y=element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                 TRUE)

plot_percent_reads_mapped_detection <- ggplot(data = melted_strain_detection, aes(x=reorder(layers, -total_reads_mapped), y=percent_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 50), breaks=seq(0,50,by=50), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Percent reads mapped", "(0 - 50 %)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                 TRUE)

# set colors for heat map
colors <- c("black", "#2166AC", "#D1E5F0", "white", "#FDDBC7", "#F4A582", "#FF0000")

colors <- c("black", "#2166AC", "white", "#F4A582", "#FF0000")


plot_strain_level_detection_non_binary <- ggplot(data = melted_strain_detection, aes(x=reorder(layers,-total_reads_mapped), y=bins, fill=value)) +
  geom_tile(height=1, width=1) + 
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") + 
  scale_fill_gradientn(colours = colors,
                       limits=c(0.1, 0.9),
                       breaks = c(0.9, 0.8, 0.7, 0.6, 0.5, 0.4, 0.3, 0.2, 0.1),
                       labels = c("> 90", "80", "70", "60", "50", "40", "30", "20", "< 10"),
                       oob=squish) +  
  scale_y_discrete(limits = rev(levels(melted_strain_detection$bins))) +
  ylab(NULL) +  
  xlab(NULL) + 
  theme_classic() + 
  guides(fill = guide_colourbar(barwidth = 1,
                                barheight = 10,
                                title = "Breadth of coverage (%)",
                                ticks = TRUE,
                                ticks.colour = "white",
                                ticks.linewidth = 0.5)) +
  theme(text = element_text(size = 12),
        axis.ticks.y = element_line(color = rev(group_tick_colors$Group_color), size = 1.2),
        axis.ticks.length = unit(1, "cm"),
        axis.text.x=element_blank(),
        axis.text.y =  element_text(size = 2, color = "black"),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "black")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text=element_text(size=12)) + 
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,1,1), "cm")) +
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                 TRUE)


Strain_plot_detection_non_binary_total_reads <- egg::ggarrange(plot_percent_reads_mapped_detection, total_reads_mapped_detection, plot_total_reads_detection, plot_strain_level_detection_non_binary,
          ncol = 1,
          heights = c(0.15,0.15,0.15,2))

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/strain_breadth_coverage_non_binary_heatmap_samples_ordered_by_no_mapped_reads.pdf",Strain_plot_detection_non_binary_total_reads,  width = 20, height = 10)

```

##### 5.d. FIGURE: Non-binary breadth of coverage - samples ordered by heirarchical clustering

```{r, eval=FALSE}

##### Sample order; need to cluster each site seperately, then get order, and then concatenate the orders ###

# created vector with 5 characters
columns= c("order")

# pass this vector length to ncol parameter
# and nrow with 0
site_order_df = data.frame(matrix(nrow = 0, ncol = length(columns)))

# assign column names
colnames(site_order_df) = columns


site_list <- list("SUPP", "SUBP", "KG", "BM", "TD", "PT", "TH", "SV") # ignore HP since there is only one sample


for (oral_site in site_list) {
  
  
  oral_site_factor <- trimws(oral_site)
                             
  # filter df by site
  site_df <- melted_strain_detection %>% 
    dplyr::filter(site == oral_site_factor) 
  
  # subset data frame for clustering species
  site_clustering_df <- site_df %>% 
    select(bins, layers, value)
  
  # pivot wider
  site_wide_df <- site_clustering_df %>% pivot_wider(names_from = layers, values_from = value)
  
  # make tibble into data frame object
  site_wide_df <- as.data.frame(site_wide_df)
  
  # rename row names to species IDs
  site_wide_df2 <- site_wide_df[,-1]
  rownames(site_wide_df2) <- site_wide_df[,1]
  
  # reformat to matrix
  site_wide_matrix <- as.matrix(site_wide_df2)
  
  # transpose data frame so that we can cluster samples based on euclidan distance and complete linkage
  trans_site_wide_matrix <- t(site_wide_matrix)
  
  # calculate Bray-Curtis distance among samples
  site_dist_df <- vegan::vegdist(trans_site_wide_matrix, method = "bray")
  
  #Replace na values with 0 using is.na()
  site_dist_df[site_dist_df=="NaN"]<-0
  
  # Hierarchical clustering using Complete Linkage
  site_hc <- hclust(site_dist_df, method = "complete" )
  
  # create dedrogram object
  site_dhc <- as.dendrogram(site_hc)
  
  # create rectangular lines object
  site_ddata <- dendro_data(site_dhc, type = "rectangle")
  
  site_samples_dend_plot <- ggplot(segment(site_ddata)) + 
    geom_segment(aes(x = x, y = y, xend = xend, yend = yend), linewidth = 1) + 
    coord_flip() + 
    scale_y_reverse(expand = c(0.2, 0)) +
    scale_x_reverse() +
    theme_classic() +
    theme(axis.ticks.y = element_blank(),
          axis.ticks.x=element_blank(),
          axis.text.x = element_blank(),
          axis.text.y =  element_blank(),
          axis.line.x = element_blank(),
          axis.line.y = element_blank(),
          axis.title.y = element_blank(),
          axis.title.x = element_blank()) +
    theme(panel.spacing.x=unit(0.1, "lines"),
          plot.margin=unit(c(0,0,0,0), "cm"))
  
  # save dend plot
  ggsave(paste0("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Dendrograms/Breadth_of_coverage_",oral_site,"_samples_dendrogram_all_sites.pdf"),site_samples_dend_plot, width = 5, height = 7)
  
  
  # get list of sample order for site
  
  site_order <- data.frame(order = site_hc$labels[site_hc$order])
  
  site_order_df <- rbind(site_order_df, site_order)
  
}


# add HP sample (HP_HC_HMP_S0195_01) to the end of the order 
#melted_strain_detection$layers[melted_strain_detection$site == "HP"] 
site_order_df <- rbind(site_order_df, "HP_HC_HMP_S0195_01")

# reorder samples based on concatenated hclust order
samples_order <- melted_strain_detection$layers <- factor(melted_strain_detection$layers, levels = site_order_df$order)


#### PLOT #### 

dodge <- position_dodge(width=0.9)

total_reads_mapped_detection <- ggplot(data = melted_strain_detection, aes(x=layers, y=total_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 25000000), breaks=seq(0,25000000,by=25000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("No. of reads mapped", "(0 - 25 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)
#total_reads_mapped_detection

plot_total_reads_detection <- ggplot(data = melted_strain_detection, aes(x=layers, y=total_number_reads)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 200000000), breaks=seq(0,200000000,by=200000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Total reads", "(0 - 200 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.text.y=element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)
#plot_total_reads_detection


plot_percent_reads_mapped_detection <- ggplot(data = melted_strain_detection, aes(x=layers, y=percent_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 50), breaks=seq(0,50,by=50), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Percent reads mapped", "(0 - 50 %)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)
#plot_percent_reads_mapped_detection



# set colors for heat map
paleta_new = rev(RColorBrewer::brewer.pal(n = 11, name = "RdYlBu"))

# Colors, including gray for the first segment
colors <- c("gray", paleta_new)

plot_strain_level_detection_non_binary <- ggplot(data = melted_strain_detection, aes(x=layers, y=bins, fill=value)) +
  geom_tile(height=1, width=1) + 
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") + 
  scale_fill_gradientn(colours = colors,
                       limits=c(0.005, 0.5),
                       #breaks=seq(0.005,0.5,by=0.1),
                       breaks = c(0.5, 0.4, 0.3, 0.2, 0.1, 0.005),
                       labels = c("> 50 %", "40 %", "30 %", "20 %", "10 %", "< 0.5 %"),
                       oob=squish) + 
  scale_y_discrete(limits = rev(levels(melted_strain_detection$bins))) +
  ylab(NULL) +  
  xlab(NULL) + 
  theme_classic() + 
  guides(fill = guide_colourbar(barwidth = 1,
                                barheight = 10,
                                title = "Breadth of coverage",
                                ticks = TRUE,
                                ticks.colour = "white",
                                ticks.linewidth = 0.5)) +
  theme(text = element_text(size = 12),
        axis.ticks.y = element_line(color = rev(group_tick_colors), size = 1.2),
        axis.ticks.length = unit(1, "cm"),
        axis.text.x=element_blank(),
        axis.text.y =  element_text(size = 2, color = "black"),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "black")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        #panel.border = element_rect(colour = "black", fill = NA),
        strip.text=element_text(size=12)) + 
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,1,1), "cm")) +
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)
#plot_strain_level_detection_non_binary


Strain_plot_detection_non_binary_smaples_HC_order <- egg::ggarrange(plot_percent_reads_mapped_detection, total_reads_mapped_detection, plot_total_reads_detection, plot_strain_level_detection_non_binary,
          ncol = 1,
          heights = c(0.15,0.15,0.15,2))

```


```{r, eval=FALSE}

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/strain_breadth_coverage_non_binary_heatmap_samples_ordered_by_HC.pdf",Strain_plot_detection_non_binary_smaples_HC_order, width = 13, height = 11)

```

##### 5.e. FIGURE: Non-binary breadth of coverage - genomes ordered by ANI, samples ordered by heirarchical clustering

```{r, eval=FALSE}


# group colors 
group_tick_colors <- bin_colors %>% 
  group_by(bins) %>% 
  select(bins, Group_color) %>% 
  distinct() %>% 
  mutate(bins = factor(bins, levels = ANI_tip_labels)) %>%
  arrange(bins)

##### Sample order; need to cluster each site seperately, then get order, and then concatenate the orders ###

# created vector with 5 characters
columns= c("order")

# pass this vector length to ncol parameter
# and nrow with 0
site_order_df = data.frame(matrix(nrow = 0, ncol = length(columns)))

# assign column names
colnames(site_order_df) = columns


site_list <- list("SUPP", "SUBP", "KG", "BM", "TD", "PT", "TH", "SV") # ignore HP since there is only one sample


for (oral_site in site_list) {
  
  
  oral_site_factor <- trimws(oral_site)
  
  # filter df by site
  site_df <- melted_strain_detection %>% 
    dplyr::filter(site == oral_site_factor) 
  
  # subset data frame for clustering species
  site_clustering_df <- site_df %>% 
    select(bins, layers, value)
  
  # pivot wider
  site_wide_df <- site_clustering_df %>% pivot_wider(names_from = layers, values_from = value)
  
  # make tibble into data frame object
  site_wide_df <- as.data.frame(site_wide_df)
  
  # rename row names to species IDs
  site_wide_df2 <- site_wide_df[,-1]
  rownames(site_wide_df2) <- site_wide_df[,1]
  
  # reformat to matrix
  site_wide_matrix <- as.matrix(site_wide_df2)
  
  # transpose data frame so that we can cluster samples based on euclidan distance and complete linkage
  trans_site_wide_matrix <- t(site_wide_matrix)
  
  # calculate Bray-Curtis distance among samples
  site_dist_df <- vegan::vegdist(trans_site_wide_matrix, method = "bray")
  
  #Replace na values with 0 using is.na()
  site_dist_df[site_dist_df=="NaN"]<-0
  
  # Hierarchical clustering using Complete Linkage
  site_hc <- hclust(site_dist_df, method = "complete" )
  
  # create dedrogram object
  site_dhc <- as.dendrogram(site_hc)
  
  # create rectangular lines object
  site_ddata <- dendro_data(site_dhc, type = "rectangle")
  
  site_samples_dend_plot <- ggplot(segment(site_ddata)) + 
    geom_segment(aes(x = x, y = y, xend = xend, yend = yend), linewidth = 1) + 
    coord_flip() + 
    scale_y_reverse(expand = c(0.2, 0)) +
    scale_x_reverse() +
    theme_classic() +
    theme(axis.ticks.y = element_blank(),
          axis.ticks.x=element_blank(),
          axis.text.x = element_blank(),
          axis.text.y =  element_blank(),
          axis.line.x = element_blank(),
          axis.line.y = element_blank(),
          axis.title.y = element_blank(),
          axis.title.x = element_blank()) +
    theme(panel.spacing.x=unit(0.1, "lines"),
          plot.margin=unit(c(0,0,0,0), "cm"))
  
  # save dend plot
  #ggsave(paste0("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Dendrograms/Breadth_of_coverage_",oral_site,"_samples_dendrogram_all_sites.pdf"),site_samples_dend_plot, width = 5, height = 7)
  
  
  # get list of sample order for site
  
  site_order <- data.frame(order = rev(site_hc$labels[site_hc$order]))
  
  site_order_df <- rbind(site_order_df, site_order)
  
}


# add HP sample (HP_HC_HMP_S0195_01) to the end of the order 
#melted_strain_detection$layers[melted_strain_detection$site == "HP"] 
site_order_df <- rbind(site_order_df, "HP_HC_HMP_S0195_01")

# reorder samples based on concatenated hclust order
samples_order <- melted_strain_detection$layers <- factor(melted_strain_detection$layers, levels = site_order_df$order)


#### PLOT #### 

dodge <- position_dodge(width=0.9)

total_reads_mapped_detection <- ggplot(data = melted_strain_detection, aes(x=layers, y=total_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 25000000), breaks=seq(0,25000000,by=25000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("No. of reads mapped", "(0 - 25 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                   TRUE)
#total_reads_mapped_detection

plot_total_reads_detection <- ggplot(data = melted_strain_detection, aes(x=layers, y=total_number_reads)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 200000000), breaks=seq(0,200000000,by=200000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Total reads", "(0 - 200 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.text.y=element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                   TRUE)
#plot_total_reads_detection


plot_percent_reads_mapped_detection <- ggplot(data = melted_strain_detection, aes(x=layers, y=percent_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 50), breaks=seq(0,50,by=50), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Percent reads mapped", "(0 - 50 %)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                   TRUE)
#plot_percent_reads_mapped_detection



# set colors for heat map
paleta_new = rev(RColorBrewer::brewer.pal(n = 11, name = "RdBu"))

# Colors, including gray for the first segment
colors <- c("black", paleta_new)

plot_strain_level_detection_non_binary <- ggplot(data = melted_strain_detection, aes(x=layers, y=bins, fill=value)) +
  geom_tile(height=1, width=1) + 
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") + 
  scale_fill_gradientn(colours = colors,
                       limits=c(0.1, 0.9),
                       #breaks=seq(0.005,0.5,by=0.1),
                       breaks = c(0.9, 0.8, 0.7, 0.6, 0.5, 0.4, 0.3, 0.2, 0.1),
                       labels = c("> 90 %", "80 %", "70 %", "60 %", "50 %", "40 %", "30 %", "20 %", "< 10 %"),
                       oob=squish) + 
  scale_y_discrete(limits = levels(melted_strain_detection$bins)) +
  ylab(NULL) +  
  xlab(NULL) + 
  theme_classic() + 
  guides(fill = guide_colourbar(barwidth = 1,
                                barheight = 10,
                                title = "Breadth of coverage",
                                ticks = TRUE,
                                ticks.colour = "white",
                                ticks.linewidth = 0.5)) +
  theme(text = element_text(size = 12),
        axis.ticks.y = element_line(color = group_tick_colors$Group_color, size = 1.2),
        axis.ticks.length = unit(1, "cm"),
        axis.text.x=element_blank(),
        axis.text.y =  element_text(size = 2, color = "black"),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "black")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        #panel.border = element_rect(colour = "black", fill = NA),
        strip.text=element_text(size=12)) + 
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,1,1), "cm")) +
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                   TRUE)
#plot_strain_level_detection_non_binary
plot_strain_level_detection_non_binary

Strain_plot_detection_non_binary_smaples_HC_order <- egg::ggarrange(plot_percent_reads_mapped_detection, total_reads_mapped_detection, plot_total_reads_detection, plot_strain_level_detection_non_binary,
                                                                    ncol = 1,
                                                                    heights = c(0.15,0.15,0.15,2))

```


```{r, eval=FALSE}

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/strain_breadth_coverage_non_binary_heatmap_samples_ordered_by_HC.pdf",Strain_plot_detection_non_binary_smaples_HC_order, width = 13, height = 11)

```

# 6. Mean depth of coverage (Q2Q3)

##### 6.a. Oraganize coverage data 
 

```{r, eval=FALSE}

# convert coverage data frame from wide to long format
melted_df_coverage <- melt(df_coverage)

# make oral site variable for facet grid
melted_df_coverage_2 <- melted_df_coverage %>% mutate(site = case_when(grepl('TD_', variable) ~ 'TD', 
                                      grepl('KG_', variable) ~ 'KG', 
                                      grepl('PP_', variable) ~ 'SUPP',
                                      grepl('PB_', variable) ~ 'SUBP',
                                      grepl('SA_', variable) ~ 'SV',
                                      grepl('TH_', variable) ~ 'TH',
                                      grepl('PT_', variable) ~ 'PT',
                                      grepl('HP_', variable) ~ 'HP',
                                      grepl('BM_', variable) ~ 'BM')) 


# specify order for site terms
melted_df_coverage_2$site <- factor(melted_df_coverage_2$site, levels = c("SUPP", "SUBP", "KG", "BM", "TD", "PT", "HP", "TH", "SV"))

# sort "variable" based on pre-sorted values of "site"
melted_df_coverage_2 <- melted_df_coverage_2 %>%
  arrange(site, variable) %>%               # sort your dataframe
  mutate(variable = factor(variable, unique(variable))) # reset your factor-column based on that order

# join metadata with to get total reads mapped into data frame
melted_strain_coverage = merge(mapping_metadata_merged, melted_df_coverage_2, by.x = "layers", by.y = "variable") 

# reorder genomes based on pangenome order
melted_strain_coverage$bins <- factor(melted_strain_coverage$bins, levels = tip_labels)


```

Save organized mean coverage data
```{r, eval=FALSE}

write.csv(melted_strain_coverage,"/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Strain_level_updated_Q2Q3_mean_coverage_data.csv", row.names = FALSE, quote = FALSE )

```


#### 6.b. FIGURE: Q2Q3 mean depth of coverage heatmap - Strain-level

```{r, eval=FALSE}

dodge <- position_dodge(width=0.9)

total_reads_mapped_coverage<- ggplot(data = melted_strain_coverage, aes(x=reorder(layers, -total_number_reads), y=total_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 25000000), breaks=seq(0,25000000,by=25000000), oob=squish, labels = scales::comma) +
  xlab(NULL)+
  ylab(expression(atop("No. of reads mapped", "(0 - 25 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)
#total_reads_mapped_detection

plot_total_reads_coverage <- ggplot(data = melted_strain_coverage, aes(x=reorder(layers, -total_number_reads), y=total_number_reads)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 200000000), breaks=seq(0,200000000,by=200000000), oob=squish, labels = scales::comma) +
  xlab(NULL)+
  ylab(expression(atop("Total reads", "(0 - 200 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.text.y=element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)
#plot_total_reads_detection


plot_percent_reads_mapped_coverage <- ggplot(data = melted_strain_coverage, aes(x=reorder(layers, -total_number_reads), y=percent_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 50), breaks=seq(0,50,by=50), oob=squish, labels = scales::comma) +
  xlab(NULL)+
  ylab(expression(atop("Percent reads mapped", "(0 - 50 %)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)
#plot_percent_reads_mapped_detection



# set colors for heat map
paleta_new = rev(RColorBrewer::brewer.pal(n = 11, name = "RdYlBu"))

# Colors, including gray for the first segment
colors <- c("gray", paleta_new)

plot_strain_level_coverage <- ggplot(data = melted_strain_coverage, aes(x=reorder(layers,-total_number_reads), y=bins, fill=value)) +
  geom_tile(height=1, width=1) + 
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") + 
  scale_fill_gradientn(colours = colors,
                       limits=c(1, 10),
                       #breaks=seq(0.005,0.5,by=0.1),
                       breaks = c(1, 2, 4, 6, 8, 10),
                       labels = c("< 1 X", "2 X", "4 X", "6 X", "8 X", "> 10 X"),
                       oob=squish) + 
  scale_y_discrete(limits = rev(levels(melted_strain_coverage$bins))) +
  ylab(NULL) +  
  xlab(NULL) + 
  theme_classic() + 
  guides(fill = guide_colourbar(barwidth = 1,
                                barheight = 10,
                                title = expression(atop("Mean depth of coverage", "(Q2Q3)")),
                                ticks = TRUE,
                                ticks.colour = "white",
                                ticks.linewidth = 0.5)) +
  theme(text = element_text(size = 12),
        axis.ticks.y = element_line(color = rev(group_tick_colors), size = 1),
        axis.ticks.length = unit(1, "cm"),
        axis.text.x=element_blank(),
        axis.text.y =  element_text(size = 2, color = "black"),
        axis.line.y = element_blank(),
        axis.line.x = element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "black")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text=element_text(size=12)) + 
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,1,1), "cm")) +
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)
#plot_strain_level_coverage


Strain_plot_coverage_sorted_by_total_reads <- egg::ggarrange(plot_percent_reads_mapped_coverage, total_reads_mapped_coverage, plot_total_reads_coverage, plot_strain_level_coverage,
          ncol = 1,
          heights = c(0.1,0.1,0.1,1))

```
Save plot and data:

```{r}

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/strain_Q2Q3_depth_heatmap.pdf",Strain_plot_coverage_sorted_by_total_reads, width = 13, height = 11)


```

# 7. Relative abundance 

##### 7.a. Organize data

```{r, eval=FALSE}


# Load data

# Load csv file with genome IDs and colors; this file was made when the tanglegram was made
bins_and_colors <-read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/pangenome_custom_gene_freq_order.csv", header = TRUE)

# Load breadth data
melted_df_detect <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Strain_level_updated_detection_data.csv", header = TRUE)

# Load mean Q2Q3 depth of coverage data
melted_df_coverage <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Strain_level_updated_Q2Q3_mean_coverage_data.csv", header = TRUE)


# new species IDs based detection profiles
group_IDs=as.factor(bins_and_colors$Pangenome_group)

# new species colors based detection profiles
group_tick_colors=bins_and_colors$Group_color

# TO EXCLUDE UN_DETECTED GENOMES
melted_df_detect_filtered <- melted_df_detect %>% 
  filter(detected == "detected")
filtered_melted_df <-merge(melted_df_coverage, melted_df_detect_filtered, by=c("layers","bins")) 
filtered_melted_df <- filtered_melted_df %>% dplyr::rename(coverage=value.x, breadth=value.y)

# select columns and transpose from long to wide
meanCov <- filtered_melted_df %>% 
  dplyr::select(layers, bins, coverage) %>% 
  pivot_wider(id_cols=bins, names_from=layers, values_from=coverage)

# chnage NA values to zeros
meanCov[is.na(meanCov)] <- 0

# transpose data frame
rownames <- meanCov$bins
meanCovT <- as.data.frame(t(meanCov[-1]))
colnames(meanCovT) <- rownames

# mean coverage of a genome / sum of mean coverages of all genomes in a sample
relAbsT <- decostand(meanCovT, method = "total")

# transpose relative abundance matrix
relAbs <- t(relAbsT)

# multiply by 100
relAbs <- 100*relAbs 

# change coverage data frame from wide to long format
melted_relAbs <- reshape2::melt(relAbs) %>% 
  dplyr::rename(variable=Var2, Genome_ID=Var1, relAbundance=value)

# merge with relabundance
melted_relAbs_merged <- merge(melted_relAbs,bins_and_colors, by.x = 'Genome_ID', by.y = 'Pangenome_ID')

# set Pangenome_group ID as factor
melted_relAbs_merged$Pangenome_group <- as.factor(melted_relAbs_merged$Pangenome_group)

# rename metagenomes to just site name
#melted_relAbs_merged$variable=gsub("_HC_HMP","",melted_relAbs_merged$variable)

# make oral site variable for facet grid
melted_relAbs_merged <- melted_relAbs_merged %>% mutate(site = case_when(grepl('TD_', variable) ~ 'TD', 
                                      grepl('KG_', variable) ~ 'KG', 
                                      grepl('PP_', variable) ~ 'SUPP',
                                      grepl('PB_', variable) ~ 'SUBP',
                                      grepl('SA_', variable) ~ 'SV',
                                      grepl('TH_', variable) ~ 'TH',
                                      grepl('PT_', variable) ~ 'PT',
                                      grepl('HP_', variable) ~ 'HP',
                                      grepl('BM_', variable) ~ 'BM')) 

# Save relative abundance data for genomes
write.csv(melted_relAbs_merged, "/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Relative_abundance_Neisseriaceae_genomes.csv", row.names = FALSE, quote = FALSE)

#Sum abundance by oral site, groupID and variable to obtain the relative abundance of each species (aka groupID) for each metagenome (aka variable)
melted_relAbs_merged_summed <- melted_relAbs_merged %>% 
  group_by(site, Pangenome_group, variable) %>% 
  dplyr::summarise(sum = sum(relAbundance))
  

 # merge total reads mapped into data frame
melted_relAbs_merged_summed = merge.data.frame(mapping_metadata_merged, melted_relAbs_merged_summed, by.x = "layers", by.y = "variable") 

# reorder oral sites
melted_relAbs_merged_summed$site <- factor(melted_relAbs_merged_summed$site, levels = c("SUPP", "SUBP", "KG", "BM", "TD", "PT", "HP", "TH", "SV"))


```

##### 7.b. Order species groups and samples by heirarchical clustering 

```{r, eval=FALSE}

# subset data frame for clustering species
melted_relAbs_merged_summed_clustering_df <- melted_relAbs_merged_summed %>% 
  dplyr::select(Pangenome_group, layers, sum)

# pivot wider
wide_df <- melted_relAbs_merged_summed_clustering_df %>% pivot_wider(names_from = layers, values_from = sum)

# make tibble into data frame object
wide_df <- as.data.frame(wide_df)

# rename row names to species IDs
wide_df2 <- wide_df[,-1]
rownames(wide_df2) <- wide_df[,1]
# add rows of zeroes for undetected species groups
  
#N_meningitidis_gonorrhoeae_group
#N_polysaccharea_group
#K_potus
#Non_oral_Neisseria_spp
#E_glucosivorans

  
N_meningitidis_gonorrhoeae_group <- t(data.frame("N_meningitidis_gonorrhoeae_group" = seq(from = 0,to = 0, length.out = (ncol(wide_df)) - 1)))
colnames(N_meningitidis_gonorrhoeae_group) <- colnames(wide_df2)

N_polysaccharea_group <- t(data.frame("N_polysaccharea_group" = seq(from = 0,to = 0, length.out = (ncol(wide_df)) - 1)))
colnames(N_polysaccharea_group) <- colnames(wide_df2)

K_potus <- t(data.frame("K_potus" = seq(from = 0,to = 0, length.out = (ncol(wide_df)) - 1)))
colnames(K_potus) <- colnames(wide_df2)

Non_oral_Neisseria_spp <- t(data.frame("Non_oral_Neisseria_spp" = seq(from = 0,to = 0, length.out = (ncol(wide_df)) - 1)))
colnames(Non_oral_Neisseria_spp) <- colnames(wide_df2)

E_glucosivorans <- t(data.frame("E_glucosivorans" = seq(from = 0,to = 0, length.out = (ncol(wide_df)) - 1)))
colnames(E_glucosivorans) <- colnames(wide_df2)

wide_df2 <- rbind(wide_df2, N_meningitidis_gonorrhoeae_group, N_polysaccharea_group, K_potus, Non_oral_Neisseria_spp, E_glucosivorans)  

wide_matrix <- as.matrix(wide_df2)

#create distance matrix for species  based on aitchison distances
#dist_df <- coda.base::dist(wide_matrix, method = "aitchison")

# calculate Bray-Curtis distance among samples
dist_df <- vegan::vegdist(wide_matrix, method = "bray")


#Replace na values with 0 using is.na()
dist_df[dist_df=="NaN"]<-0
  
# Hierarchical clustering using Complete Linkage
hc <- hclust(dist_df, method = "complete" )

# create dedrogram object
dhc <- as.dendrogram(hc)

# create rectangular lines object
ddata <- dendro_data(dhc, type = "rectangle")

species_groups_dend_plot <- ggplot(segment(ddata)) + 
  geom_segment(aes(x = x, y = y, xend = xend, yend = yend), linewidth = 1) + 
  coord_flip() + 
  scale_y_reverse(expand = c(0.2, 0)) +
  scale_x_reverse() +
  theme_classic() +
  theme(axis.ticks.y = element_blank(),
        axis.ticks.x=element_blank(),
        axis.text.x = element_blank(),
        axis.text.y =  element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.title.y = element_blank(),
        axis.title.x = element_blank()) +
  theme(panel.spacing.x=unit(0.05, "lines"),
        plot.margin=unit(c(0,0,0,0), "cm"))

# save dend plot
ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Dendrograms/Relative_abundance_species_groups_dendrogram_all_sites.pdf",species_groups_dend_plot, width = 5, height = 7)



# reorder genomes for triples plot based on hclust
species_order <- hc$labels[hc$order]
melted_relAbs_merged_summed$Pangenome_group <- factor(melted_relAbs_merged_summed$Pangenome_group, levels = species_order)


##### Sample order; need to cluster each site seperately, then get order, and then concatenate the orders ###

# created vector with 5 characters
columns= c("order")

# pass this vector length to ncol parameter
# and nrow with 0
site_order_df = data.frame(matrix(nrow = 0, ncol = length(columns)))

# assign column names
colnames(site_order_df) = columns


site_list <- list("SUPP", "SUBP", "KG", "BM", "TD", "PT", "TH", "SV") # ignore HP since there is only one sample


for (oral_site in site_list) {
  
  oral_site_factor <- trimws(oral_site)
  
  # filter df by site
  site_df <- melted_relAbs_merged_summed %>% 
    filter(site == oral_site_factor)
  
  # subset data frame for clustering species
  site_clustering_df <- site_df %>% 
    dplyr::select(Pangenome_group, layers, sum)
  
  # pivot wider
  site_wide_df <- site_clustering_df %>% pivot_wider(names_from = layers, values_from = sum)
  
  # make tibble into data frame object
  site_wide_df <- as.data.frame(site_wide_df)
  
  # rename row names to species IDs
  site_wide_df2 <- site_wide_df[,-1]
  rownames(site_wide_df2) <- site_wide_df[,1]
  
  # add rows of zeroes for undetected species
  #N_meningitidis_gonorrhoeae_group
#N_polysaccharea_group
#K_potus
#Non_oral_Neisseria_spp
#E_glucosivorans
  
  N_meningitidis_gonorrhoeae_group <- t(data.frame("N_meningitidis_gonorrhoeae_group" = seq(from = 0,to = 0, length.out = (ncol(site_wide_df)) - 1)))
  colnames(N_meningitidis_gonorrhoeae_group) <- colnames(site_wide_df2)
  
  N_polysaccharea_group <- t(data.frame("N_polysaccharea_group" = seq(from = 0,to = 0, length.out = (ncol(site_wide_df)) - 1)))
  colnames(N_polysaccharea_group) <- colnames(site_wide_df2)
  
  K_potus <- t(data.frame("K_potus" = seq(from = 0,to = 0, length.out = (ncol(site_wide_df)) - 1)))
  colnames(K_potus) <- colnames(site_wide_df2)
  
  Non_oral_Neisseria_spp <- t(data.frame("Non_oral_Neisseria_spp" = seq(from = 0,to = 0, length.out = (ncol(site_wide_df)) - 1)))
  colnames(Non_oral_Neisseria_spp) <- colnames(site_wide_df2)
  
  E_glucosivorans <- t(data.frame("E_glucosivorans" = seq(from = 0,to = 0, length.out = (ncol(site_wide_df)) - 1)))
  colnames(E_glucosivorans) <- colnames(site_wide_df2)
  
  site_wide_df2 <- rbind(site_wide_df2, N_meningitidis_gonorrhoeae_group, N_polysaccharea_group, K_potus, Non_oral_Neisseria_spp, E_glucosivorans)  
  
  site_wide_matrix <- as.matrix(site_wide_df2)
  
  # transpose data frame so that we can cluster samples based on euclidan distance and complete linkage
  trans_site_wide_matrix <- t(site_wide_matrix)
  
  # calculate Bray-Curtis distance among samples
  site_dist_df <- vegan::vegdist(trans_site_wide_matrix, method = "bray")
  
  #Replace na values with 0 using is.na()
  site_dist_df[site_dist_df=="NaN"]<-0
  
  # Hierarchical clustering using Complete Linkage
  site_hc <- hclust(site_dist_df, method = "complete" )
  
  # create dedrogram object
  site_dhc <- as.dendrogram(site_hc)
  
  # create rectangular lines object
  site_ddata <- dendro_data(site_dhc, type = "rectangle")
  
  site_samples_dend_plot <- ggplot(segment(site_ddata)) + 
    geom_segment(aes(x = x, y = y, xend = xend, yend = yend), linewidth = 1) + 
    coord_flip() + 
    scale_y_reverse(expand = c(0.2, 0)) +
    scale_x_reverse() +
    theme_classic() +
    theme(axis.ticks.y = element_blank(),
          axis.ticks.x=element_blank(),
          axis.text.x = element_blank(),
          axis.text.y =  element_blank(),
          axis.line.x = element_blank(),
          axis.line.y = element_blank(),
          axis.title.y = element_blank(),
          axis.title.x = element_blank()) +
    theme(panel.spacing.x=unit(0.05, "lines"),
          plot.margin=unit(c(0,0,0,0), "cm"))
  
  # save dend plot
  ggsave(paste0("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Dendrograms/Relative_abundance_",oral_site,"_samples_dendrogram_all_sites.pdf"),site_samples_dend_plot, width = 5, height = 7)
  
  
  # get list of sample order for site
  
  site_order <- data.frame(order = site_hc$labels[site_hc$order])
  
  site_order_df <- rbind(site_order_df, site_order)
  
}


# add HP sample to the end of the order 
melted_relAbs_merged_summed$layers[melted_relAbs_merged_summed$site == "HP"] 
site_order_df <- rbind(site_order_df, "HP_HC_HMP_S0195_01")

# reorder samples based on concatenated hclust order
samples_order <- melted_relAbs_merged_summed$layers <- factor(melted_relAbs_merged_summed$layers, levels = site_order_df$order)

```


 Edit rel ab data so that un-detected species groups have zeroes for relative abundance values..
 
```{r, eval=FALSE}
# Assuming melted_relAbs_merged_summed is your data frame
melted_relAbs_merged_summed_copy <- melted_relAbs_merged_summed

# Create a complete data frame with all combinations
complete_data <- expand.grid(Pangenome_group = levels(melted_relAbs_merged_summed$Pangenome_group), layers = unique(melted_relAbs_merged_summed$layers)) # Add other variables as needed

# Merge with original data
full_data <- merge(complete_data, melted_relAbs_merged_summed_copy, by = c("Pangenome_group", "layers"), all.x = TRUE)

# Replace NA in 'sum' with -1
full_data$sum[is.na(full_data$sum)] <- 0

# select columns that we need
full_data_select <- full_data %>% 
  dplyr::select(Pangenome_group, layers, sum)

# make site variable for new dataframe
full_data_select <- full_data_select %>% mutate(site = case_when(grepl('TD_', layers) ~ 'TD', 
                                      grepl('KG_', layers) ~ 'KG', 
                                      grepl('PP_', layers) ~ 'SUPP',
                                      grepl('PB_', layers) ~ 'SUBP',
                                      grepl('SA_', layers) ~ 'SV',
                                      grepl('TH_', layers) ~ 'TH',
                                      grepl('PT_', layers) ~ 'PT',
                                      grepl('HP_', layers) ~ 'HP',
                                      grepl('BM_', layers) ~ 'BM')) 

# reorder oral sites
full_data_select$site <- factor(full_data_select$site, levels = c("SUPP", "SUBP", "KG", "BM", "TD", "PT", "HP", "TH", "SV"))


```



##### 7.c. FIGURE: Relative Abundance heatmap

```{r, eval=FALSE}


dodge <- position_dodge(width=0.9)

total_reads_mapped_rel_ab <- ggplot(data = melted_relAbs_merged_summed, aes(x=layers, y=total_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 25000000), breaks=seq(0,25000000,by=25000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("No. of reads mapped", "(0 - 25 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)


plot_total_reads_rel_ab <- ggplot(data = melted_relAbs_merged_summed, aes(x=layers, y=total_number_reads)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 200000000), breaks=seq(0,200000000,by=200000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Total reads", "(0 - 200 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.text.y=element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)

plot_percent_reads_mapped_rel_ab <- ggplot(data = melted_relAbs_merged_summed, aes(x=layers, y=percent_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 50), breaks=seq(0,50,by=50), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Percent reads mapped", "(0 - 50 %)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)



######### Relative abundance plot ######### 

# set colors for heat map
paleta_new = rev(RColorBrewer::brewer.pal(n = 11, name = "RdYlBu"))

# Colors, including gray for the first segment
colors <- c("gray", paleta_new)


plot_group_level_rel_ab <- ggplot(data = full_data_select, aes(x=layers, y=Pangenome_group, fill=sum)) +
  geom_tile(height=1, width=1) + 
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") + 
  scale_fill_gradientn(colours = colors,
                       limits=c(0, 100),
                       #breaks=seq(0.005,0.5,by=0.1),
                       breaks = c(0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100),
                       labels = c("0 %", "10 %", "20 %", "30 %", "40 %", "50 %", "60 %", "70 %", "80 %", "90 %", "100 %"),
                       oob=squish) + 
  scale_y_discrete(limits = levels(full_data_select$Pangenome_group)) +
  ylab(NULL) +  
  xlab(NULL) + 
  theme_classic() + 
  guides(fill = guide_colourbar(barwidth = 1,
                                barheight = 10,
                                title = "Relative abundance",
                                ticks = TRUE,
                                ticks.colour = "white",
                                ticks.linewidth = 0.5)) +
  theme(text = element_text(size = 12),
        #axis.ticks.y = element_line(color = rev(group_tick_colors), size = 1.2),
        axis.ticks.y = element_blank(),
        #axis.ticks.length = unit(1, "cm"),
        axis.text.x=element_blank(),
        axis.text.y =  element_text(size = 12, color = "black"),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "black")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        #panel.border = element_rect(colour = "black", fill = NA),
        strip.text=element_text(size=12)) + 
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,1,1), "cm")) +
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)


Group_level_plot_rel_ab <- egg::ggarrange(plot_percent_reads_mapped_rel_ab, total_reads_mapped_rel_ab, plot_total_reads_rel_ab, plot_group_level_rel_ab,
          ncol = 1,
          heights = c(0.1,0.1,0.1,1))
```


```{r, eval=FALSE}

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Relative_abundance_of_detected_species_groups.pdf", Group_level_plot_rel_ab, width = 14, height = 8)

```


# 8. Breadth of coverage density plots


```{r, eval=FALSE}

library(ggplot2)
library(ggridges)
library(dplyr)

# Assuming 'melted_strain_detection' is your original dataset
# Step 1: Calculate the sum of total_number_reads for each layer
layer_reads <- melted_strain_detection %>% 
  filter(site == "SUPP") %>% 
  group_by(layers) %>% 
  summarize(total_reads = sum(total_number_reads, na.rm = TRUE)) %>% 
  arrange(desc(total_reads))

# Step 2: Identify the top 10 layers based on total reads
top_layers <- layer_reads %>% 
  top_n(10, total_reads) %>% 
  pull(layers)

# Step 3: Filter the original dataset for these top layers
Breadth_density_data <- melted_strain_detection %>%
  filter(site == "SUPP", layers %in% top_layers)

# Plot
plot <- ggplot(Breadth_density_data, aes(x = value, y = layers, fill = after_stat(x))) +
  geom_density_ridges_gradient(scale = 3, rel_min_height = 0.001, size = 0.4) +
  scale_fill_viridis_c(name = "Breadth (%)", option = "C") +
  scale_x_continuous(expand = c(0, 0), limits = c(0, 1)) +
  scale_y_discrete(expand = expansion(mult = c(0.01, .7))) +
  xlab(NULL) + 
  ylab(NULL) + 
  theme_ridges(center_axis_labels = TRUE) +
  theme(legend.position="none")

# Display the plot
print(plot)


```



Group by sites - 16827 rows of data, thus each vertical line in the density plot for a site represents a sample and genome combination. 

```{r, eval=FALSE}

# Assuming 'melted_strain_detection' is your original dataset
# Step 1: Calculate the sum of total_number_reads for each layer
layer_reads <- melted_strain_detection %>% 
  group_by(site, layers) %>% 
  summarize(total_reads = sum(total_number_reads, na.rm = TRUE)) %>% 
  arrange(desc(total_reads))

# Step 2: Identify the top 10 layers based on total reads
top_layers <- layer_reads %>% 
  top_n(10, total_reads) %>% 
  pull(layers)

# Step 3: Filter the original dataset for these top layers
Breadth_density_data <- melted_strain_detection %>%
  filter(layers %in% top_layers)

# Plot
plot <- ggplot(Breadth_density_data, aes(x = value, y = site, fill = after_stat(x))) +
  geom_density_ridges_gradient(scale = 5, rel_min_height = 0, size = 0.4) +
  scale_fill_viridis_c(name = "Breadth (%)", option = "C") +
  scale_x_continuous(expand = c(0, 0), limits = c(0, 1)) +
  #scale_y_discrete(expand = expansion(mult = c(0.01, 0.02))) +
  xlab(NULL) + 
  ylab(NULL) + 
  theme_ridges(center_axis_labels = TRUE) +
  theme(legend.position="none")

# Display the plot
print(plot)


```


For each genome, group by sites: thus each vertical line in the density plot for a site represents a sample.


```{r, eval=FALSE}

# Create list of genome IDs

genomes <- melted_strain_detection %>% 
  group_by(bins) %>% 
  summarise(n = n()) %>% 
  droplevels() %>% 
  select(bins)

for (genome in genomes$bins) {
  
      # # Assuming 'melted_strain_detection' is your original dataset
      # # Step 1: Calculate the sum of total_number_reads for each layer
      # layer_reads <- melted_strain_detection %>% 
      #   group_by(site, layers) %>% 
      #   summarize(total_reads = sum(total_number_reads, na.rm = TRUE)) %>% 
      #   arrange(desc(total_reads))
      # 
      # # Step 2: Identify the top 10 layers based on total reads
      # top_layers <- layer_reads %>% 
      #   top_n(10, total_reads) %>% 
      #   pull(layers)
      
      # Step 3: Filter the original dataset for these top layers and the choice genome
      Breadth_density_data <- melted_strain_detection %>%
        #filter(layers %in% top_layers) %>% 
        filter(bins ==  genome)
      
      # Plot
      plot <- ggplot(Breadth_density_data, aes(x = value, y = site, fill = after_stat(x))) +
        geom_density_ridges_gradient(rel_min_height = 0, size = 0.4, alpha = 0.7,
                                 jittered_points = TRUE,
                                 position = position_points_jitter(width = 0.0001, height = 0),
                                 point_shape = '|', point_size = 1, point_alpha = 1) +
        scale_fill_viridis_c(name = "Breadth (%)", option = "C") +
        scale_x_continuous(expand = c(0, 0), limits = c(0, 1)) +
        xlab(NULL) + 
        ylab(NULL) + 
        theme_ridges(center_axis_labels = TRUE) 
      
     # save plot
        ggsave(paste0("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Breadth_densities_per_genome/",genome,"_ridgeline_plot.pdf"), plot = plot, width = 7, height = 5)
        
        
}



ggplot(Breadth_density_data, aes(x = value, y = site, fill = factor(stat(quantile)))) + stat_density_ridges(
    geom = "density_ridges_gradient", calc_ecdf = TRUE,
    quantiles = 4, quantile_lines = TRUE) +
  scale_fill_viridis_d(name = "Quartiles")


ggplot(Breadth_density_data, aes(x = value, y = site)) + geom_density_ridges(jittered_points = TRUE)

```


## Species-level average breadth density ridgeline plots

Breadth values per site include all Breadth values per sample for all genomes in species group.

```{r}

library(ggplot2)
library(ggridges)
library(dplyr)

plot_path = "/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Breadth_densities_per_species/"

####
Breadth_coverage_original <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Strain_level_updated_detection_data.csv", header = TRUE)

Final_Groups <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/TESTS/Classifying_genomes_to_oral_sites/Neisseriaceae_genome_groups.txt", header = TRUE, sep = "\t")
 
# select relevant columns
Breadth_original_selected <- Breadth_coverage_original %>% 
  dplyr::select(layers, site, bins, value)

# Merge with species groups info
Breadth_original_selected_with_groups <- merge(Breadth_original_selected, Final_Groups, by.x = "bins", by.y = "Pangenome_ID")


Breadth_original_selected_with_groups$site <- as.factor(Breadth_original_selected_with_groups$site)
Breadth_original_selected_with_groups$Group_ID <- as.factor(Breadth_original_selected_with_groups$Group_ID)

# Get a list of unique Group IDs
unique_groups <- unique(Breadth_original_selected_with_groups$Group_ID)

for (group in unique_groups) {

  
  Breadth_density_data <- Breadth_original_selected_with_groups %>%
      filter(Group_ID ==  group) %>% 
      filter(site != "HP") %>% 
      group_by(layers) %>%
      filter(value == max(value)) %>%
      ungroup() %>% 
      mutate(Breadth_percent = 100 * value)
  
  # Calculate the max histogram height for each site
  max_heights <- Breadth_density_data %>%
    filter(Group_ID == group, site != "HP") %>%
    mutate(bin = cut_width(Breadth_percent, width = 1, boundary = 0)) %>% # Adjust 'width' based on your data range and desired precision
    group_by(site, bin) %>%
    summarise(count = n(), .groups = 'drop') %>%
    group_by(site) %>%
    summarise(max_height = max(count))
  
  # Reorder 'site' based on max_height
  Breadth_density_data$site <- with(Breadth_density_data,
                                                     factor(site, levels = max_heights$site[order(max_heights$max_height, decreasing = TRUE)])
  )
  
  # Assuming 'Site' is your factor that has been ordered
  Breadth_density_data$site <- forcats::fct_rev(Breadth_density_data$site)
  
  plot <- ggplot(Breadth_density_data, aes(x = Breadth_percent, y = site, height = ..count.., fill = site)) + 
    geom_density_ridges(stat = "binline", scale = 15, bins = 100, draw_baseline = FALSE, size = 0, alpha = 0.75) +
    scale_fill_brewer(palette = "Dark2",direction=-1) +
    scale_x_continuous(expand = c(0, 0), limits = c(-1, 100)) +
    xlab("Breadth of Coverage (%)") + 
    ylab(NULL) + 
    theme_ridges(center_axis_labels = TRUE, line_size = 0.25) 
 
      
     # save plot
        ggsave(paste0(plot_path,group,"_ridgeline_plot.pdf"), plot = plot, width = 7, height = 5)
        
        
}

```

## Proportion of samples 


"/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/P_0003_Neisseriaceae-breadth_X_depth_hadamard.txt"

## Species-level Hadamard Breadth X Depth density ridgeline plots

First, consolidate hadamard data at the species level

```{r}

library(tidyr)
library(dplyr)

Hadamard_coverage_original <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/P_0003_Neisseriaceae-breadth_X_depth_hadamard.txt", header = TRUE, sep = "\t")

Final_Groups <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/TESTS/Classifying_genomes_to_oral_sites/Neisseriaceae_genome_groups.txt", header = TRUE, sep = "\t")
 
# select relevant columns
Hadamard_coverage_original_selected <- Hadamard_coverage_original %>% 
  dplyr::select(variable, bins, value)

# pivot wider
Hadamard_coverage_original_selected_wide <- Hadamard_coverage_original_selected %>%  
  pivot_wider(names_from = bins, values_from = value)

# mutate variable so that there is only site ID

Hadamard_coverage_original_selected_wide_mutated <- Hadamard_coverage_original_selected_wide %>% 
  mutate(site = case_when(grepl('TD_', variable) ~ 'TD', 
                                      grepl('KG_', variable) ~ 'KG', 
                                      grepl('PP_', variable) ~ 'SUPP',
                                      grepl('PB_', variable) ~ 'SUBP',
                                      grepl('SA_', variable) ~ 'SV',
                                      grepl('TH_', variable) ~ 'TH',
                                      grepl('PT_', variable) ~ 'PT',
                                      grepl('HP_', variable) ~ 'HP',
                                      grepl('BM_', variable) ~ 'BM'))

Hadamard_coverage_original_selected_wide_mutated <- Hadamard_coverage_original_selected_wide_mutated %>% dplyr::select(site, 2:(ncol(Hadamard_coverage_original_selected_wide_mutated) - 1))
  
# Create file that contains list of genome IDs in first column and mean relative abundances for each site in other columns???
# Pivot the data to longer format
long_df <- pivot_longer(Hadamard_coverage_original_selected_wide_mutated, cols = -site, names_to = "genomeIDs", values_to = "value")

long_df_with_groups <- merge(long_df, Final_Groups, by.x = "genomeIDs", by.y = "Pangenome_ID")

# Calculate the average relative abundance for each Genome_ID at each site
avg_Hadamard_coverage_df_per_group <- long_df_with_groups %>%
  group_by(site, Group_ID) %>%
  summarise(avg_Hadamard_coverage = mean(value, na.rm = TRUE)) %>%
  pivot_wider(names_from = site, values_from = avg_Hadamard_coverage)

avg_Hadamard_coverage_df_per_group <- avg_Hadamard_coverage_df_per_group %>% 
  rename(groupIDs = Group_ID) 

# Save file
write.table(avg_Hadamard_coverage_df_per_group, "/Users/home/SPECIES_LEVEL_PANGENOMES/TESTS/Classifying_genomes_to_oral_sites/Groups/With_HP/Neisseriaceae_mean_Hadamard_coverage_per_group.txt", quote = FALSE, row.names = FALSE, sep = "\t")


# Remove HP site
avg_Hadamard_coverage_df_per_group_no_HP <- avg_Hadamard_coverage_df_per_group %>% 
  dplyr::select(-HP)


# Save file
write.table(avg_Hadamard_coverage_df_per_group_no_HP, "/Users/home/SPECIES_LEVEL_PANGENOMES/TESTS/Classifying_genomes_to_oral_sites/Groups/Without_HP/Neisseriaceae_mean_Hadamard_coverage_per_group_no_HP.txt", quote = FALSE, row.names = FALSE, sep = "\t")


########## Now sum coverages per group ########## 
Hadamard_coverage_original_with_groups <- merge(Hadamard_coverage_original_selected, Final_Groups, by.x = "bins", by.y = "Pangenome_ID")

sum_Hadamard_coverage_df_per_group <- Hadamard_coverage_original_with_groups %>%
  group_by(variable, Group_ID) %>%
  summarise(sum_Hadamard_coverage = sum(value, na.rm = TRUE)) 

sum_Hadamard_coverage_df_per_group <- sum_Hadamard_coverage_df_per_group %>% 
  mutate(site = case_when(grepl('TD_', variable) ~ 'TD', 
                                      grepl('KG_', variable) ~ 'KG', 
                                      grepl('PP_', variable) ~ 'SUPP',
                                      grepl('PB_', variable) ~ 'SUBP',
                                      grepl('SA_', variable) ~ 'SV',
                                      grepl('TH_', variable) ~ 'TH',
                                      grepl('PT_', variable) ~ 'PT',
                                      grepl('HP_', variable) ~ 'HP',
                                      grepl('BM_', variable) ~ 'BM'))

# pivot wider
sum_Hadamard_coverage_df_per_group_wide <- sum_Hadamard_coverage_df_per_group %>%  
  dplyr::select(site, Group_ID, sum_Hadamard_coverage) %>% 
  droplevels() %>% 
  pivot_wider(names_from = Group_ID, values_from = sum_Hadamard_coverage) 

sum_Hadamard_coverage_df_per_group_wide <- sum_Hadamard_coverage_df_per_group_wide[,-1]


write.table(sum_Hadamard_coverage_df_per_group_wide, "/Users/home/SPECIES_LEVEL_PANGENOMES/TESTS/Classifying_genomes_to_oral_sites/Groups/With_HP/Neisseriaceae_sum_Hadamard_coverage_per_group.txt", quote = FALSE, row.names = FALSE, sep = "\t")
  
# Drop all rows that contain "HP" in "site: column

sum_Hadamard_coverage_df_per_group_wide_no_HP <- sum_Hadamard_coverage_df_per_group_wide %>% 
  filter(site != "HP")

write.table(sum_Hadamard_coverage_df_per_group_wide_no_HP, "/Users/home/SPECIES_LEVEL_PANGENOMES/TESTS/Classifying_genomes_to_oral_sites/Groups/Without_HP/Neisseriaceae_sum_Hadamard_coverage_per_group_no_HP.txt", quote = FALSE, row.names = FALSE, sep = "\t")

```

```{r}

library(ggplot2)
library(ggridges)
library(dplyr)

plot_path = "/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Hadamard_densities_per_species/"

df <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/TESTS/Classifying_genomes_to_oral_sites/Groups/Without_HP/Neisseriaceae_sum_Hadamard_coverage_per_group_no_HP.txt", header = TRUE, sep = "\t")

# Convert data to long form
df_long <- pivot_longer(df, cols = -site, names_to = "groupIDs", values_to = "value")
df_long$site <- as.factor(df_long$site)
df_long$groupIDs <- as.factor(df_long$groupIDs)

# Get a list of unique Group IDs
unique_groups <- unique(df_long$groupIDs)


for (group in unique_groups) {
  
       Hadamard_density_data <- df_long %>%
        filter(groupIDs ==  group)
      
      # Plot
      plot <- ggplot(Hadamard_density_data, aes(x = value, y = site, fill = after_stat(x))) +
        geom_density_ridges_gradient(rel_min_height = 0, size = 0.4, alpha = 0.7,
                                 jittered_points = TRUE,
                                 position = position_points_jitter(width = 0.0001, height = 0),
                                 point_shape = '|', point_size = 1, point_alpha = 1) +
        scale_fill_viridis_c(name = "Hadamard (Depth X Breadth)", option = "C") +
        scale_x_continuous(expand = c(0, 0)) +
        xlab(NULL) + 
        ylab(NULL) + 
        theme_ridges(center_axis_labels = TRUE) 
      
     # save plot
        ggsave(paste0(plot_path,group,"_ridgeline_plot.pdf"), plot = plot, width = 7, height = 5)
        
        
}

```

# 9. Hadamard - Depth * Breadth


```{r, eval=FALSE}

# Load mean coverage Q2Q3 data into R
df_coverage = read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/P_0003_Neisseriaceae-mean_coverage_Q2Q3.txt", header=TRUE, sep = "\t")

# Load detection data into R
df_detect = read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/P_0003_Neisseriaceae-detection.txt", header=TRUE, sep = "\t")

# Set the first column as row names
row.names(df_coverage) = df_coverage[,1]

# Remove the first column from the data
df_coverage = df_coverage[,-1]

# For df_detect
# Set the first column as row names
row.names(df_detect) = df_detect[,1]

# Remove the first column from the data
df_detect = df_detect[,-1]

# Convert data frames to matrices (if they are not already)
matrix_coverage = as.matrix(df_coverage)
matrix_detect = as.matrix(df_detect)

# Perform element-wise multiplication to form the Hadamard matrix
hadamard_matrix = matrix_coverage * matrix_detect


# convert hadamard_matrix to data frame 
hadamard_df <- as.data.frame(hadamard_matrix)

# Convert row names to a column named 'bins'
hadamard_df$bins <- row.names(hadamard_df)

# Convert 'bins' to be the first column
hadamard_df <- hadamard_df[, c('bins', names(hadamard_df)[!names(hadamard_df) %in% 'bins'])]

# convert from wide to long format
melted_hadamard_df <- melt(hadamard_df)


# make oral site variable for facet grid
melted_hadamard_df_2 <- melted_hadamard_df %>% mutate(site = case_when(grepl('TD_', variable) ~ 'TD', 
                                      grepl('KG_', variable) ~ 'KG', 
                                      grepl('PP_', variable) ~ 'SUPP',
                                      grepl('PB_', variable) ~ 'SUBP',
                                      grepl('SA_', variable) ~ 'SV',
                                      grepl('TH_', variable) ~ 'TH',
                                      grepl('PT_', variable) ~ 'PT',
                                      grepl('HP_', variable) ~ 'HP',
                                      grepl('BM_', variable) ~ 'BM')) 


# specify order for site terms
melted_hadamard_df_2$site <- factor(melted_hadamard_df_2$site, levels = c("SUPP", "SUBP", "KG", "BM", "TD", "PT", "HP", "TH", "SV"))

# sort "variable" based on pre-sorted values of "site"
melted_hadamard_df_2 <- melted_hadamard_df_2 %>%
  arrange(site, variable) %>%               # sort your dataframe
  mutate(variable = factor(variable, unique(variable))) # reset your factor-column based on that order

# Save data
write.table(melted_hadamard_df_2, "/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/P_0003_Neisseriaceae-breadth_X_depth_hadamard.txt", sep = "\t", quote = FALSE, row.names = FALSE)

# join metadata with to get total reads mapped into data frame
#melted_hadamard_df_3 = merge(mapping_metadata_merged, melted_hadamard_df_2, by.x = "layers", by.y = "variable") 

# reorder genomes based on pangenome order
#melted_hadamard_df_3$bins <- factor(melted_hadamard_df_3$bins, levels = tip_labels)




```

##### 9.a. FIGURE: Order samples by total reads

```{r}
dodge <- position_dodge(width=0.9)

total_reads_mapped_hadamard <- ggplot(data = melted_hadamard_df_3, aes(x=reorder(layers, -total_number_reads), y=total_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 25000000), breaks=seq(0,25000000,by=25000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("No. of reads mapped", "(0 - 25 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x = element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x = unit(0.1, "lines"),
        plot.margin = unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)

plot_total_reads_hadamard <- ggplot(data = melted_hadamard_df_3, aes(x=reorder(layers, -total_number_reads), y=total_number_reads)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 200000000), breaks=seq(0,200000000,by=200000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Total reads", "(0 - 200 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.text.y=element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)

plot_percent_reads_mapped_hadamard <- ggplot(data = melted_hadamard_df_3, aes(x=reorder(layers, -total_number_reads), y=percent_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 50), breaks=seq(0,50,by=50), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Percent reads mapped", "(0 - 50 %)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)


# set colors for heat map
paleta_new = rev(RColorBrewer::brewer.pal(n = 11, name = "RdYlBu"))

# Colors, including gray for the first segment
colors <- c("gray", paleta_new)

plot_strain_level_hadamard <- ggplot(data = melted_hadamard_df_3, aes(x=reorder(layers,-total_number_reads), y=bins, fill=value)) +
  geom_tile(height=1, width=1) + 
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") + 
  scale_fill_gradientn(colours = colors,
                       limits=c(1, 10),
                       #breaks=seq(0.005,0.5,by=0.1),
                       breaks = c(1, 2, 4, 6, 8, 10),
                       labels = c("< 1 X", "2 X", "4 X", "6 X", "8 X", "> 10 X"),
                       oob=squish) + 
  scale_y_discrete(limits = rev(levels(melted_hadamard_df_3$bins))) +
  ylab(NULL) +  
  xlab(NULL) + 
  theme_classic() + 
  guides(fill = guide_colourbar(barwidth = 1,
                                barheight = 10,
                                title = expression(atop("Hadamard product", "Q2Q3 DoC * BoC")),
                                ticks = TRUE,
                                ticks.colour = "white",
                                ticks.linewidth = 0.5)) +
  theme(text = element_text(size = 12),
        axis.ticks.y = element_line(color = rev(group_tick_colors), size = 1),
        axis.ticks.length = unit(1, "cm"),
        axis.text.x=element_blank(),
        axis.text.y =  element_text(size = 2, color = "black"),
        axis.line.y = element_blank(),
        axis.line.x = element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "black")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text=element_text(size=12)) + 
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,1,1), "cm")) +
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)


Strain_plot_hadamard_sorted_by_total_reads <- egg::ggarrange(plot_percent_reads_mapped_hadamard, total_reads_mapped_hadamard, plot_total_reads_hadamard, plot_strain_level_hadamard,
          ncol = 1,
          heights = c(0.1,0.1,0.1,1))

```



Save plot and data:

```{r, eval=FALSE}

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/strain_hadamard_samples_orderd_by_total_reads.pdf",Strain_plot_hadamard_sorted_by_total_reads, width = 13, height = 11)

```



##### 9.b. FIGURE: Order samples by heirarchical clustering

```{r}

##### Sample order; need to cluster each site seperately, then get order, and then concatenate the orders ###

# created vector with 5 characters
columns= c("order")

# pass this vector length to ncol parameter
# and nrow with 0
site_order_df = data.frame(matrix(nrow = 0, ncol = length(columns)))

# assign column names
colnames(site_order_df) = columns


site_list <- list("SUPP", "SUBP", "KG", "BM", "TD", "PT", "TH", "SV") # ignore HP since there is only one sample


for (oral_site in site_list) {
  
  
  oral_site_factor <- trimws(oral_site)
                             
  # filter df by site
  site_df <- melted_hadamard_df_3 %>% 
    dplyr::filter(site == oral_site_factor) 
  
  # subset data frame for clustering species
  site_clustering_df <- site_df %>% 
    select(bins, layers, value)
  
  # pivot wider
  site_wide_df <- site_clustering_df %>% pivot_wider(names_from = layers, values_from = value)
  
  # make tibble into data frame object
  site_wide_df <- as.data.frame(site_wide_df)
  
  # rename row names to species IDs
  site_wide_df2 <- site_wide_df[,-1]
  rownames(site_wide_df2) <- site_wide_df[,1]
  
  # reformat to matrix
  site_wide_matrix <- as.matrix(site_wide_df2)
  
  # transpose data frame so that we can cluster samples based on euclidan distance and complete linkage
  trans_site_wide_matrix <- t(site_wide_matrix)
  
  # calculate Bray-Curtis distance among samples
  site_dist_df <- vegan::vegdist(trans_site_wide_matrix, method = "bray")
  
  #Replace na values with 0 using is.na()
  site_dist_df[site_dist_df=="NaN"]<-0
  
  # Hierarchical clustering using Complete Linkage
  site_hc <- hclust(site_dist_df, method = "complete" )
  
  
  # get list of sample order for site
  
  site_order <- data.frame(order = site_hc$labels[site_hc$order])
  
  site_order_df <- rbind(site_order_df, site_order)
  
}


# add HP sample (HP_HC_HMP_S0195_01) to the end of the order 
#melted_strain_detection$layers[melted_strain_detection$site == "HP"] 
site_order_df <- rbind(site_order_df, "HP_HC_HMP_S0195_01")

# reorder samples based on concatenated hclust order
samples_order <- melted_hadamard_df_3$layers <- factor(melted_hadamard_df_3$layers, levels = site_order_df$order)



```


```{r}
dodge <- position_dodge(width=0.9)

total_reads_mapped_hadamard <- ggplot(data = melted_hadamard_df_3, aes(x=layers, y=total_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 25000000), breaks=seq(0,25000000,by=25000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("No. of reads mapped", "(0 - 25 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x = element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x = unit(0.1, "lines"),
        plot.margin = unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)

plot_total_reads_hadamard <- ggplot(data = melted_hadamard_df_3, aes(x=layers, y=total_number_reads)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 200000000), breaks=seq(0,200000000,by=200000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Total reads", "(0 - 200 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.text.y=element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)

plot_percent_reads_mapped_hadamard <- ggplot(data = melted_hadamard_df_3, aes(x=layers, y=percent_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 50), breaks=seq(0,50,by=50), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Percent reads mapped", "(0 - 50 %)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)


# set colors for heat map
paleta_new = rev(RColorBrewer::brewer.pal(n = 11, name = "RdYlBu"))

# Colors, including gray for the first segment
colors <- c("gray", paleta_new)

plot_strain_level_hadamard <- ggplot(data = melted_hadamard_df_3, aes(x=layers, y=bins, fill=value)) +
  geom_tile(height=1, width=1) + 
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") + 
  scale_fill_gradientn(colours = colors,
                       limits=c(1, 10),
                       #breaks=seq(0.005,0.5,by=0.1),
                       breaks = c(1, 2, 4, 6, 8, 10),
                       labels = c("< 1 X", "2 X", "4 X", "6 X", "8 X", "> 10 X"),
                       oob=squish) + 
  scale_y_discrete(limits = rev(levels(melted_hadamard_df_3$bins))) +
  ylab(NULL) +  
  xlab(NULL) + 
  theme_classic() + 
  guides(fill = guide_colourbar(barwidth = 1,
                                barheight = 10,
                                title = expression(atop("Hadamard product", "Q2Q3 DoC * BoC")),
                                ticks = TRUE,
                                ticks.colour = "white",
                                ticks.linewidth = 0.5)) +
  theme(text = element_text(size = 12),
        axis.ticks.y = element_line(color = rev(group_tick_colors), size = 1),
        axis.ticks.length = unit(1, "cm"),
        axis.text.x=element_blank(),
        axis.text.y =  element_text(size = 2, color = "black"),
        axis.line.y = element_blank(),
        axis.line.x = element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "black")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text=element_text(size=12)) + 
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,1,1), "cm")) +
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)


Strain_plot_hadamard_sorted_by_HC <- egg::ggarrange(plot_percent_reads_mapped_hadamard, total_reads_mapped_hadamard, plot_total_reads_hadamard, plot_strain_level_hadamard,
          ncol = 1,
          heights = c(0.1,0.1,0.1,1))

```

Save plot and data:

```{r, eval=FALSE}

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/strain_hadamard_samples_orderd_by_HC.pdf",Strain_plot_hadamard_sorted_by_HC, width = 13, height = 11)

```


# 10. Depth X Breadth plots

```{r}

library(scales)

melted_strain_detection <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Strain_level_updated_detection_data.csv", header = TRUE)

melted_strain_coverage <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Strain_level_updated_Q2Q3_mean_coverage_data.csv", header = TRUE)

melted_strain_detection$bins <- as.factor(melted_strain_detection$bins)
genomes <- levels(melted_strain_detection$bins)

Breadth_vs_Depth_plot <- function(genome) {
  
  # Step 1: Select a genome and melt dfs
  breadth <- melted_strain_detection %>% 
    filter(bins == genome)
  
  depth <- melted_strain_coverage %>% 
    filter(bins == genome)
  
  df <- data.frame(x = depth$value,
                    y = (breadth$value * 100),
                    site=as.factor(breadth$site))

  
  ggplot(df, aes(x, y, col = site, shape = site)) + 
  geom_point(size = 3, alpha = 0.5) + 
    geom_vline(xintercept = 1, col = "red", lty = 2, lwd = 0.5) + 
    geom_hline(yintercept = 50, col = "blue", lty = 2, lwd = 0.5) + 
    xlab("Depth of coverage (X)") + 
    ylab("Breadth of coverage (%)") +
    scale_shape_manual(values = c(15,0,16,1,17, 18,2,8, 3))+
    scale_color_manual(values = c("navy", "gray20","red4", "yellow4", "darkgreen", "magenta", "cyan", "purple4", "sienna"))+
    theme_classic()

  
  ggsave(paste0("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Breadth_vs_Depth/", genome,".pdf"), width = 7, height = 7)
  
}


for (i in genomes) {
  Breadth_vs_Depth_plot(i)
}

```



# 11. Depth X Breadth plots (Q1234 mean coverage)

```{r}


library(scales)

# Load mean coverage Q2Q3 data into R
df_coverage = read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/P_0003_Neisseriaceae-mean_coverage.txt", header=TRUE, sep = "\t")


# convert coverage data frame from wide to long format
melted_df_coverage <- melt(df_coverage)

# make oral site variable for facet grid
melted_df_coverage_2 <- melted_df_coverage %>% mutate(site = case_when(grepl('TD_', variable) ~ 'TD', 
                                      grepl('KG_', variable) ~ 'KG', 
                                      grepl('PP_', variable) ~ 'SUPP',
                                      grepl('PB_', variable) ~ 'SUBP',
                                      grepl('SA_', variable) ~ 'SV',
                                      grepl('TH_', variable) ~ 'TH',
                                      grepl('PT_', variable) ~ 'PT',
                                      grepl('HP_', variable) ~ 'HP',
                                      grepl('BM_', variable) ~ 'BM')) 


# specify order for site terms
melted_df_coverage_2$site <- factor(melted_df_coverage_2$site, levels = c("SUPP", "SUBP", "KG", "BM", "TD", "PT", "HP", "TH", "SV"))

# sort "variable" based on pre-sorted values of "site"
melted_df_coverage_2 <- melted_df_coverage_2 %>%
  arrange(site, variable) %>%               # sort your dataframe
  mutate(variable = factor(variable, unique(variable))) # reset your factor-column based on that order

# join metadata with to get total reads mapped into data frame
melted_strain_coverage = merge(mapping_metadata_merged, melted_df_coverage_2, by.x = "layers", by.y = "variable") 

melted_strain_detection <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Strain_level_updated_detection_data.csv", header = TRUE)


melted_strain_detection$bins <- as.factor(melted_strain_detection$bins)
genomes <- levels(melted_strain_detection$bins)

Breadth_vs_Depth_plot_Q1234_mean_coverage <- function(genome) {
  
  # Step 1: Select a genome and melt dfs
  breadth <- melted_strain_detection %>% 
    filter(bins == genome)
  
  depth <- melted_strain_coverage %>% 
    filter(bins == genome)
  
  df <- data.frame(x = depth$value,
                    y = (breadth$value * 100),
                    site=as.factor(breadth$site))

  
  ggplot(df, aes(x, y, col = site, shape = site)) + 
  geom_point(size = 3, alpha = 0.5) + 
    geom_vline(xintercept = 1, col = "red", lty = 2, lwd = 0.5) + 
    geom_hline(yintercept = 50, col = "blue", lty = 2, lwd = 0.5) + 
    xlab("Depth of coverage (X)") + 
    ylab("Breadth of coverage (%)") +
    scale_shape_manual(values = c(15,0,16,1,17, 18,2,8, 3))+
    scale_color_manual(values = c("navy", "gray20","red4", "yellow4", "darkgreen", "magenta", "cyan", "purple4", "sienna"))+
    theme_classic()

  
  ggsave(paste0("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Breadth_vs_Depth_Q1234/", genome,".pdf"), width = 7, height = 7)
  
}


for (i in genomes) {
  Breadth_vs_Depth_plot_Q1234_mean_coverage(i)
}

```

# 12. Species level plots

Species groups are based on ANI and phylogeny.

## Detection (50% breadth)

```{r, eval = FALSE}

library(tidyr)
library(dplyr)

Breadth_coverage_original <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Strain_level_updated_detection_data.csv", header = TRUE)

Final_Groups <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/TESTS/Classifying_genomes_to_oral_sites/Neisseriaceae_genome_groups.txt", header = TRUE, sep = "\t")

# select relevant columns
Detection_original_selected <- Breadth_coverage_original %>% 
  dplyr::select(layers, bins, detection_binary, total_number_reads, total_reads_mapped, percent_reads_mapped)

# Merge with species groups info
Detection_original_with_groups <- merge(Detection_original_selected, Final_Groups, by.x = "bins", by.y = "Pangenome_ID", all.x = TRUE)

# Sum detection and mutate detection so that values >=1 become binary
sum_detection_df_per_group <- Detection_original_with_groups %>%
  group_by(layers, Group_ID) %>%
  summarise(sum_detection = sum(detection_binary)) %>% 
  mutate(accumulated_detection = ifelse(sum_detection >= 1, 1, 0))


# Add site column
sum_detection_df_per_group <- sum_detection_df_per_group %>% 
  mutate(site = case_when(grepl('TD_', layers) ~ 'TD', 
                          grepl('KG_', layers) ~ 'KG', 
                          grepl('PP_', layers) ~ 'SUPP',
                          grepl('PB_', layers) ~ 'SUBP',
                          grepl('SA_', layers) ~ 'SV',
                          grepl('TH_', layers) ~ 'TH',
                          grepl('PT_', layers) ~ 'PT',
                          grepl('HP_', layers) ~ 'HP',
                          grepl('BM_', layers) ~ 'BM'))

# Get sample read count data
sample_data <- Breadth_coverage_original %>% 
  dplyr::select(layers, total_number_reads, total_reads_mapped, percent_reads_mapped) %>% 
  distinct(layers, .keep_all = TRUE)

# merge back with sample info
sum_detection_df_per_group_merged_back <- merge(sum_detection_df_per_group, sample_data, by = "layers", all.x = TRUE)

# Set order of sites
sum_detection_df_per_group_merged_back$site <- factor(sum_detection_df_per_group_merged_back$site, levels = c("SUPP", "SUBP", "KG", "BM", "TD", "PT", "HP", "TH", "SV"))


##### REORDER SPECIES #####

# subset data frame for clustering species & pivot wider
wide_df <- sum_detection_df_per_group_merged_back %>% 
  select(Group_ID, layers, accumulated_detection) %>%
  pivot_wider(names_from = layers, values_from = accumulated_detection)

# make tibble into data frame object
wide_df <- as.data.frame(wide_df)

# rename row names to species IDs
wide_df2 <- wide_df[,-1]
rownames(wide_df2) <- wide_df[,1]

# Convert to matrix
wide_matrix <- as.matrix(wide_df2)

# calculate Bray-Curtis distance among samples
dist_df <- vegan::vegdist(wide_matrix, method = "bray")

#Replace na values with 0 using is.na()
dist_df[dist_df=="NaN"]<-0

# Hierarchical clustering using Complete Linkage
hc <- hclust(dist_df, method = "complete" )

# Species order based on HC
species_manual_reorder <- c("N_subflava_VII","N_subflava_VIII","N_subflava_IV",
                            "N_subflava_III","N_subflava_I",  "N_subflava_II",
                            "N_mucosa_II",   "N_mucosa_III",  "N_mucosa_I",   
                            "N_subflava_V",  "N_cinerea",     "S_muelleri",  
                            "K_negevensis",  "K_kingae", "E_exigua", "N_sp_HMT_020", 
                            "N_mucosa_VII",  "K_bonacorsii_oralis", "N_elongata",   
                            "E_halliae",     "N_bacilliformis","N_mucosa_IV",  
                            "N_mucosa_V",    "K_sp_str_SNUBH_2017", "N_mucosa_VI",  
                            "E_corrodens",   "K_denitrificans","N_oralis",
                            "N_weaveri",    
                            "N_wadsworthii", "N_viridiae",  "N_uirgultaei", 
                            "N_subflava_VI", "N_subflava_IX", "N_sp_str_MVDL20_010259",
                            "N_sp_str_HSC_16F19","N_sp_str_3986_and_str_51_81","N_shayeganii", 
                            "N_polysaccharea","N_meningitidis","N_maigaei",    
                            "N_lactamica_str_NS19", "N_lactamica","N_gonorrhoeae",
                            "N_dumasiana",   "N_dentiae",     "N_cinerea_str_CCUG_5746",
                            "N_brasiliensis","N_blantyrii",   "N_bergeri",    
                            "N_benedictiae", "N_basseii","E_glucosivorans",  
                            "K_potus")

sum_detection_df_per_group_merged_back$Group_ID <- factor(sum_detection_df_per_group_merged_back$Group_ID, levels = species_manual_reorder)


##### FIGURE ##### 

dodge <- position_dodge(width=0.9)
Medium_panel_sizes <-c(4,1.5,1.5,4,4,1.5,0.5,1.5,1.5)
Smaller_panel_sizes <- unlist(lapply(Medium_panel_sizes,"*",0.75))
Larger_panel_sizes <- unlist(lapply(Medium_panel_sizes,"*",1.2))


total_reads_mapped_detection <- ggplot(data = sum_detection_df_per_group_merged_back, aes(x=reorder(layers, -total_number_reads), y=total_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 25000000), breaks=seq(0,25000000,by=25000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("No. of reads mapped", "(0 - 25 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 6),
        axis.ticks.y = element_blank(),
        axis.text.x = element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x = unit(0.1, "lines"),
        plot.margin = unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)
#total_reads_mapped_detection

plot_total_reads_detection <- ggplot(data = sum_detection_df_per_group_merged_back, aes(x=reorder(layers, -total_number_reads), y=total_number_reads)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 200000000), breaks=seq(0,200000000,by=200000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Total reads", "(0 - 200 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 6),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.text.y=element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)
#plot_total_reads_detection


plot_percent_reads_mapped_detection <- ggplot(data = sum_detection_df_per_group_merged_back, aes(x=reorder(layers, -total_number_reads), y=percent_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 50), breaks=seq(0,50,by=50), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Percent reads mapped", "(0 - 50 %)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 6),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)
#plot_percent_reads_mapped_detection


plot_species_level_detection <- ggplot(data = sum_detection_df_per_group_merged_back, aes(x=reorder(layers,-total_number_reads), y=Group_ID, fill=as.factor(accumulated_detection))) +
  geom_tile(height=1, width=1) + 
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") + 
  scale_fill_manual(values = c("black", "cyan"),
                    labels = c("Un-detected (Breadth < 50%)", "Detected (Breadth > 50%)"),
                    name = "Species detection") +
  scale_y_discrete(limits = rev(levels(sum_detection_df_per_group_merged_back$Group_ID))) +
  ylab(NULL) +  
  xlab(NULL) + 
  theme_classic() + 
  theme(text = element_text(size = 10),
        legend.position = "bottom",
        axis.ticks.y = element_line(linewidth = 1, linetype = "solid"),
        axis.ticks.length = unit(0.25, "cm"),
        axis.text.x=element_blank(),
        axis.text.y =  element_text(size = 10, color = "black"),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "black")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text=element_text(size=10)) + 
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,1,1), "cm")) +
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)
#plot_species_level_detection


Species_detection_total_reads <- egg::ggarrange(plot_percent_reads_mapped_detection, total_reads_mapped_detection,plot_total_reads_detection, plot_species_level_detection,
                                                    ncol = 1,
                                                    heights = c(0.05,0.05,0.05,1))

```


Save plot and data:
  
```{r, eval=FALSE}

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Species_groups_detection_binary_samples_orderd_by_total_reads.pdf",Species_detection_total_reads, width = 11, height = 8)

write.table(sum_detection_df_per_group_merged_back, "/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Species_groups_Accumulated_detection.txt", row.names = FALSE, quote = FALSE, sep = "\t")

```

## Binary Detection - Samples HC order

```{r,eval=FALSE}

library(tidyr)
library(dplyr)

Breadth_coverage_original <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Strain_level_updated_detection_data.csv", header = TRUE)

Final_Groups <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/TESTS/Classifying_genomes_to_oral_sites/Neisseriaceae_genome_groups.txt", header = TRUE, sep = "\t")

# select relevant columns
Detection_original_selected <- Breadth_coverage_original %>% 
  dplyr::select(layers, bins, detection_binary, total_number_reads, total_reads_mapped, percent_reads_mapped)

# Merge with species groups info
Detection_original_with_groups <- merge(Detection_original_selected, Final_Groups, by.x = "bins", by.y = "Pangenome_ID", all.x = TRUE)

# Sum detection and mutate detection so that values >=1 become binary
sum_detection_df_per_group <- Detection_original_with_groups %>%
  group_by(layers, Group_ID) %>%
  summarise(sum_detection = sum(detection_binary)) %>% 
  mutate(accumulated_detection = ifelse(sum_detection >= 1, 1, 0))


# Add site column
sum_detection_df_per_group <- sum_detection_df_per_group %>% 
  mutate(site = case_when(grepl('TD_', layers) ~ 'TD', 
                          grepl('KG_', layers) ~ 'KG', 
                          grepl('PP_', layers) ~ 'SUPP',
                          grepl('PB_', layers) ~ 'SUBP',
                          grepl('SA_', layers) ~ 'SV',
                          grepl('TH_', layers) ~ 'TH',
                          grepl('PT_', layers) ~ 'PT',
                          grepl('HP_', layers) ~ 'HP',
                          grepl('BM_', layers) ~ 'BM'))

# Get sample read count data
sample_data <- Breadth_coverage_original %>% 
  dplyr::select(layers, total_number_reads, total_reads_mapped, percent_reads_mapped) %>% 
  distinct(layers, .keep_all = TRUE)

# merge back with sample info
sum_detection_df_per_group_merged_back <- merge(sum_detection_df_per_group, sample_data, by = "layers", all.x = TRUE)

# Set order of sites
sum_detection_df_per_group_merged_back$site <- factor(sum_detection_df_per_group_merged_back$site, levels = c("SUPP", "SUBP", "KG", "BM", "TD", "PT", "HP", "TH", "SV"))



species_manual_reorder <- c( "N_cinerea",
                            "N_subflava_I", "N_subflava_II","N_subflava_III","N_subflava_IV",
                            "N_subflava_V","N_subflava_VII","N_subflava_VIII",
                            "N_mucosa_I", "N_mucosa_II", "N_mucosa_III", "N_mucosa_IV",
                            "N_mucosa_V", "N_mucosa_VI", "N_mucosa_VII", 
                            "N_elongata", "N_bacilliformis","N_sp_HMT_020", "N_oralis",
                            "K_bonacorsii_oralis", "K_denitrificans","K_sp_str_SNUBH_2017",
                            "K_negevensis",  "K_kingae",  "E_corrodens",  "E_halliae", "E_exigua", 
                            "S_muelleri")
                            
                          
                           
                           

sum_detection_df_per_group_merged_back$Group_ID <- factor(sum_detection_df_per_group_merged_back$Group_ID, levels = species_manual_reorder)



##### Sample order; need to cluster each site separately, then get order, and then concatenate the orders ###

# created vector with 5 characters
columns= c("order")

# pass this vector length to ncol parameter
# and nrow with 0
site_order_df = data.frame(matrix(nrow = 0, ncol = length(columns)))

# assign column names
colnames(site_order_df) = columns


site_list <- list("SUPP", "SUBP", "KG", "BM", "TD", "PT", "TH", "SV") # ignore HP since there is only one sample


for (oral_site in site_list) {
  
  
  oral_site_factor <- trimws(oral_site)
  
  sum_detection_df_per_group_merged_back$accumulated_detection <- as.numeric(as.character(sum_detection_df_per_group_merged_back$accumulated_detection))
                             
  # filter df by site
  site_df <- sum_detection_df_per_group_merged_back %>% 
    dplyr::filter(site == oral_site_factor) 
  
  # subset data frame for clustering species
  site_clustering_df <- site_df %>% 
    select(Group_ID, layers, accumulated_detection)
  
  # pivot wider
  site_wide_df <- site_clustering_df %>% pivot_wider(names_from = layers, values_from = accumulated_detection)
  
  # make tibble into data frame object
  site_wide_df <- as.data.frame(site_wide_df)
  
  # rename row names to species IDs
  site_wide_df2 <- site_wide_df[,-1]
  rownames(site_wide_df2) <- site_wide_df[,1]
  
  # reformat to matrix
  site_wide_matrix <- as.matrix(site_wide_df2)
  
  # transpose data frame so that we can cluster samples based on euclidian distance and complete linkage
  trans_site_wide_matrix <- t(site_wide_matrix)
  
  # calculate Bray-Curtis distance among samples
  site_dist_df <- vegan::vegdist(trans_site_wide_matrix, method = "jaccard")
  
  #Replace na values with 0 using is.na()
  site_dist_df[site_dist_df=="NaN"]<-0
  
  # Hierarchical clustering using Complete Linkage
  site_hc <- hclust(site_dist_df, method = "complete" )
  

  # get list of sample order for site
  
  site_order <- data.frame(order = site_hc$labels[site_hc$order])
  
  site_order_df <- rbind(site_order_df, site_order)
  
}


# add HP sample (HP_HC_HMP_S0195_01) to the end of the order 
#melted_strain_detection$layers[melted_strain_detection$site == "HP"] 
site_order_df <- rbind(site_order_df, "HP_HC_HMP_S0195_01")

# reorder samples based on concatenated hclust order
samples_order <- sum_detection_df_per_group_merged_back$layers <- factor(sum_detection_df_per_group_merged_back$layers, levels = site_order_df$order)

##### FIGURE ##### 

dodge <- position_dodge(width=0.9)
Medium_panel_sizes <-c(4,1.5,1.5,4,4,1.5,0.5,1.5,1.5)
Smaller_panel_sizes <- unlist(lapply(Medium_panel_sizes,"*",0.75))
Larger_panel_sizes <- unlist(lapply(Medium_panel_sizes,"*",1.2))


total_reads_mapped_detection <- ggplot(data = sum_detection_df_per_group_merged_back, aes(x=layers, y=total_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 25000000), breaks=seq(0,25000000,by=25000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("No. of reads mapped", "(0 - 25 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 6),
        axis.ticks.y = element_blank(),
        axis.text.x = element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x = unit(0.1, "lines"),
        plot.margin = unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)
#total_reads_mapped_detection

plot_total_reads_detection <- ggplot(data = sum_detection_df_per_group_merged_back, aes(x=layers, y=total_number_reads)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 200000000), breaks=seq(0,200000000,by=200000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Total reads", "(0 - 200 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 6),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.text.y=element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)
#plot_total_reads_detection


plot_percent_reads_mapped_detection <- ggplot(data = sum_detection_df_per_group_merged_back, aes(x=layers, y=percent_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 50), breaks=seq(0,50,by=50), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Percent reads mapped", "(0 - 50 %)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 6),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)
#plot_percent_reads_mapped_detection


plot_species_level_detection <- ggplot(data = sum_detection_df_per_group_merged_back, aes(x=layers, y=Group_ID, fill=as.factor(accumulated_detection))) +
  geom_tile(height=1, width=1) + 
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") + 
  scale_fill_manual(values = c("black", "cyan"),
                    labels = c("Un-detected (Breadth < 50%)", "Detected (Breadth > 50%)"),
                    name = "Species detection") +
  scale_y_discrete(limits = rev(levels(sum_detection_df_per_group_merged_back$Group_ID))) +
  ylab(NULL) +  
  xlab(NULL) + 
  theme_classic() + 
  theme(text = element_text(size = 10),
        legend.position = "bottom",
        axis.ticks.y = element_line(linewidth = 1, linetype = "solid"),
        axis.ticks.length = unit(0.25, "cm"),
        axis.text.x=element_blank(),
        axis.text.y =  element_text(size = 10, color = "black"),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "black")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text=element_text(size=10)) + 
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,1,1), "cm")) +
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)
#plot_species_level_detection


Species_detection_HC <- egg::ggarrange(plot_percent_reads_mapped_detection, total_reads_mapped_detection,plot_total_reads_detection, plot_species_level_detection,
                                                    ncol = 1,
                                                    heights = c(0.05,0.05,0.05,1))

```


Save plot and data:
  
```{r, eval=FALSE}

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Species_groups_detection_binary_samples_orderd_by_HC.pdf",Species_detection_HC, width = 11, height = 8)

```


## Q2Q3 Mean Coverage

```{r, eval = FALSE}

library(tidyr)
library(dplyr)

# Load data
Q2Q3_coverage_original <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Strain_level_updated_Q2Q3_mean_coverage_data.csv", header = TRUE)
Final_Groups <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/TESTS/Classifying_genomes_to_oral_sites/Neisseriaceae_genome_groups.txt", header = TRUE, sep = "\t")

# select relevant columns
Q2Q3_coverage_original_selected <- Q2Q3_coverage_original %>% 
  dplyr::select(layers, site, bins, value, total_number_reads, total_reads_mapped, percent_reads_mapped)

# Merge with species groups info
Q2Q3_coverage_original_with_groups <- merge(Q2Q3_coverage_original_selected, Final_Groups, by.x = "bins", by.y = "Pangenome_ID", all.x = TRUE)


# Sum coverages per group 
sum_Q2Q3_coverage_df_per_group <- Q2Q3_coverage_original_with_groups %>%
  group_by(layers, Group_ID) %>%
  summarise(sum_Q2Q3_coverage = sum(value, na.rm = TRUE)) 


# Get sample read count data
sample_data <- Breadth_coverage_original %>% 
  dplyr::select(layers, site, total_number_reads, total_reads_mapped, percent_reads_mapped) %>% 
  distinct(layers, .keep_all = TRUE)

# merge back with sample info
sum_Q2Q3_coverage_df_per_group_merged_back <- merge(sum_Q2Q3_coverage_df_per_group, sample_data, by = "layers", all.x = TRUE)

# Set order of sites
sum_Q2Q3_coverage_df_per_group_merged_back$site <- factor(sum_Q2Q3_coverage_df_per_group_merged_back$site, levels = c("SUPP", "SUBP", "KG", "BM", "TD", "PT", "HP", "TH", "SV"))


##### REORDER SPECIES #####

# subset data frame for clustering species & pivot wider
wide_df <- sum_Q2Q3_coverage_df_per_group_merged_back %>% 
  select(Group_ID, layers, sum_Q2Q3_coverage) %>%
  pivot_wider(names_from = layers, values_from = sum_Q2Q3_coverage)

# make tibble into data frame object
wide_df <- as.data.frame(wide_df)

# rename row names to species IDs
wide_df2 <- wide_df[,-1]
rownames(wide_df2) <- wide_df[,1]

# Convert to matrix
wide_matrix <- as.matrix(wide_df2)

# calculate Bray-Curtis distance among samples
dist_df <- vegan::vegdist(wide_matrix, method = "bray")

#Replace na values with 0 using is.na()
dist_df[dist_df=="NaN"]<-0

# Hierarchical clustering using Complete Linkage
hc <- hclust(dist_df, method = "complete" )

dhc <- as.dendrogram(hc)
dhc2 <- click_rotate(dhc)


order.dendrogram(dhc2)
labels(dhc2)

# Species order based on HC
species_manual_reorder <- labels(dhc2)

sum_Q2Q3_coverage_df_per_group_merged_back$Group_ID <- factor(sum_Q2Q3_coverage_df_per_group_merged_back$Group_ID, levels = species_manual_reorder)


##### FIGURE ##### 

dodge <- position_dodge(width=0.9)
Medium_panel_sizes <-c(4,1.5,1.5,4,4,1.5,0.5,1.5,1.5)
Smaller_panel_sizes <- unlist(lapply(Medium_panel_sizes,"*",0.75))
Larger_panel_sizes <- unlist(lapply(Medium_panel_sizes,"*",1.2))

# set colors for heat map
paleta_new = rev(RColorBrewer::brewer.pal(n = 11, name = "RdYlBu"))

# Colors, including gray for the first segment
colors <- c("gray", paleta_new)
color_no_gray <- c(paleta_new)

total_reads_mapped_coverage <- ggplot(data = sum_Q2Q3_coverage_df_per_group_merged_back, aes(x=reorder(layers, -total_number_reads), y=total_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 25000000), breaks=seq(0,25000000,by=25000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("No. of reads mapped", "(0 - 25 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 6),
        axis.ticks.y = element_blank(),
        axis.text.x = element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x = unit(0.1, "lines"),
        plot.margin = unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)


plot_total_reads_coverage <- ggplot(data = sum_Q2Q3_coverage_df_per_group_merged_back, aes(x=reorder(layers, -total_number_reads), y=total_number_reads)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 200000000), breaks=seq(0,200000000,by=200000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Total reads", "(0 - 200 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 6),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.text.y=element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)


plot_percent_reads_mapped_coverage <- ggplot(data = sum_Q2Q3_coverage_df_per_group_merged_back, aes(x=reorder(layers, -total_number_reads), y=percent_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 50), breaks=seq(0,50,by=50), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Percent reads mapped", "(0 - 50 %)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 6),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)

plot_species_level_coverage <- ggplot(data = sum_Q2Q3_coverage_df_per_group_merged_back, aes(x=reorder(layers,-total_number_reads), y=Group_ID, fill=sum_Q2Q3_coverage)) +
  geom_tile(height=1, width=1) + 
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") + 
  scale_fill_gradientn(colours = color_no_gray,
                       limits=c(0, 10),
                       breaks = c(0, 2, 4, 6, 8, 10),
                       labels = c("0X", "2X", "4X", "6X", "8X", "> 10X"),
                       oob=squish) + 
  scale_y_discrete(limits = levels(sum_Q2Q3_coverage_df_per_group_merged_back$Group_ID)) +
  ylab(NULL) +  
  xlab(NULL) + 
  theme_classic() + 
  guides(fill = guide_colourbar(barwidth = 10,
                                barheight = 1,
                                title = "Sum of Mean Q2Q3 Depth of Coverage",
                                ticks = TRUE,
                                ticks.colour = "white",
                                ticks.linewidth = 0.5)) +
  theme(text = element_text(size = 10),
        legend.position = "bottom",
        axis.ticks.y = element_line(linewidth = 1, linetype = "solid"),
        axis.ticks.length = unit(0.25, "cm"),
        axis.text.x=element_blank(),
        axis.text.y =  element_text(size = 7, color = "black"),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "black")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text=element_text(size=10)) + 
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,1,1), "cm")) +
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)


Species_coverage_total_reads <- egg::ggarrange(plot_percent_reads_mapped_coverage, total_reads_mapped_coverage,plot_total_reads_coverage, plot_species_level_coverage,
                                                ncol = 1,
                                                heights = c(0.05,0.05,0.05,1))

```


Save plot and data:
  
```{r, eval=FALSE}

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Species_groups_Sum_Q2Q3_coverage_samples_orderd_by_total_reads.pdf",Species_coverage_total_reads, width = 11, height = 8)

write.table(sum_Q2Q3_coverage_df_per_group_merged_back, "/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Species_groups_Sum_Q2Q3_coverage.txt", row.names = FALSE, quote = FALSE, sep = "\t")

```

## Relative abundance V1


```{r, eval = FALSE}

library("vegan")
library("ggdendro")

# Load data

# Load csv file with genome IDs and colors; this file was made when the tanglegram was made
bins_and_colors <-read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/bins_colors.csv", header = TRUE)

# Load breadth data
melted_df_detect <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Strain_level_updated_detection_data.csv", header = TRUE)

# Load mean Q2Q3 depth of coverage data
melted_df_coverage <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Strain_level_updated_Q2Q3_mean_coverage_data.csv", header = TRUE)

  
  # Load sample metadata into R
sample_metadata <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/HMP_metadata/Final_1297_HMP_metagenomes_metadata.csv", header=TRUE)

# load mapping stats
mapping_stats <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/P_0003_Neisseriaceae-mapping_stats_new.txt", header=TRUE, sep = "\t")

# Merge the two 
mapping_metadata_merged <- merge(mapping_stats, sample_metadata, by.x = "layers", by.y = "Internal_ID")

# Add variable for percentage reads mapped
mapping_metadata_merged <- mapping_metadata_merged %>% 
  mutate(total_number_reads = (total_pairs_passed*2)) %>% 
  mutate(percent_reads_mapped = 100*(total_reads_mapped/total_number_reads))

# new species IDs based detection profiles
group_IDs=as.factor(bins_and_colors$Species_group)

# new species colors based detection profiles
group_tick_colors=bins_and_colors$Group_color

genus_tick_colors=bins_and_colors$Genus_color

# TO EXCLUDE UN_DETECTED GENOMES
melted_df_detect_filtered <- melted_df_detect %>% 
  filter(detected == "detected")
filtered_melted_df <-merge(melted_df_coverage, melted_df_detect_filtered, by=c("layers","bins")) 
filtered_melted_df <- filtered_melted_df %>% dplyr::rename(coverage=value.x, breadth=value.y)

# select columns and transpose from long to wide
meanCov <- filtered_melted_df %>% 
  dplyr::select(layers, bins, coverage) %>% 
  pivot_wider(id_cols=bins, names_from=layers, values_from=coverage)

# chnage NA values to zeros
meanCov[is.na(meanCov)] <- 0

# transpose data frame
rownames <- meanCov$bins
meanCovT <- as.data.frame(t(meanCov[-1]))
colnames(meanCovT) <- rownames

# mean coverage of a genome / sum of mean coverages of all genomes in a sample
relAbsT <- decostand(meanCovT, method = "total")

# transpose relative abundance matrix
relAbs <- t(relAbsT)

# multiply by 100
relAbs <- 100*relAbs 

# change coverage data frame from wide to long format
melted_relAbs <- reshape2::melt(relAbs) %>% 
  dplyr::rename(variable=Var2, Genome_ID=Var1, relAbundance=value)

# merge with relabundance
melted_relAbs_merged <- merge(melted_relAbs,bins_and_colors, by.x = 'Genome_ID', by.y = 'Pangenome_ID')

# set Pangenome_group ID as factor
melted_relAbs_merged$Species_group <- as.factor(melted_relAbs_merged$Species_group)

# rename metagenomes to just site name
#melted_relAbs_merged$variable=gsub("_HC_HMP","",melted_relAbs_merged$variable)

# make oral site variable for facet grid
melted_relAbs_merged <- melted_relAbs_merged %>% mutate(site = case_when(grepl('TD_', variable) ~ 'TD', 
                                      grepl('KG_', variable) ~ 'KG', 
                                      grepl('PP_', variable) ~ 'SUPP',
                                      grepl('PB_', variable) ~ 'SUBP',
                                      grepl('SA_', variable) ~ 'SV',
                                      grepl('TH_', variable) ~ 'TH',
                                      grepl('PT_', variable) ~ 'PT',
                                      grepl('HP_', variable) ~ 'HP',
                                      grepl('BM_', variable) ~ 'BM')) 

# Save relative abundance data for genomes
write.csv(melted_relAbs_merged, "/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Relative_abundance_Neisseriaceae_genomes.csv", row.names = FALSE, quote = FALSE)

#Sum abundance by oral site, groupID and variable to obtain the relative abundance of each species (aka groupID) for each metagenome (aka variable)
melted_relAbs_merged_summed <- melted_relAbs_merged %>% 
  group_by(site, Species_group, variable) %>% 
  dplyr::summarise(sum = sum(relAbundance))
  

 # merge total reads mapped into data frame
melted_relAbs_merged_summed = merge.data.frame(mapping_metadata_merged, melted_relAbs_merged_summed, by.x = "layers", by.y = "variable") 

# reorder oral sites
melted_relAbs_merged_summed$site <- factor(melted_relAbs_merged_summed$site, levels = c("SUPP", "SUBP", "KG", "BM", "TD", "PT", "HP", "TH", "SV"))


##### Sample order; need to cluster each site seperately, then get order, and then concatenate the orders ###

# Extract the levels of the original factor
original_levels <- levels(melted_relAbs_merged_summed$Species_group)

# Extract the levels of the factor created from bins_and_colors$Species_group
new_levels <- levels(as.factor(bins_and_colors$Species_group))

# Find the differences between the two sets of levels
differences_reverse <- setdiff(new_levels, original_levels)

# created vector with 5 characters
columns= c("order")

# pass this vector length to ncol parameter
# and nrow with 0
site_order_df = data.frame(matrix(nrow = 0, ncol = length(columns)))

# assign column names
colnames(site_order_df) = columns


site_list <- list("SUPP", "SUBP", "KG", "BM", "TD", "PT", "TH", "SV") # ignore HP since there is only one sample


for (oral_site in site_list) {
  
  oral_site_factor <- trimws(oral_site)
  
  # filter df by site
  site_df <- melted_relAbs_merged_summed %>% 
    filter(site == oral_site_factor)
  
  # subset data frame for clustering species
  site_clustering_df <- site_df %>% 
    dplyr::select(Species_group, layers, sum)
  
  # pivot wider
  site_wide_df <- site_clustering_df %>% pivot_wider(names_from = layers, values_from = sum)
  
  # make tibble into data frame object
  site_wide_df <- as.data.frame(site_wide_df)
  
  # rename row names to species IDs
  site_wide_df2 <- site_wide_df[,-1]
  rownames(site_wide_df2) <- site_wide_df[,1]
  
  # add rows of zeroes for undetected species
  for (level in differences_reverse) {
  
    # Create a row with zeros for all columns in wide_df2
    missing_group <- as.data.frame(matrix(0, nrow=1, ncol=ncol(site_wide_df2)))
    
    # Assign column names to match wide_df2
    colnames(missing_group) <- colnames(site_wide_df2)
    
    # Add the missing group as a row name or append the group name as an attribute if you need to keep track
    rownames(missing_group) <- level
    
    # Append the missing group row to wide_df2
    site_wide_df2 <- rbind(site_wide_df2, missing_group)
  }

  site_wide_matrix <- as.matrix(site_wide_df2)
  
  # transpose data frame so that we can cluster samples based on euclidan distance and complete linkage
  trans_site_wide_matrix <- t(site_wide_matrix)
  
  # calculate Bray-Curtis distance among samples
  site_dist_df <- vegan::vegdist(trans_site_wide_matrix, method = "bray")
  
  #Replace na values with 0 using is.na()
  site_dist_df[site_dist_df=="NaN"]<-0
  
  # Hierarchical clustering using Complete Linkage
  site_hc <- hclust(site_dist_df, method = "complete" )
  
  # get list of sample order for site
  
  site_order <- data.frame(order = site_hc$labels[site_hc$order])
  
  site_order_df <- rbind(site_order_df, site_order)
  
}


# add HP sample to the end of the order 
melted_relAbs_merged_summed$layers[melted_relAbs_merged_summed$site == "HP"] 
site_order_df <- rbind(site_order_df, "HP_HC_HMP_S0195_01")

# reorder samples based on concatenated hclust order
samples_order <- melted_relAbs_merged_summed$layers <- factor(melted_relAbs_merged_summed$layers, levels = site_order_df$order)


# Extract unique combinations of columns excluding Species_group and sum
unique_combinations <- unique(melted_relAbs_merged_summed[, !names(melted_relAbs_merged_summed) %in% c("Species_group", "sum")])

# Initialize an empty dataframe to store the missing groups
missing_groups_df <- data.frame()

# Loop over each level in differences_reverse to create rows for missing species
for (level in differences_reverse) {
  
  # Create a dataframe with zeros for sum and the current missing Species_group
  new_rows <- unique_combinations
  new_rows$Species_group <- level
  new_rows$sum <- 0
  
  # Bind these new rows to the missing_groups_df
  missing_groups_df <- rbind(missing_groups_df, new_rows)
}

# Bind the missing groups dataframe to the original dataframe
melted_relAbs_merged_summed <- rbind(melted_relAbs_merged_summed, missing_groups_df)


# manually reorder species based on pangenome/phylogney/ani  
species_order <- rev(levels(melted_relAbs_merged_summed$Species_group))

melted_relAbs_merged_summed$Species_group <- factor(melted_relAbs_merged_summed$Species_group, levels = species_order)




```

```{r, eval=FALSE}


dodge <- position_dodge(width=0.9)

total_reads_mapped_rel_ab <- ggplot(data = melted_relAbs_merged_summed, aes(x=layers, y=total_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 25000000), breaks=seq(0,25000000,by=25000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("No. of reads mapped", "(0 - 25 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)


plot_total_reads_rel_ab <- ggplot(data = melted_relAbs_merged_summed, aes(x=layers, y=total_number_reads)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 200000000), breaks=seq(0,200000000,by=200000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Total reads", "(0 - 200 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.text.y=element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)

plot_percent_reads_mapped_rel_ab <- ggplot(data = melted_relAbs_merged_summed, aes(x=layers, y=percent_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 50), breaks=seq(0,50,by=50), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Percent reads mapped", "(0 - 50 %)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)



######### Relative abundance plot ######### 


# Load csv file with genome IDs and colors; this file was made when the tanglegram was made
bins_and_colors <-read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/bins_colors.csv", header = TRUE)

bins_and_colors_species <- bins_and_colors %>% 
  select(Species_group, Group_color, Genus_color) %>% 
  distinct()

melted_relAbs_merged_summed_tick_colors <- merge(melted_relAbs_merged_summed, bins_and_colors_species, by = "Species_group")

# set colors for heat map
paleta_new = rev(RColorBrewer::brewer.pal(n = 11, name = "RdBu"))

# Add black for the first segment
colors <- c("black", paleta_new)

colors2 <- c("black", "#2166AC","#D1E5F0", "white", "#FDDBC7", "#F4A582", "#FF0000")


plot_group_level_rel_ab <- ggplot(data = melted_relAbs_merged_summed_tick_colors, aes(x=layers, y=Species_group, fill=sum)) +
  geom_tile(height=1, width=1) + 
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") + 
  scale_fill_gradientn(colours = colors2,
                       limits=c(0, 100),
                       #breaks=seq(0.005,0.5,by=0.1),
                       breaks = c(0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100),
                       labels = c("0 %", "10 %", "20 %", "30 %", "40 %", "50 %", "60 %", "70 %", "80 %", "90 %", "100 %"),
                       oob=squish) + 
  scale_y_discrete(limits = levels(melted_relAbs_merged_summed_tick_colors$Species_group)) +
  ylab(NULL) +  
  xlab(NULL) + 
  theme_classic() + 
  guides(fill = guide_colourbar(barwidth = 1,
                                barheight = 10,
                                title = "Relative abundance",
                                ticks = TRUE,
                                ticks.colour = "white",
                                ticks.linewidth = 0.5)) +
  theme(text = element_text(size = 10),
        axis.ticks.y = element_line(color = rev(melted_relAbs_merged_summed_tick_colors$Group_color), linewidth = 1.2),
        #axis.ticks.y = element_blank(),
        axis.ticks.length = unit(1, "cm"),
        axis.text.x=element_blank(),
        axis.text.y =  element_text(size = 9, color = "black"),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "black")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        #panel.border = element_rect(colour = "black", fill = NA),
        strip.text=element_text(size=12)) + 
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,1,1), "cm")) +
  force_panelsizes(cols = unit(c(4,2,2,4,4,2,0.75,2,1), "cm"), 
                 TRUE)


Group_level_plot_rel_ab <- egg::ggarrange(plot_percent_reads_mapped_rel_ab, total_reads_mapped_rel_ab, plot_total_reads_rel_ab, plot_group_level_rel_ab,
          ncol = 1,
          heights = c(0.1,0.1,0.1,1))
```


```{r, eval=FALSE}

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Relative_abundance_of_detected_species_groups.pdf", Group_level_plot_rel_ab, width = 14, height = 10)

```

## Relative abundance V2

Samples sorted by heirarchical clustering per site

```{r, eval = FALSE}

library("vegan")
library("ggdendro")

# Load data

# Load csv file with genome IDs and colors; this file was made when the tanglegram was made
bins_and_colors <-read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/bins_colors.csv", header = TRUE)

# Load breadth data
melted_df_detect <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Strain_level_updated_detection_data.csv", header = TRUE)

# Load mean Q2Q3 depth of coverage data
melted_df_coverage <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Strain_level_updated_Q2Q3_mean_coverage_data.csv", header = TRUE)


# Load sample metadata into R
sample_metadata <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/HMP_metadata/Final_1297_HMP_metagenomes_metadata.csv", header=TRUE)

# load mapping stats
mapping_stats <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/P_0003_Neisseriaceae-mapping_stats_new.txt", header=TRUE, sep = "\t")

# Merge the two 
mapping_metadata_merged <- merge(mapping_stats, sample_metadata, by.x = "layers", by.y = "Internal_ID")

# Add variable for percentage reads mapped
mapping_metadata_merged <- mapping_metadata_merged %>% 
  mutate(total_number_reads = (total_pairs_passed*2)) %>% 
  mutate(percent_reads_mapped = 100*(total_reads_mapped/total_number_reads))

# new species IDs based detection profiles
group_IDs=as.factor(bins_and_colors$Species_group)

# new species colors based detection profiles
group_tick_colors=bins_and_colors$Group_color

genus_tick_colors=bins_and_colors$Genus_color

# TO EXCLUDE UN_DETECTED GENOMES
melted_df_detect_filtered <- melted_df_detect %>% 
  filter(detected == "detected")
filtered_melted_df <-merge(melted_df_coverage, melted_df_detect_filtered, by=c("layers","bins")) 
filtered_melted_df <- filtered_melted_df %>% dplyr::rename(coverage=value.x, breadth=value.y)

# select columns and transpose from long to wide
meanCov <- filtered_melted_df %>% 
  dplyr::select(layers, bins, coverage) %>% 
  pivot_wider(id_cols=bins, names_from=layers, values_from=coverage)

# chnage NA values to zeros
meanCov[is.na(meanCov)] <- 0

# transpose data frame
rownames <- meanCov$bins
meanCovT <- as.data.frame(t(meanCov[-1]))
colnames(meanCovT) <- rownames

# mean coverage of a genome / sum of mean coverages of all genomes in a sample
relAbsT <- decostand(meanCovT, method = "total")

# transpose relative abundance matrix
relAbs <- t(relAbsT)

# multiply by 100
relAbs <- 100*relAbs 

# change coverage data frame from wide to long format
melted_relAbs <- reshape2::melt(relAbs) %>% 
  dplyr::rename(variable=Var2, Genome_ID=Var1, relAbundance=value)

# merge with relabundance
melted_relAbs_merged <- merge(melted_relAbs,bins_and_colors, by.x = 'Genome_ID', by.y = 'Pangenome_ID')

# set Pangenome_group ID as factor
melted_relAbs_merged$Species_group <- as.factor(melted_relAbs_merged$Species_group)

# rename metagenomes to just site name
#melted_relAbs_merged$variable=gsub("_HC_HMP","",melted_relAbs_merged$variable)

# make oral site variable for facet grid
melted_relAbs_merged <- melted_relAbs_merged %>% mutate(site = case_when(grepl('TD_', variable) ~ 'TD', 
                                                                         grepl('KG_', variable) ~ 'KG', 
                                                                         grepl('PP_', variable) ~ 'SUPP',
                                                                         grepl('PB_', variable) ~ 'SUBP',
                                                                         grepl('SA_', variable) ~ 'SV',
                                                                         grepl('TH_', variable) ~ 'TH',
                                                                         grepl('PT_', variable) ~ 'PT',
                                                                         grepl('HP_', variable) ~ 'HP',
                                                                         grepl('BM_', variable) ~ 'BM')) 

# Save relative abundance data for genomes
#write.csv(melted_relAbs_merged, "/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Relative_abundance_Neisseriaceae_genomes.csv", row.names = FALSE, quote = FALSE)


# Change mucosa and subflava groups 

melted_relAbs_merged <- melted_relAbs_merged %>% 
  dplyr::mutate(Group_ID_2 = case_when(
    grepl("N_subflava_V", Species_group) ~ "N_subflava_V",
    grepl("N_subflava_I", Species_group) ~ "N_subflava_I-VIII",
    grepl("N_subflava_II", Species_group) ~ "N_subflava_I-VIII",
    grepl("N_subflava_III", Species_group) ~ "N_subflava_I-VIII",
    grepl("N_subflava_IV", Species_group) ~ "N_subflava_I-VIII",
    grepl("N_subflava_VII", Species_group) ~ "N_subflava_I-VIII",
    grepl("N_subflava_VIII", Species_group) ~ "N_subflava_I-VIII",
    grepl("N_mucosa", Species_group) ~ "N_mucosa",
    TRUE ~ Species_group  # Keeps all other Group_IDs the same
  ))


#Sum abundance by oral site, groupID and variable to obtain the relative abundance of each species (aka groupID) for each metagenome (aka variable)
melted_relAbs_merged_summed <- melted_relAbs_merged %>% 
  group_by(site, Group_ID_2, variable) %>% 
  dplyr::summarise(sum = sum(relAbundance))


# merge total reads mapped into data frame
melted_relAbs_merged_summed = merge.data.frame(mapping_metadata_merged, melted_relAbs_merged_summed, by.x = "layers", by.y = "variable") 

# reorder oral sites
melted_relAbs_merged_summed$site <- factor(melted_relAbs_merged_summed$site, levels = c("SUPP", "SUBP", "KG", "BM", "TD", "PT", "HP", "TH", "SV"))


##### Sample order; need to cluster each site seperately, then get order, and then concatenate the orders ###


# created vector with 5 characters
columns= c("order")

# pass this vector length to ncol parameter
# and nrow with 0
site_order_df = data.frame(matrix(nrow = 0, ncol = length(columns)))

# assign column names
colnames(site_order_df) = columns


site_list <- list("SUPP", "SUBP", "KG", "BM", "TD", "PT", "TH", "SV") # ignore HP since there is only one sample


for (oral_site in site_list) {
  
  oral_site_factor <- trimws(oral_site)
  
  # filter df by site
  site_df <- melted_relAbs_merged_summed %>% 
    filter(site == oral_site_factor)
  
  # subset data frame for clustering species
  site_clustering_df <- site_df %>% 
    dplyr::select(Group_ID_2, layers, sum)
  
  # pivot wider
  site_wide_df <- site_clustering_df %>% pivot_wider(names_from = layers, values_from = sum)
  
  # make tibble into data frame object
  site_wide_df <- as.data.frame(site_wide_df)
  
  # rename row names to species IDs
  site_wide_df2 <- site_wide_df[,-1]
  rownames(site_wide_df2) <- site_wide_df[,1]
  
  site_wide_matrix <- as.matrix(site_wide_df2)
  
  # transpose data frame so that we can cluster samples based on euclidan distance and complete linkage
  trans_site_wide_matrix <- t(site_wide_matrix)
  
  # calculate Bray-Curtis distance among samples
  site_dist_df <- vegan::vegdist(trans_site_wide_matrix, method = "bray")
  
  #Replace na values with 0 using is.na()
  site_dist_df[site_dist_df=="NaN"]<-0
  
  # Hierarchical clustering using Complete Linkage
  site_hc <- hclust(site_dist_df, method = "complete" )
  
  # get list of sample order for site
  
  site_order <- data.frame(order = site_hc$labels[site_hc$order])
  
  site_order_df <- rbind(site_order_df, site_order)
  
}


# add HP sample to the end of the order 
melted_relAbs_merged_summed$layers[melted_relAbs_merged_summed$site == "HP"] 
site_order_df <- rbind(site_order_df, "HP_HC_HMP_S0195_01")

# reorder samples based on concatenated hclust order
samples_order <- melted_relAbs_merged_summed$layers <- factor(melted_relAbs_merged_summed$layers, levels = site_order_df$order)

# manually reorder species based on pangenome/phylogney/ani  

species_manual_reorder <- c("N_cinerea",
                            "N_subflava_V", "N_subflava_I-VIII",
                            "N_mucosa", 
                            "N_elongata", "N_bacilliformis","N_sp_HMT_020", "K_bonacorsii_oralis",
                            "K_denitrificans",  "K_sp_str_SNUBH_2017", "S_muelleri", 
                            "N_oralis", "E_corrodens", "E_halliae", "E_exigua")

melted_relAbs_merged_summed$Group_ID_2 <- factor(melted_relAbs_merged_summed$Group_ID_2, levels = rev(species_manual_reorder))


dodge <- position_dodge(width=0.9)
Medium_panel_sizes <-c(10,1.5,1.5,10,10,1.5,0.5,1.5,1.5)

total_reads_mapped_rel_ab <- ggplot(data = melted_relAbs_merged_summed, aes(x=layers, y=total_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 25000000), breaks=seq(0,25000000,by=25000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("No. of reads mapped", "(0 - 25 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)


plot_total_reads_rel_ab <- ggplot(data = melted_relAbs_merged_summed, aes(x=layers, y=total_number_reads)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 200000000), breaks=seq(0,200000000,by=200000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Total reads", "(0 - 200 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.text.y=element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)

plot_percent_reads_mapped_rel_ab <- ggplot(data = melted_relAbs_merged_summed, aes(x=layers, y=percent_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 50), breaks=seq(0,50,by=50), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Percent reads mapped", "(0 - 50 %)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)



######### Relative abundance plot ######### 

# set colors for heat map
colors2 <- c("black", "#2166AC","#D1E5F0", "white", "#FDDBC7", "#F4A582", "#FF0000")


plot_group_level_rel_ab <- ggplot(data = melted_relAbs_merged_summed, aes(x=layers, y=Group_ID_2, fill=sum)) +
  geom_tile(height=1, width=1) + 
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") + 
  scale_fill_gradientn(colours = colors2,
                       limits=c(0, 100),
                       #breaks=seq(0.005,0.5,by=0.1),
                       breaks = c(0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100),
                       labels = c("0 %", "10 %", "20 %", "30 %", "40 %", "50 %", "60 %", "70 %", "80 %", "90 %", "100 %"),
                       oob=squish) + 
  scale_y_discrete(limits = levels(melted_relAbs_merged_summed$Group_ID_2)) +
  ylab(NULL) +  
  xlab(NULL) + 
  theme_classic() + 
  guides(fill = guide_colourbar(barwidth = 1,
                                barheight = 10,
                                title = "Relative abundance",
                                ticks = TRUE,
                                ticks.colour = "white",
                                ticks.linewidth = 0.5)) +
  theme(text = element_text(size = 10),
        axis.ticks.y = element_blank(),
        axis.ticks.length = unit(1, "cm"),
        axis.text.x=element_blank(),
        axis.text.y =  element_text(size = 9, color = "black"),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "black")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        #panel.border = element_rect(colour = "black", fill = NA),
        strip.text=element_text(size=12)) + 
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,1,1), "cm")) +
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)


Group_level_plot_rel_ab <- egg::ggarrange(plot_percent_reads_mapped_rel_ab, total_reads_mapped_rel_ab, plot_total_reads_rel_ab, plot_group_level_rel_ab,
                                          ncol = 1,
                                          heights = c(0.1,0.1,0.1,1))

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Relative_abundance_of_detected_species_groups_V2.pdf", Group_level_plot_rel_ab, width = 20, height = 10)

```


## Relative abundance V3

Samples sorted by no. of mapped reads and samples with zeroes included. 

```{r, eval = FALSE}

library("vegan")
library("ggdendro")

# Load data

# Load csv file with genome IDs and colors; this file was made when the tanglegram was made
bins_and_colors <-read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/bins_colors.csv", header = TRUE)

# Load breadth data
melted_df_detect <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Strain_level_updated_detection_data.csv", header = TRUE)

# Load mean Q2Q3 depth of coverage data
melted_df_coverage <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Strain_level_updated_Q2Q3_mean_coverage_data.csv", header = TRUE)


# Load sample metadata into R
sample_metadata <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/HMP_metadata/Final_1297_HMP_metagenomes_metadata.csv", header=TRUE)

# load mapping stats
mapping_stats <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/P_0003_Neisseriaceae-mapping_stats_new.txt", header=TRUE, sep = "\t")

# Merge the two 
mapping_metadata_merged <- merge(mapping_stats, sample_metadata, by.x = "layers", by.y = "Internal_ID")

# Add variable for percentage reads mapped
mapping_metadata_merged <- mapping_metadata_merged %>% 
  mutate(total_number_reads = (total_pairs_passed*2)) %>% 
  mutate(percent_reads_mapped = 100*(total_reads_mapped/total_number_reads))


# Keep samples where no genome was detected by setting coverage value to zero if genome "un-detected"
merged_melted_df <-merge(melted_df_coverage, melted_df_detect, by=c("layers","bins")) 
filtered_melted_df <- merged_melted_df %>% 
  dplyr::rename(coverage=value.x, breadth=value.y) %>% 
  dplyr::mutate(coverage = ifelse(detected == "un-detected", 0, coverage))
  

# select columns and transpose from long to wide
meanCov <- filtered_melted_df %>% 
  dplyr::select(layers, bins, coverage) %>% 
  pivot_wider(id_cols=bins, names_from=layers, values_from=coverage)

# chnage NA values to zeros
meanCov[is.na(meanCov)] <- 0

# transpose data frame
rownames <- meanCov$bins
meanCovT <- as.data.frame(t(meanCov[-1]))
colnames(meanCovT) <- rownames

# mean coverage of a genome / sum of mean coverages of all genomes in a sample
relAbsT <- decostand(meanCovT, method = "total")

# transpose relative abundance matrix
relAbs <- t(relAbsT)

# multiply by 100
relAbs <- 100*relAbs 

# change coverage data frame from wide to long format
melted_relAbs <- reshape2::melt(relAbs) %>% 
  dplyr::rename(variable=Var2, Genome_ID=Var1, relAbundance=value)

# merge with relabundance
melted_relAbs_merged <- merge(melted_relAbs,bins_and_colors, by.x = 'Genome_ID', by.y = 'Pangenome_ID')

# set Pangenome_group ID as factor
melted_relAbs_merged$Species_group <- as.factor(melted_relAbs_merged$Species_group)


# make oral site variable for facet grid
melted_relAbs_merged <- melted_relAbs_merged %>% mutate(site = case_when(grepl('TD_', variable) ~ 'TD', 
                                                                         grepl('KG_', variable) ~ 'KG', 
                                                                         grepl('PP_', variable) ~ 'SUPP',
                                                                         grepl('PB_', variable) ~ 'SUBP',
                                                                         grepl('SA_', variable) ~ 'SV',
                                                                         grepl('TH_', variable) ~ 'TH',
                                                                         grepl('PT_', variable) ~ 'PT',
                                                                         grepl('HP_', variable) ~ 'HP',
                                                                         grepl('BM_', variable) ~ 'BM')) 


# Change mucosa and subflava groups 
melted_relAbs_merged <- melted_relAbs_merged %>% 
  dplyr::mutate(Group_ID_2 = case_when(
    grepl("N_subflava_V", Species_group) ~ "N_subflava_V",
    grepl("N_subflava_I", Species_group) ~ "N_subflava_I-VIII",
    grepl("N_subflava_II", Species_group) ~ "N_subflava_I-VIII",
    grepl("N_subflava_III", Species_group) ~ "N_subflava_I-VIII",
    grepl("N_subflava_IV", Species_group) ~ "N_subflava_I-VIII",
    grepl("N_subflava_VII", Species_group) ~ "N_subflava_I-VIII",
    grepl("N_subflava_VIII", Species_group) ~ "N_subflava_I-VIII",
    grepl("N_mucosa", Species_group) ~ "N_mucosa",
    TRUE ~ Species_group  # Keeps all other Group_IDs the same
  ))


#Sum abundance by oral site, groupID and variable to obtain the relative abundance of each species (aka groupID) for each metagenome (aka variable)
melted_relAbs_merged_summed <- melted_relAbs_merged %>% 
  group_by(site, Group_ID_2, variable) %>% 
  dplyr::summarise(sum = sum(relAbundance))

# merge total reads mapped into data frame
melted_relAbs_merged_summed = merge.data.frame(mapping_metadata_merged, melted_relAbs_merged_summed, by.x = "layers", by.y = "variable") 

# reorder oral sites
melted_relAbs_merged_summed$site <- factor(melted_relAbs_merged_summed$site, levels = c("SUPP", "BM", "TD", "SUBP", "KG","PT","TH", "SV","HP"))


# manually reorder species based on pangenome/phylogney/ani  
species_manual_reorder <- c("N_cinerea",
                            "N_subflava_V", "N_subflava_I-VIII",
                            "N_mucosa", 
                            "N_elongata", "N_bacilliformis","N_sp_HMT_020", "K_bonacorsii_oralis",
                            "K_denitrificans",  "K_sp_str_SNUBH_2017", "S_muelleri", 
                            "N_oralis", "E_corrodens", "E_halliae", "E_exigua")

melted_relAbs_merged_summed$Group_ID_2 <- factor(melted_relAbs_merged_summed$Group_ID_2, levels = rev(species_manual_reorder))


dodge <- position_dodge(width=0.9)
Medium_panel_sizes <-c(10,10,10,2,1.5,2,1.75,0.75,0.5)
#levels = c("SUPP", "BM", "TD", "SUBP", "KG","PT","TH", "SV","HP"))

total_reads_mapped_rel_ab <- ggplot(data = melted_relAbs_merged_summed, aes(x=reorder(layers, -total_reads_mapped), y=total_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 25000000), breaks=seq(0,25000000,by=25000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("No. of reads mapped", "(0 - 25 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)


plot_total_reads_rel_ab <- ggplot(data = melted_relAbs_merged_summed, aes(x=reorder(layers, -total_reads_mapped), y=total_number_reads)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 200000000), breaks=seq(0,200000000,by=200000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Total reads", "(0 - 200 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.text.y=element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)

plot_percent_reads_mapped_rel_ab <- ggplot(data = melted_relAbs_merged_summed, aes(x=reorder(layers, -total_reads_mapped), y=percent_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 50), breaks=seq(0,50,by=50), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Percent reads mapped", "(0 - 50 %)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)



######### Relative abundance plot ######### 

# set colors for heat map
colors2 <- c("black", "#2166AC","#D1E5F0", "white", "#FDDBC7", "#F4A582", "#FF0000")


plot_group_level_rel_ab <- ggplot(data = melted_relAbs_merged_summed, aes(x=reorder(layers, -total_reads_mapped), y=Group_ID_2, fill=sum)) +
  geom_tile(height=1, width=1) + 
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") + 
  scale_fill_gradientn(colours = colors2,
                       limits=c(0, 100),
                       #breaks=seq(0.005,0.5,by=0.1),
                       breaks = c(0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100),
                       labels = c("0 %", "10 %", "20 %", "30 %", "40 %", "50 %", "60 %", "70 %", "80 %", "90 %", "100 %"),
                       oob=squish) + 
  scale_y_discrete(limits = levels(melted_relAbs_merged_summed$Group_ID_2)) +
  ylab(NULL) +  
  xlab(NULL) + 
  theme_classic() + 
  guides(fill = guide_colourbar(barwidth = 1,
                                barheight = 10,
                                title = "Relative abundance",
                                ticks = TRUE,
                                ticks.colour = "white",
                                ticks.linewidth = 0.5)) +
  theme(text = element_text(size = 10),
        axis.ticks.y = element_blank(),
        axis.ticks.length = unit(1, "cm"),
        axis.text.x=element_blank(),
        axis.text.y =  element_text(size = 9, color = "black"),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "black")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        #panel.border = element_rect(colour = "black", fill = NA),
        strip.text=element_text(size=12)) + 
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,1,1), "cm")) +
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)


Group_level_plot_rel_ab <- egg::ggarrange(plot_percent_reads_mapped_rel_ab, total_reads_mapped_rel_ab, plot_total_reads_rel_ab, plot_group_level_rel_ab,
                                          ncol = 1,
                                          heights = c(0.1,0.1,0.1,1))

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Relative_abundance_of_detected_species_groups_V3.pdf", Group_level_plot_rel_ab, width = 20, height = 10)

```


## Relative abundance V5 


Samples sorted by percentage mapped reads and samples with zeroes included. 

```{r, eval = FALSE}

dodge <- position_dodge(width=0.9)
Medium_panel_sizes <-c(10,10,10,2,1.5,2,1.75,0.75,0.5)
#levels = c("SUPP", "BM", "TD", "SUBP", "KG","PT","TH", "SV","HP"))

total_reads_mapped_rel_ab <- ggplot(data = melted_relAbs_merged_summed, aes(x=reorder(layers, -percent_reads_mapped), y=total_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 25000000), breaks=seq(0,25000000,by=25000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("No. of reads mapped", "(0 - 25 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)


plot_total_reads_rel_ab <- ggplot(data = melted_relAbs_merged_summed, aes(x=reorder(layers, -percent_reads_mapped), y=total_number_reads)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 200000000), breaks=seq(0,200000000,by=200000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Total reads", "(0 - 200 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.text.y=element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)

plot_percent_reads_mapped_rel_ab <- ggplot(data = melted_relAbs_merged_summed, aes(x=reorder(layers, -percent_reads_mapped), y=percent_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 50), breaks=seq(0,50,by=50), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Percent reads mapped", "(0 - 50 %)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)



######### Relative abundance plot ######### 

# set colors for heat map
colors2 <- c("black", "#2166AC","#D1E5F0", "white", "#FDDBC7", "#F4A582", "#FF0000")


plot_group_level_rel_ab <- ggplot(data = melted_relAbs_merged_summed, aes(x=reorder(layers, -percent_reads_mapped), y=Group_ID_2, fill=sum)) +
  geom_tile(height=1, width=1) + 
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") + 
  scale_fill_gradientn(colours = colors2,
                       limits=c(0, 100),
                       #breaks=seq(0.005,0.5,by=0.1),
                       breaks = c(0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100),
                       labels = c("0 %", "10 %", "20 %", "30 %", "40 %", "50 %", "60 %", "70 %", "80 %", "90 %", "100 %"),
                       oob=squish) + 
  scale_y_discrete(limits = levels(melted_relAbs_merged_summed$Group_ID_2)) +
  ylab(NULL) +  
  xlab(NULL) + 
  theme_classic() + 
  guides(fill = guide_colourbar(barwidth = 1,
                                barheight = 10,
                                title = "Relative abundance",
                                ticks = TRUE,
                                ticks.colour = "white",
                                ticks.linewidth = 0.5)) +
  theme(text = element_text(size = 10),
        axis.ticks.y = element_blank(),
        axis.ticks.length = unit(1, "cm"),
        axis.text.x=element_blank(),
        axis.text.y =  element_text(size = 9, color = "black"),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "black")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        #panel.border = element_rect(colour = "black", fill = NA),
        strip.text=element_text(size=12)) + 
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,1,1), "cm")) +
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)


Group_level_plot_rel_ab <- egg::ggarrange(plot_percent_reads_mapped_rel_ab, total_reads_mapped_rel_ab, plot_total_reads_rel_ab, plot_group_level_rel_ab,
                                          ncol = 1,
                                          heights = c(0.1,0.1,0.1,1))

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Relative_abundance_of_detected_species_groups_V5.pdf", Group_level_plot_rel_ab, width = 20, height = 10)

```
## Relative abundance V6

Samples sorted by heirarchical clustering and samples with zeroes included. 

```{r, eval = FALSE}

# Load data

# Load csv file with genome IDs and colors; this file was made when the tanglegram was made
bins_and_colors <-read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/bins_colors.csv", header = TRUE)

# Load breadth data
melted_df_detect <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Strain_level_updated_detection_data.csv", header = TRUE)

# Load mean Q2Q3 depth of coverage data
melted_df_coverage <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Strain_level_updated_Q2Q3_mean_coverage_data.csv", header = TRUE)


# Load sample metadata into R
sample_metadata <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/HMP_metadata/Final_1297_HMP_metagenomes_metadata.csv", header=TRUE)

# load mapping stats
mapping_stats <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/P_0003_Neisseriaceae-mapping_stats_new.txt", header=TRUE, sep = "\t")

# Merge the two 
mapping_metadata_merged <- merge(mapping_stats, sample_metadata, by.x = "layers", by.y = "Internal_ID")

# Add variable for percentage reads mapped
mapping_metadata_merged <- mapping_metadata_merged %>% 
  mutate(total_number_reads = (total_pairs_passed*2)) %>% 
  mutate(percent_reads_mapped = 100*(total_reads_mapped/total_number_reads))


# Keep samples where no genome was detected by setting coverage value to zero if genome "un-detected"
merged_melted_df <-merge(melted_df_coverage, melted_df_detect, by=c("layers","bins")) 
filtered_melted_df <- merged_melted_df %>% 
  dplyr::rename(coverage=value.x, breadth=value.y) %>% 
  dplyr::mutate(coverage = ifelse(detected == "un-detected", 0, coverage))


# select columns and transpose from long to wide
meanCov <- filtered_melted_df %>% 
  dplyr::select(layers, bins, coverage) %>% 
  pivot_wider(id_cols=bins, names_from=layers, values_from=coverage)

# chnage NA values to zeros
meanCov[is.na(meanCov)] <- 0

# transpose data frame
rownames <- meanCov$bins
meanCovT <- as.data.frame(t(meanCov[-1]))
colnames(meanCovT) <- rownames

# mean coverage of a genome / sum of mean coverages of all genomes in a sample
relAbsT <- decostand(meanCovT, method = "total")

# transpose relative abundance matrix
relAbs <- t(relAbsT)

# multiply by 100
relAbs <- 100*relAbs 

# change coverage data frame from wide to long format
melted_relAbs <- reshape2::melt(relAbs) %>% 
  dplyr::rename(variable=Var2, Genome_ID=Var1, relAbundance=value)

# merge with relabundance
melted_relAbs_merged <- merge(melted_relAbs,bins_and_colors, by.x = 'Genome_ID', by.y = 'Pangenome_ID')

# set Pangenome_group ID as factor
melted_relAbs_merged$Species_group <- as.factor(melted_relAbs_merged$Species_group)


# make oral site variable for facet grid
melted_relAbs_merged <- melted_relAbs_merged %>% mutate(site = case_when(grepl('TD_', variable) ~ 'TD', 
                                                                         grepl('KG_', variable) ~ 'KG', 
                                                                         grepl('PP_', variable) ~ 'SUPP',
                                                                         grepl('PB_', variable) ~ 'SUBP',
                                                                         grepl('SA_', variable) ~ 'SV',
                                                                         grepl('TH_', variable) ~ 'TH',
                                                                         grepl('PT_', variable) ~ 'PT',
                                                                         grepl('HP_', variable) ~ 'HP',
                                                                         grepl('BM_', variable) ~ 'BM')) 


# Change mucosa and subflava groups 
melted_relAbs_merged <- melted_relAbs_merged %>% 
  dplyr::mutate(Group_ID_2 = case_when(
    grepl("N_subflava_V", Species_group) ~ "N_subflava_V",
    grepl("N_subflava_I", Species_group) ~ "N_subflava_I-VIII",
    grepl("N_subflava_II", Species_group) ~ "N_subflava_I-VIII",
    grepl("N_subflava_III", Species_group) ~ "N_subflava_I-VIII",
    grepl("N_subflava_IV", Species_group) ~ "N_subflava_I-VIII",
    grepl("N_subflava_VII", Species_group) ~ "N_subflava_I-VIII",
    grepl("N_subflava_VIII", Species_group) ~ "N_subflava_I-VIII",
    grepl("N_mucosa", Species_group) ~ "N_mucosa",
    TRUE ~ Species_group  # Keeps all other Group_IDs the same
  ))


#Sum abundance by oral site, groupID and variable to obtain the relative abundance of each species (aka groupID) for each metagenome (aka variable)
melted_relAbs_merged_summed <- melted_relAbs_merged %>% 
  group_by(site, Group_ID_2, variable) %>% 
  dplyr::summarise(sum = sum(relAbundance))

# merge total reads mapped into data frame
melted_relAbs_merged_summed = merge.data.frame(mapping_metadata_merged, melted_relAbs_merged_summed, by.x = "layers", by.y = "variable") 

# reorder oral sites
melted_relAbs_merged_summed$site <- factor(melted_relAbs_merged_summed$site, levels = c("SUPP", "BM", "TD", "SUBP", "KG","PT","TH", "SV","HP"))

##### Sample order; need to cluster each site seperately, then get order, and then concatenate the orders ###


# created vector with 5 characters
columns= c("order")

# pass this vector length to ncol parameter
# and nrow with 0
site_order_df = data.frame(matrix(nrow = 0, ncol = length(columns)))

# assign column names
colnames(site_order_df) = columns


site_list <- list("SUPP", "SUBP", "KG", "BM", "TD", "PT", "TH", "SV") # ignore HP since there is only one sample


for (oral_site in site_list) {
  
  oral_site_factor <- trimws(oral_site)
  
  # filter df by site
  site_df <- melted_relAbs_merged_summed %>% 
    filter(site == oral_site_factor)
  
  # subset data frame for clustering species
  site_clustering_df <- site_df %>% 
    dplyr::select(Group_ID_2, layers, sum)
  
  # pivot wider
  site_wide_df <- site_clustering_df %>% pivot_wider(names_from = layers, values_from = sum)
  
  # make tibble into data frame object
  site_wide_df <- as.data.frame(site_wide_df)
  
  # rename row names to species IDs
  site_wide_df2 <- site_wide_df[,-1]
  rownames(site_wide_df2) <- site_wide_df[,1]
  
  site_wide_matrix <- as.matrix(site_wide_df2)
  
  # transpose data frame so that we can cluster samples based on euclidan distance and complete linkage
  trans_site_wide_matrix <- t(site_wide_matrix)
  
  # replace the missing values with 0 
  trans_site_wide_matrix[is.na(trans_site_wide_matrix)] <- 0 
  
  # calculate Bray-Curtis distance among samples
  site_dist_df <- vegan::vegdist(trans_site_wide_matrix, method = "bray")
  
  #Replace na values with 0 using is.na()
  site_dist_df[site_dist_df=="NaN"]<-0
  
  # Hierarchical clustering using Complete Linkage
  site_hc <- hclust(site_dist_df, method = "complete" )
  
  # get list of sample order for site
  
  site_order <- data.frame(order = rev(site_hc$labels[site_hc$order]))
  
  site_order_df <- rbind(site_order_df, site_order)
  
}


# add HP sample to the end of the order 
melted_relAbs_merged_summed$layers[melted_relAbs_merged_summed$site == "HP"] 
site_order_df <- rbind(site_order_df, "HP_HC_HMP_S0195_01")

# reorder samples based on concatenated hclust order
samples_order <- melted_relAbs_merged_summed$layers <- factor(melted_relAbs_merged_summed$layers, levels = site_order_df$order)



# manually reorder species based on pangenome/phylogney/ani  
species_manual_reorder <- c("N_cinerea",
                            "N_subflava_V", "N_subflava_I-VIII",
                            "N_mucosa", 
                            "N_elongata", "N_bacilliformis","N_sp_HMT_020", "K_bonacorsii_oralis",
                            "K_denitrificans",  "K_sp_str_SNUBH_2017", "S_muelleri", 
                            "N_oralis", "E_corrodens", "E_halliae", "E_exigua")

melted_relAbs_merged_summed$Group_ID_2 <- factor(melted_relAbs_merged_summed$Group_ID_2, levels = rev(species_manual_reorder))


dodge <- position_dodge(width=0.9)
#Medium_panel_sizes <-c(10,10,10,2,1.5,2,1.75,0.75,0.5)
Medium_panel_sizes <-c(5.444236,5.444236,5.444236,1.435862,1.129538,1.36271,1.377696,1.133856,1.11125)
#levels = c("SUPP", "BM", "TD", "SUBP", "KG","PT","TH", "SV","HP"))

total_reads_mapped_rel_ab <- ggplot(data = melted_relAbs_merged_summed, aes(x=layers, y=total_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 25000000), breaks=seq(0,25000000,by=25000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("No. of reads mapped", "(0 - 25 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)


plot_total_reads_rel_ab <- ggplot(data = melted_relAbs_merged_summed, aes(x=layers, y=total_number_reads)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 200000000), breaks=seq(0,200000000,by=200000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Total reads", "(0 - 200 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.text.y=element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)

plot_percent_reads_mapped_rel_ab <- ggplot(data = melted_relAbs_merged_summed, aes(x=layers, y=percent_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 50), breaks=seq(0,50,by=50), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Percent reads mapped", "(0 - 50 %)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 8),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)



######### Relative abundance plot ######### 

# set colors for heat map
colors <- c("black", "#2166AC", "white", "#F4A582", "#FF0000")


plot_group_level_rel_ab <- ggplot(data = melted_relAbs_merged_summed, aes(x=layers, y=Group_ID_2, fill=sum)) +
  geom_tile(height=1, width=1) + 
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") + 
  scale_fill_gradientn(colours = colors,
                       limits=c(0, 100),
                       #breaks=seq(0.005,0.5,by=0.1),
                       breaks = c(0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100),
                       labels = c("0 %", "10 %", "20 %", "30 %", "40 %", "50 %", "60 %", "70 %", "80 %", "90 %", "100 %"),
                       oob=squish) + 
  scale_y_discrete(limits = levels(melted_relAbs_merged_summed$Group_ID_2)) +
  ylab(NULL) +  
  xlab(NULL) + 
  theme_classic() + 
  guides(fill = guide_colourbar(barwidth = 1,
                                barheight = 10,
                                title = "Relative abundance",
                                ticks = TRUE,
                                ticks.colour = "white",
                                ticks.linewidth = 0.5)) +
  theme(text = element_text(size = 10),
        axis.ticks.y = element_blank(),
        axis.ticks.length = unit(1, "cm"),
        axis.text.x=element_blank(),
        axis.text.y =  element_text(size = 9, color = "black"),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "black")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        #panel.border = element_rect(colour = "black", fill = NA),
        strip.text=element_text(size=12)) + 
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,1,1), "cm")) +
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)


Group_level_plot_rel_ab <- egg::ggarrange(plot_percent_reads_mapped_rel_ab, total_reads_mapped_rel_ab, plot_total_reads_rel_ab, plot_group_level_rel_ab,
                                          ncol = 1,
                                          heights = c(0.1,0.1,0.1,1))

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Relative_abundance_of_detected_species_groups_V6.pdf", Group_level_plot_rel_ab, width = 20, height = 10)

```


# 13. Dominance & Reciprocity 

Analysis of the relative abundance of taxa (Figure 4) suggested patterns of dominance, where one species consistently had the highest abundance, or reciprocity, where shifts in one species' abundance were linked to changes in another, indicating potential interactions. For example, in the buccal mucosa (BM), N. subflava, N. mucosa, or K. oralis dominated each sample, implying competition for similar ecological niches and potential reciprocal influences on each others population levels.


Do individuals that have K. oralis in the plaque, also have it in the BM? 

If individuals who have one of the three taxa (K. oralis, N. mucosa, or N. subflava) in the buccal mucosa (BM) also have the same taxa in their respective preferred site (plaque for K. oralis and N. mucosa, tongue for N. subflava), it could suggest that these taxa are capable of colonizing multiple niches within the oral cavity, reflecting either their broad ecological tolerance or a functional role across different sites. If taxa are only found in BM and not their preferred site:

1. Specialized niche adaptation in BM: It could suggest that some populations of these species may have adapted specifically to thrive in BM, even though they are typically associated with plaque or tongue.

2. Transient colonization: This might indicate that their presence in BM is temporary or opportunistic, potentially reflecting a disruption in the individual's oral microbiome or an intermediate state before returning to their preferred site.

3. Lack of competitive success in preferred site: If K. oralis, N. mucosa, or N. subflava are only present in BM, it might imply that they are being outcompeted in their primary niches, or that other factors (e.g., oral hygiene, immune responses) are preventing their colonization in plaque or the tongue.



Findings:

We investigated the co-occurrence of K. oralis, N. mucosa, and N. subflava in both their respective preferred habitats (plaque for K. oralis and N. mucosa, tongue dorsum for N. subflava) and the buccal mucosa (BM). Our results indicate that a substantial proportion of individuals who have one of these taxa in the BM also have the same species in its preferred site. Specifically, 83.33% of individuals had K. oralis in the plaque, and 5.56% had it in the BM, with 4.63% of individuals having it in both the plaque and BM. Similarly, 42.59% of individuals had N. mucosa in the plaque, 12.96% had it in the BM, and 11.11% had it in both. For N. subflava, 70.83% of individuals had it on the tongue dorsum, 11.57% had it in the BM, and all individuals with it in the BM also had it on the tongue dorsum.

These results demonstrate that colonization of the BM complements colonization of the primary site, with dual presence in the BM and preferred sites across a subset of individuals. This finding suggests that the BM serves as a secondary niche for these taxa, where they can coexist or compete without displacing their dominance in their preferred habitats.

Dominance and Reciprocity: The reciprocal pattern observed in the BMwhere K. oralis, N. mucosa, and N. subflava alternate as dominant speciescould reflect localized competition within the BM. However, the consistent presence of these species in their respective preferred sites while they also colonize the BM suggests that competition within the BM does not preclude dominance in the preferred sites. This might mean that the BM serves as an extension or secondary niche for these taxa rather than a primary site of competition.

In this scenario, the reciprocal pattern observed in the BM could be due to transient fluctuations in environmental conditions or competitive pressures that allow one taxon to outcompete the others temporarily. However, because each species is still stably present in its preferred site, the BM may be a more dynamic or opportunistic niche, where different species may dominate at different times, but long-term stability is maintained in their preferred sites.

Microhabitats and Niche Flexibility: The fact that species like K. oralis, N. mucosa, and N. subflava are often present in both the BM and their preferred sites suggests that they may have a broader ecological range than originally thought. This could indicate that these taxa are flexible in their environmental requirements and capable of adapting to multiple microhabitats. For example:

K. oralis might utilize plaque as its main habitat for persistence but can also spread to the BM when conditions (e.g., nutrient availability, space) permit.
N. subflava may primarily specialize in the tongue dorsum but could colonize the BM if conditions are favorable.
This dual colonization suggests that these species have evolved mechanisms that allow them to move between or colonize multiple niches, which could provide a competitive advantage, especially in the context of fluctuating environmental conditions in the oral cavity.

```{r, eval = FALSE}

melted_relAbs_merged <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Relative_abundance_Neisseriaceae_genomes.csv", header = TRUE, stringsAsFactors = TRUE)

# Load sample metadata into R
sample_metadata <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/HMP_metadata/Final_1297_HMP_metagenomes_metadata.csv", header=TRUE, stringsAsFactors = TRUE)

# load mapping stats
mapping_stats <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/P_0003_Neisseriaceae-mapping_stats_new.txt", header=TRUE, sep = "\t", stringsAsFactors = TRUE)

# Merge the two 
mapping_metadata_merged <- merge(mapping_stats, sample_metadata, by.x = "layers", by.y = "Internal_ID")

# Add variable for percentage reads mapped
mapping_metadata_merged <- mapping_metadata_merged %>% 
  mutate(total_number_reads = (total_pairs_passed*2)) %>% 
  mutate(percent_reads_mapped = 100*(total_reads_mapped/total_number_reads))

#Sum abundance by oral site, groupID and variable to obtain the relative abundance of each species (aka groupID) for each metagenome (aka variable)
melted_relAbs_merged_summed <- melted_relAbs_merged %>% 
  group_by(site, Species_group, variable) %>% 
  dplyr::summarise(sum = sum(relAbundance))
  

 # merge total reads mapped into data frame
melted_relAbs_merged_summed = merge.data.frame(mapping_metadata_merged, melted_relAbs_merged_summed, by.x = "layers", by.y = "variable") 

select_melted_relAbs_merged_summed <- melted_relAbs_merged_summed %>% 
  dplyr::select(subject_id_new,layers,visit_number, subject_gender, site, total_number_reads, total_reads_mapped, percent_reads_mapped, sum, Species_group )


```


```{r, eval = FALSE}

# Load necessary packages
library(dplyr)

# Define the preferred sites for each species
preferred_sites <- c(
  "K_bonacorsii_oralis" = "SUPP",
  "N_mucosa_VII" = "SUPP",
  "N_subflava_IV" = "TD"
)

# Summarize data for each species (K. oralis, N. mucosa, N. subflava)
species_summary <- select_melted_relAbs_merged_summed %>%
  filter(Species_group %in% names(preferred_sites)) %>%
  group_by(Species_group, subject_id_new) %>%
  summarize(
    has_in_preferred = sum(site == preferred_sites[as.character(Species_group)] & sum > 0) > 0,
    has_in_BM = sum(site == "BM" & sum > 0) > 0
  ) %>%
  mutate(
    exclusive_to_preferred = has_in_preferred & !has_in_BM,
    exclusive_to_BM = !has_in_preferred & has_in_BM,
    in_both_preferred_and_BM = has_in_preferred & has_in_BM
  ) %>%
  group_by(Species_group) %>%
  summarize(
    total_subjects = n(),
    percent_in_preferred = 100 * sum(has_in_preferred) / total_subjects,
    percent_exclusive_to_preferred = 100 * sum(exclusive_to_preferred) / total_subjects,
    percent_in_BM = 100 * sum(has_in_BM) / total_subjects,
    percent_exclusive_to_BM = 100 * sum(exclusive_to_BM) / total_subjects,
    percent_in_both_preferred_and_BM = 100 * sum(in_both_preferred_and_BM) / total_subjects
  )

# View the summary
species_summary

write.csv(species_summary, "/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/FIGURES/BM_reciprocity_species_summary.csv", quote = FALSE, row.names = FALSE)
```




## Probability of detecting both K. oralis and N. mucosa in BM

Also, N. subflava and N. mucosa or K oralis...

```{r, eval = FALSE}

library(dplyr)

melted_strain_detection <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Relative_abundance_Neisseriaceae_genomes.csv", header = TRUE, stringsAsFactors = TRUE)

layers_subject_IDs <- melted_strain_detection %>% 
  dplyr::select(layers, subject_id_new) %>% 
  dplyr::distinct()

Relative_abundance_Neisseriaceae_genomes <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Relative_abundance_Neisseriaceae_genomes.csv", header = TRUE, stringsAsFactors = TRUE)


merged_df <- merge(Relative_abundance_Neisseriaceae_genomes,layers_subject_IDs,by.x = "variable", by.y = "layers", all.x = TRUE)


```

```{r, eval = FALSE}

df = merged_df

# Load dplyr library
library(dplyr)
library(tidyr)


# Filter data to only include rows for the BM site
BM_data <- df %>% filter(site == "BM")

# Calculate observed prevalence of K. bonacorsii_oralis in BM
K_oralis_prevalence <- BM_data %>%
  filter(Species_group == "K_bonacorsii_oralis") %>% 
  mutate(detected = if_else(relAbundance > 0, 1, 0)) %>% 
  summarize(K_oralis_prevalence = mean(detected)) %>%
  pull(K_oralis_prevalence)

# Calculate observed prevalence of N. mucosa_VII in BM
N_mucosa_prevalence <- BM_data %>%
  filter(Species_group == "N_mucosa_VII") %>%
  mutate(detected = if_else(relAbundance > 0, 1, 0)) %>% 
  summarize(N_mucosa_prevalence = mean(detected)) %>%
  pull(N_mucosa_prevalence)

# Calculate expected prevalence (if independent)
expected_prevalence <- K_oralis_prevalence * N_mucosa_prevalence

# Calculate observed prevalence of both K. bonacorsii_oralis and N. mucosa_VII in the same individual
observed_prevalence_both <- BM_data %>%
  filter(Species_group %in% c("K_bonacorsii_oralis", "N_mucosa_VII")) %>%
  group_by(subject_id_new, Species_group) %>%
  summarize(detected = if_else(any(relAbundance > 0), 1, 0)) %>%
  pivot_wider(names_from = Species_group, values_from = detected, values_fill = 0) %>%
  mutate(both_detected = if_else(K_bonacorsii_oralis == 1 & N_mucosa_VII == 1, 1, 0)) %>%
  summarize(both_detected_prevalence = mean(both_detected)) %>%
  pull(both_detected_prevalence)

# Print the results
cat("Observed prevalence of K. bonacorsii_oralis in BM:", K_oralis_prevalence, "\n")
cat("Observed prevalence of N. mucosa_VII in BM:", N_mucosa_prevalence, "\n")
cat("Expected prevalence of both K. bonacorsii_oralis and N. mucosa_VII:", expected_prevalence, "\n")
cat("Observed prevalence of both in BM:", mean(observed_prevalence_both), "\n")

```

Observed prevalence of K. bonacorsii_oralis in BM: 0.1061947 
Observed prevalence of N. mucosa_VII in BM: 0.06342183 
Expected prevalence of both K. bonacorsii_oralis and N. mucosa_VII: 0.006735061 
Observed prevalence of both in BM: 0.03614458 






# 14. Detection figure for poster

The goal here is to make a binary detection figure for a poster presented at Cold Spring Harbor Microbiome conference 2024. It needs to be simplified by removing undetected species and collapsing at to the species level. 
```{r}


species_detected <- c("N_cinerea",
"N_subflava_V", "N_subflava_VII","N_subflava_VIII","N_subflava_IV",
"N_subflava_III","N_subflava_I",  "N_subflava_II",
"N_mucosa_II",   "N_mucosa_III",  "N_mucosa_I", "N_mucosa_VII","N_mucosa_IV", "N_mucosa_V", "N_mucosa_VI",  
"N_elongata", "N_bacilliformis","N_sp_HMT_020", "K_bonacorsii_oralis",
"K_denitrificans",  "K_sp_str_SNUBH_2017", "S_muelleri", 
"N_oralis", "E_corrodens", "E_halliae", "E_exigua")
```



## Detection (50% breadth) Total Reads order

```{r, eval = FALSE}

library(tidyr)
library(dplyr)

Breadth_coverage_original <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/Strain_level_updated_detection_data.csv", header = TRUE)

Final_Groups <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/TESTS/Classifying_genomes_to_oral_sites/Neisseriaceae_genome_groups.txt", header = TRUE, sep = "\t")

# select relevant columns
Detection_original_selected <- Breadth_coverage_original %>% 
  dplyr::select(layers, bins, detection_binary, total_number_reads, total_reads_mapped, percent_reads_mapped)

# Merge with species groups info
Detection_original_with_groups <- merge(Detection_original_selected, Final_Groups, by.x = "bins", by.y = "Pangenome_ID", all.x = TRUE)

# Remove groups that are not detected
Detection_original_with_groups_filtered <- Detection_original_with_groups %>% 
  dplyr::filter(Group_ID %in% species_detected)

# Change mucosa and subflava groups 

Detection_original_with_groups_filtered <- Detection_original_with_groups_filtered %>% 
  dplyr::mutate(Group_ID_2 = case_when(
    grepl("N_subflava_V", Group_ID) ~ "N_subflava_V",
    grepl("N_subflava_I", Group_ID) ~ "N_subflava_I-VIII",
    grepl("N_subflava_II", Group_ID) ~ "N_subflava_I-VIII",
    grepl("N_subflava_III", Group_ID) ~ "N_subflava_I-VIII",
    grepl("N_subflava_IV", Group_ID) ~ "N_subflava_I-VIII",
    grepl("N_subflava_VII", Group_ID) ~ "N_subflava_I-VIII",
    grepl("N_subflava_VIII", Group_ID) ~ "N_subflava_I-VIII",
    grepl("N_mucosa", Group_ID) ~ "N_mucosa",
    TRUE ~ Group_ID  # Keeps all other Group_IDs the same
  ))




# Sum detection and mutate detection so that values >=1 become binary
sum_detection_df_per_group <- Detection_original_with_groups_filtered %>%
  group_by(layers, Group_ID_2) %>%
  summarise(sum_detection = sum(detection_binary)) %>% 
  mutate(accumulated_detection = ifelse(sum_detection >= 1, 1, 0))


# Add site column
sum_detection_df_per_group <- sum_detection_df_per_group %>% 
  mutate(site = case_when(grepl('TD_', layers) ~ 'TD', 
                          grepl('KG_', layers) ~ 'KG', 
                          grepl('PP_', layers) ~ 'SUPP',
                          grepl('PB_', layers) ~ 'SUBP',
                          grepl('SA_', layers) ~ 'SV',
                          grepl('TH_', layers) ~ 'TH',
                          grepl('PT_', layers) ~ 'PT',
                          grepl('HP_', layers) ~ 'HP',
                          grepl('BM_', layers) ~ 'BM'))

# Get sample read count data
sample_data <- Breadth_coverage_original %>% 
  dplyr::select(layers, total_number_reads, total_reads_mapped, percent_reads_mapped) %>% 
  distinct(layers, .keep_all = TRUE)

# merge back with sample info
sum_detection_df_per_group_merged_back <- merge(sum_detection_df_per_group, sample_data, by = "layers", all.x = TRUE)

# Set order of sites
sum_detection_df_per_group_merged_back$site <- factor(sum_detection_df_per_group_merged_back$site, levels = c("SUPP", "SUBP", "KG", "BM", "TD", "PT", "HP", "TH", "SV"))



# Species order based on HC
species_manual_reorder <- c("N_cinerea",
"N_subflava_V", "N_subflava_I-VIII",
"N_mucosa", 
"N_elongata", "N_bacilliformis","N_sp_HMT_020", "K_bonacorsii_oralis",
"K_denitrificans",  "K_sp_str_SNUBH_2017", "S_muelleri", 
"N_oralis", "E_corrodens", "E_halliae", "E_exigua")


sum_detection_df_per_group_merged_back$Group_ID_2 <- factor(sum_detection_df_per_group_merged_back$Group_ID_2, levels = species_manual_reorder)


##### FIGURE ##### 

dodge <- position_dodge(width=0.9)
Medium_panel_sizes <-c(4,1.5,1.5,4,4,1.5,0.5,1.5,1.5)
Medium_panel_sizes <-c(10,1.5,1.5,10,10,1.5,0.5,1.5,1.5)
Smaller_panel_sizes <- unlist(lapply(Medium_panel_sizes,"*",0.75))
Larger_panel_sizes <- unlist(lapply(Medium_panel_sizes,"*",1.2))


total_reads_mapped_detection <- ggplot(data = sum_detection_df_per_group_merged_back, aes(x=reorder(layers, -total_number_reads), y=total_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 25000000), breaks=seq(0,25000000,by=25000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("No. of reads mapped", "(0 - 25 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 6),
        axis.ticks.y = element_blank(),
        axis.text.x = element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x = unit(0.1, "lines"),
        plot.margin = unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)
#total_reads_mapped_detection

plot_total_reads_detection <- ggplot(data = sum_detection_df_per_group_merged_back, aes(x=reorder(layers, -total_number_reads), y=total_number_reads)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 200000000), breaks=seq(0,200000000,by=200000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Total reads", "(0 - 200 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 6),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.text.y=element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)
#plot_total_reads_detection


plot_percent_reads_mapped_detection <- ggplot(data = sum_detection_df_per_group_merged_back, aes(x=reorder(layers, -total_number_reads), y=percent_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 50), breaks=seq(0,50,by=50), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Percent reads mapped", "(0 - 50 %)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 6),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)
#plot_percent_reads_mapped_detection


plot_species_level_detection <- ggplot(data = sum_detection_df_per_group_merged_back, aes(x=reorder(layers,-total_number_reads), y=Group_ID_2, fill=as.factor(accumulated_detection))) +
  geom_tile(height=1, width=1) + 
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") + 
  scale_fill_manual(values = c("black", "cyan"),
                    labels = c("Un-detected (Breadth < 50%)", "Detected (Breadth > 50%)"),
                    name = "Species detection") +
  scale_y_discrete(limits = rev(levels(sum_detection_df_per_group_merged_back$Group_ID))) +
  ylab(NULL) +  
  xlab(NULL) + 
  theme_classic() + 
  theme(text = element_text(size = 10),
        legend.position = "bottom",
        axis.ticks.y = element_line(linewidth = 1, linetype = "solid"),
        axis.ticks.length = unit(0.25, "cm"),
        axis.text.x=element_blank(),
        axis.text.y =  element_text(size = 10, color = "black"),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "black")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text=element_text(size=10)) + 
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,1,1), "cm")) +
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)
#plot_species_level_detection


Species_detection_total_reads <- egg::ggarrange(plot_percent_reads_mapped_detection, total_reads_mapped_detection,plot_total_reads_detection, plot_species_level_detection,
                                                    ncol = 1,
                                                    heights = c(0.05,0.05,0.05,1))

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/POSTER_Species_detection_plot.pdf", Species_detection_total_reads, width = 20, height = 6)



```


## Detection samples HC order
```{r}

##### Sample order; need to cluster each site separately, then get order, and then concatenate the orders ###

# created vector with 5 characters
columns= c("order")

# pass this vector length to ncol parameter
# and nrow with 0
site_order_df = data.frame(matrix(nrow = 0, ncol = length(columns)))

# assign column names
colnames(site_order_df) = columns


site_list <- list("SUPP", "SUBP", "KG", "BM", "TD", "PT", "TH", "SV") # ignore HP since there is only one sample


for (oral_site in site_list) {
  
  
  oral_site_factor <- trimws(oral_site)
  
  sum_detection_df_per_group_merged_back$accumulated_detection <- as.numeric(as.character(sum_detection_df_per_group_merged_back$accumulated_detection))
                             
  # filter df by site
  site_df <- sum_detection_df_per_group_merged_back %>% 
    dplyr::filter(site == oral_site_factor) 
  
  # subset data frame for clustering species
  site_clustering_df <- site_df %>% 
    select(Group_ID_2, layers, accumulated_detection)
  
  # pivot wider
  site_wide_df <- site_clustering_df %>% pivot_wider(names_from = layers, values_from = accumulated_detection)
  
  # make tibble into data frame object
  site_wide_df <- as.data.frame(site_wide_df)
  
  # rename row names to species IDs
  site_wide_df2 <- site_wide_df[,-1]
  rownames(site_wide_df2) <- site_wide_df[,1]
  
  # reformat to matrix
  site_wide_matrix <- as.matrix(site_wide_df2)
  
  # transpose data frame so that we can cluster samples based on euclidian distance and complete linkage
  trans_site_wide_matrix <- t(site_wide_matrix)
  
  # calculate Bray-Curtis distance among samples
  site_dist_df <- vegan::vegdist(trans_site_wide_matrix, method = "jaccard")
  
  #Replace na values with 0 using is.na()
  site_dist_df[site_dist_df=="NaN"]<-0
  
  # Hierarchical clustering using Complete Linkage
  site_hc <- hclust(site_dist_df, method = "complete" )
  

  # get list of sample order for site
  
  site_order <- data.frame(order = site_hc$labels[site_hc$order])
  
  site_order_df <- rbind(site_order_df, site_order)
  
}


# add HP sample (HP_HC_HMP_S0195_01) to the end of the order 
#melted_strain_detection$layers[melted_strain_detection$site == "HP"] 
site_order_df <- rbind(site_order_df, "HP_HC_HMP_S0195_01")

# reorder samples based on concatenated hclust order
samples_order <- sum_detection_df_per_group_merged_back$layers <- factor(sum_detection_df_per_group_merged_back$layers, levels = site_order_df$order)

##### FIGURE ##### 



total_reads_mapped_detection <- ggplot(data = sum_detection_df_per_group_merged_back, aes(x=layers, y=total_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 25000000), breaks=seq(0,25000000,by=25000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("No. of reads mapped", "(0 - 25 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 6),
        axis.ticks.y = element_blank(),
        axis.text.x = element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x = unit(0.1, "lines"),
        plot.margin = unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)
#total_reads_mapped_detection

plot_total_reads_detection <- ggplot(data = sum_detection_df_per_group_merged_back, aes(x=layers, y=total_number_reads)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 200000000), breaks=seq(0,200000000,by=200000000), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Total reads", "(0 - 200 million)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 6),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.line.x = element_blank(),
        axis.text.y=element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)
#plot_total_reads_detection


plot_percent_reads_mapped_detection <- ggplot(data = sum_detection_df_per_group_merged_back, aes(x=layers, y=percent_reads_mapped)) + 
  geom_bar(stat = 'identity',position = dodge, color = "darkblue") +
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") +
  scale_y_continuous(expand = c(0, 0), limits=c(0, 50), breaks=seq(0,50,by=50), oob=squish, labels = scales::comma) +
  #coord_cartesian(ylim=c(0,100000000)) +
  xlab(NULL)+
  ylab(expression(atop("Percent reads mapped", "(0 - 50 %)"))) +  # Modified y-axis label
  theme_classic() + 
  theme(text = element_text(size = 6),
        axis.ticks.y = element_blank(),
        axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = -270, vjust = 0.5, hjust = 0.5),  # Try -90 degrees rotation
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "lightblue")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text.x = element_blank()) +
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,-0.001,1), "cm"))+
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)
#plot_percent_reads_mapped_detection


plot_species_level_detection <- ggplot(data = sum_detection_df_per_group_merged_back, aes(x=layers, y=Group_ID_2, fill=as.factor(accumulated_detection))) +
  geom_tile(height=1, width=1) + 
  facet_grid(. ~ site, scales = "free", space = "free", switch = "x") + 
  scale_fill_manual(values = c("black", "cyan"),
                    labels = c("Un-detected (Breadth < 50%)", "Detected (Breadth > 50%)"),
                    name = "Species detection") +
  scale_y_discrete(limits = rev(levels(sum_detection_df_per_group_merged_back$Group_ID))) +
  ylab(NULL) +  
  xlab(NULL) + 
  theme_classic() + 
  theme(text = element_text(size = 10),
        legend.position = "bottom",
        axis.ticks.y = element_line(linewidth = 1, linetype = "solid"),
        axis.ticks.length = unit(0.25, "cm"),
        axis.text.x=element_blank(),
        axis.text.y =  element_text(size = 10, color = "black"),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x=element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "black")) + 
  theme(strip.background = element_blank(),
        panel.border = element_blank(),
        strip.text=element_text(size=10)) + 
  theme(panel.spacing.x=unit(0.1, "lines"),
        plot.margin=unit(c(-0.001,1,1,1), "cm")) +
  force_panelsizes(cols = unit(Medium_panel_sizes, "cm"), 
                   TRUE)
#plot_species_level_detection


Species_detection_HC <- egg::ggarrange(plot_percent_reads_mapped_detection, total_reads_mapped_detection,plot_total_reads_detection, plot_species_level_detection,
                                                    ncol = 1,
                                                    heights = c(0.05,0.05,0.05,1))

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/13_DETECTED_GENOMES/POSTER_Species_detection_plot_HC_rorder.pdf", Species_detection_HC, width = 20, height = 6)



```