---
title: "COG20 & Pfam Cross-check"
author: "J. J. Giacomini"
date: "2024-11-15"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

PURPOSE: To generate a COG20 and Pfam table that can be used to cross-validate functional enrichment reuslts from metapangenomics analyses. 

METHODS: Using the functional annotation data within the Contigs or Genomes DB, in which each gene caller ID has a set of functional annotations, one row for each functional source, we generate a table that associates each COG20 accession with Pfam accessions. 

```{r, eval=FALSE}

library("dplyr")

Pangenome_metadata

gene_functions_df <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/20_FUNCTIONAL_ENRICHMENT/N_subflava_vs_N_mucosa_pangenome/gene_clusters_with_functional_annotations.tsv", header = TRUE, sep = "\t",stringsAsFactors = TRUE, fill = TRUE, quote = "")

gene_functions_df_COG_Pfam <- gene_functions_df %>% 
  dplyr::filter(source == "COG20_FUNCTION" | source == "Pfam") 
  
gene_functions_df_COG_Pfam_2 <- merge(gene_functions_df_COG_Pfam, Pangenome_metadata, by.x = "genome_name", by.y = "item", all.x = TRUE)

# Step 1: Filter out rows for COG20 and Pfam accessions
cog20_df <- gene_functions_df_COG_Pfam %>%
  filter(source == "COG20_FUNCTION") %>%
  select(gene_caller_id, accession) %>%
  rename(COG20_accession = accession)

pfam_df <- gene_functions_df_COG_Pfam %>%
  filter(source == "Pfam") %>%
  select(gene_caller_id, accession) %>%
  rename(Pfam_accession = accession)

# Step 2: Join the data frames on gene_caller_id to find shared annotations
cog_pfam_associations <- inner_join(cog20_df, pfam_df, by = "gene_caller_id")

# Step 3: Create a summary table associating each COG20 with its Pfam accessions
# Group by COG20_accession to create a unique list of associated Pfam_accessions
cog_pfam_summary <- cog_pfam_associations %>%
  group_by(COG20_accession) %>%
  summarise(Pfam_accessions = paste(unique(Pfam_accession), collapse = "; "))

# Step 4: Handle cases where COG20 accessions do not have any Pfam associations
# Find all unique COG20 accessions that were not matched in the join
cog20_only <- cog20_df %>%
  anti_join(cog_pfam_associations, by = "gene_caller_id") %>%
  distinct(COG20_accession) %>%
  mutate(Pfam_accessions = NA)

# Step 5: Combine COG20s with and without Pfam associations
final_summary <- bind_rows(cog_pfam_summary, cog20_only)

# View the resulting summary table
print(final_summary)

# Optionally, write the result to a CSV file
write.table(final_summary, "/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/20_FUNCTIONAL_ENRICHMENT/N_subflava_vs_N_mucosa_pangenome/COG20_Pfam_associations.txt", row.names = FALSE, quote = FALSE, sep = "\t")


```