---
title: "R Notebook"
output: html_notebook
---


Regarding the Human Oral Metapangenomics Project, an important concern arises when considering bacterial genomic assemblies â€“ the presence of short contigs, typically measuring between 300 to 1000 nucleotides in length. Two fundamental questions emerge: firstly, how frequently do these short contigs occur in bacterial isolate reference genomes, and secondly, what impact, if any, do these short contigs have on metagenomic read recruitment and subsequent coverage estimates, including genome mean coverage, Q2Q3 mean coverage, and genome breadth of coverage (also known as detection)?

To address these questions, I conducted an in-depth analysis of contig lengths within selected Neisseriaceae RefSeq isolate genome assemblies. The results are presented in the accompanying bar chart, illustrating the prevalence of short contigs in Neisseria, Eikenella, and Kingella assemblies.

Key findings from this analysis include: 

Out of the 488 genomes scrutinized, 291 genomes harbor at least one contig shorter than 1000 nucleotides (as indicated by the blue bars on the chart).
Additionally, among the 488 genomes, 230 feature at least one contig shorter than 500 nucleotides (highlighted in green).
Furthermore, 205 genomes within the dataset contain at least one contig shorter than 300 nucleotides (shown in red).
These observations suggest that short contigs are a common occurrence in a significant proportion of publicly available bacterial RefSeq isolate reference genomes hosted on NCBI.


Here is the code used to generate the results above:

The follwing script uses anvio reformat program to determine how many contigs are less than a sequence of lengths between 100 and 1000 by intervals of 100. 

```{bash Get distribution of short contigs}

clusterize -n 20 -l LOGS/contig_lengths_per_genome.log /workspace/jmarkwelchlab/P_0003_Neisseriaceae/SCRIPTS/contig_lengths_per_genome.sh

cp CONTIG_LENGTH_TEST/final_table.txt CONTIG_LENGTH_TEST/final_table_og_copy.txt
cat CONTIG_LENGTH_TEST/final_table.txt | sed -e 's/%//g' > CONTIG_LENGTH_TEST/final_table.txt
rm CONTIG_LENGTH_TEST/final_table_og_copy.txt
```

```{bash Get distribution of short contigs continued}

clusterize -n 10 -l LOGS/contig_stats_per_genome.log /workspace/jmarkwelchlab/P_0003_Neisseriaceae/SCRIPTS/contig_stats_per_genome.sh

anvi-display-contigs-stats $DIR_Contigs/G_0278-contigs.db
```


```{bash send reuslts to local}

scp -r jgiacomini@evol5.mbl.edu:/workspace/jmarkwelchlab/P_0003_Neisseriaceae/CONTIG_LENGTH_TEST /Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/
```


```{r Generate plot of distribution of short contigs in genomes}

# Load the dplyr package
library(dplyr)

#### Number of genomes with contigs less than specified lengths...


df = read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/final_table.txt", header = TRUE)

# Group the data by length and summarize the count
summary_table <- df %>%
  group_by(contig_length_threshold) %>%
  summarize(NonZeroGenomes = sum(contigs_removed > 0))

# Print the summary table
print(summary_table)


# Filter rows where count is greater than 0, group by length, and calculate the average count
average_table <- df %>%
  filter(contigs_removed > 0) %>%
  group_by(contig_length_threshold) %>%
  summarize(AverageCount = mean(contigs_removed),
            StDevCount = sd(contigs_removed))

# Print the average table
print(average_table)


# Load the necessary libraries
library(ggplot2)

# Create the heatmap using ggplot2
heatmap_plot <- ggplot(df, aes(x = reorder(genome, -contigs_removed), y = contig_length_threshold, fill = contigs_removed)) +
  geom_tile() +
  scale_fill_gradient(low = "blue", high = "red") +
  labs(
    title = "Heatmap of short contigs within isolate genome assemblies",
    x = "Genome",
    y = "Contig length (No. of nucelotides)",
    fill = "Count of contigs less than each length") 
# Print the heatmap
print(heatmap_plot)



# Create the heatmap using ggplot2
heatmap_plot_perecent <- ggplot(df, aes(x = reorder(genome, -percent_contigs_removed), y = contig_length_threshold, fill = percent_contigs_removed)) +
  geom_tile() +
  scale_fill_gradient(low = "blue", high = "red") +
  labs(
    title = "Heatmap of % short contigs within isolate genome assemblies",
    x = "Genome",
    y = "Contig length (No. of nucelotides)",
    fill = "% of contigs less than each length") 
# Print the heatmap
print(heatmap_plot_perecent)


df2 = read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/Contig_stats.txt", header = TRUE)

str(df2)

bar_plot_1000 <- ggplot(df2, aes(x = reorder(Genome, -Count_contigs_less_than_1000), y = Count_contigs_less_than_1000)) +
  geom_bar(stat = "identity", color = "steelblue") +
  labs(x = "Genome (n = 488)",
    y = "No. of contigs < 1k nts") +
  scale_y_continuous(expand = c(0, 0)) +
  theme_classic() + 
  theme(text = element_text(size = 12, color = "black"),
        axis.ticks.y = element_line(linewidth = 1),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())
  

bar_plot_1000 

bar_plot_500 <- ggplot(df2, aes(x = reorder(Genome, -Count_contigs_less_than_500), y = Count_contigs_less_than_500)) +
  geom_bar(stat = "identity", color = "orange") +
  labs(x = "Genome (n = 488)",
    y = "No. of contigs < 500 nts") +
  scale_y_continuous(expand = c(0, 0)) +
  theme_classic() + 
  theme(text = element_text(size = 12, color = "black"),
        axis.ticks.y = element_line(linewidth = 1),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())
  

bar_plot_500 

df2_subset_1000_500_300 <- df2 %>% 
  select(Genome, Count_contigs_less_than_1000, Count_contigs_less_than_500, Count_contigs_less_than_300)

library(reshape2)
melted_df2_subset_1000_500_300<-melt(df2_subset_1000_500_300, id="Genome")

library(gtools)

bar_plot_1000_500_300 <- ggplot(melted_df2_subset_1000_500_300, aes(x = reorder(Genome, -value), y = value, fill=variable)) +
  geom_bar(stat="identity",position = "identity") +
  labs(x = "Genome (n = 488)",
    y = "No. of contigs") +
  scale_y_continuous(expand = c(0,0)) +
  scale_fill_manual(values = c("blue", "green", "red"),
                    labels = c("< 1000", "< 500", "< 300"),
                    name = "Contig length (nt)") + 
  theme_classic() + 
  theme(text = element_text(size = 12, color = "black"),
        axis.ticks.y = element_line(linewidth = 1),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())
  

bar_plot_1000_500_300 

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/bar_plot_short_contigs_per_genome.pdf", bar_plot_1000_500_300, width = 5, height = 3)

```


Referring to the figure, a sizeable portion of the genomes have short contigs (defined here as contigs less than 1000 nt in length) - 291 out of 488 genomes have at least one contig shorter than 1000 nucleotides (blue); 230 out of 488 genomes have at least one contig shorter than 500 nucleotides (green); 205 out of 488 genomes have at least one contig shorter than 300 nucleotides (red).


To tackle the second question, we can compare mean coverage estimates obtained from read recruitment for two distinct scenarios: genomes that are free from short contigs and those that contain such short contigs. Additionally, we can further explore the impact of short contigs by contrasting genome coverage obtained from read recruitment for all contigs against coverage derived only from contigs exceeding a specified length threshold (e.g., 300 nucleotides).

This can be achieved by mapping reads from a subset of available HMP (Human Microbiome Project) metagenomes where we anticipate the presence of specific taxa. For instance, MetaPhlan 4 results indicate the high relative abundance of Neisseria subflava in numerous tongue dorsum samples. Our prior analysis of contig statistics per genome revealed a promising candidate genome for N. subflava (N_subflava_str_UMB1050_id_GCA_030218065_1) within our dataset, notable for its substantial number of short contigs (158 out of 229 contigs less than 300 nucleotides in length). Importantly, this genome closely clusters with other N. subflava genomes in a single copy core gene (SSCG) phylogeny, indicating that it is not an outlier genome despite the prevalence of short contigs.

Furthermore, we can expand this analysis to include a Kingella kingae genome, which is expected to occur in tongue dorsum samples but at a relatively lower abundance. This will allow us to investigate whether the influence of short contigs on coverage estimates is dependent on the relative abundance of the taxon. For both taxa (N. subflava and K. kingae), our dataset includes both fragmented and complete genomes. Notably, similar to N. subflava, a highly fragmented isolate reference genome for K. kingella also aligns well with other K. kingella genomes in the SCCG phylogeny.

To quantify the statistical impact of short contigs, we propose the following model:

`Coverage = Genome Type * Filter Type * Taxon`

Here, "Coverage" represents either mean genome coverage, mean Q2Q3 coverage, or breadth of coverage. "Genome Type" distinguishes between complete (comprising a single contig) and fragmented genomes (containing multiple short contigs). "Filter Type" represents read recruitment profiles, which can be applied to all contigs or only those exceeding 300 nucleotides in length. Finally, "Taxon" specifies either N. subflava or K. kingella.

By implementing this model, we aim to elucidate the nuanced influence of short contigs on various coverage metrics across these distinct scenarios.



# 1. Set up directories, variables and files needed
screen -S qrsh_sesh_2
qrsh -pe allslots 10

module purge
module load miniconda/3
source /bioware/miniconda3/bashrc
conda activate anvio-7.1
module load clusters/barhal
module load jbpc



```{bash}

projectID=P_0003_Neisseriaceae
mainDIR=/workspace/jmarkwelchlab
DIR_CLE=$mainDIR/$projectID/CONTIG_LENGTH_TEST

DIR_Assemblies=$mainDIR/$projectID/02_ASSEMBLIES
DIR_Contigs=$DIR_CLE/03_GENOMES_EDITED
DIR_ContigsDB=$DIR_CLE/04_CONTIGS_DB
DIR_Mapping=$DIR_CLE/05_MAPPING
DIR_Profiling=$DIR_CLE/06_PROFILES
DIR_Summary=$DIR_CLE/08_PROFILE_SUMMARY


DIR_Data=$mainDIR/$projectID/DATA
CLE_genomesID=$DIR_Data/CONTIG_LENGTH_TEST_id_genomes.txt

mkdir $DIR_Contigs $DIR_ContigsDB $DIR_Mapping $DIR_Profiling $DIR_Summary


# make CONTIG_LENGTH_TEST_id_genomes.txt file
G_0421	N_subflava_str_UMB1050_id_GCA_030218065_1
G_0400	K_kingae_str_KK100_id_GCA_030182025_1
G_0271	K_kingae_str_ATCC_23332_id_GCA_011612385_1
G_0250	N_subflava_str_ATCC_49275_id_GCA_005221305_1

echo 'G_0421' > $DIR_Data/CONTIG_LENGTH_TEST_id_genomes.txt
echo 'G_0400' >> $DIR_Data/CONTIG_LENGTH_TEST_id_genomes.txt
echo 'G_0271' >> $DIR_Data/CONTIG_LENGTH_TEST_id_genomes.txt 
echo 'G_0250' >> $DIR_Data/CONTIG_LENGTH_TEST_id_genomes.txt



# make genome ID name conversions file
CONTIG_LENGTH_TEST_name_conversions=$DIR_Data/CONTIG_LENGTH_TEST_name_conversions.txt 
cat > $CONTIG_LENGTH_TEST_name_conversions

G_0421	N_subflava_str_UMB1050_id_GCA_030218065_1
G_0400	K_kingae_str_KK100_id_GCA_030182025_1
G_0271	K_kingae_str_ATCC_23332_id_GCA_011612385_1
G_0250	N_subflava_str_ATCC_49275_id_GCA_005221305_1

<!-- control + D -->

```


# 2. Reformat deflines

```{bash}

for genome in `cat $CLE_genomesID`
do
RAW_GENOMES=$DIR_Assemblies/${genome}-RAW.fa
EDITED_GENOMES=$DIR_Contigs/${genome}.fa
REPORT=$DIR_Contigs/${genome}.report.tsv
anvi-script-reformat-fasta -o $EDITED_GENOMES --simplify-names --prefix ${genome} --seq-type NT -r $REPORT $RAW_GENOMES 
done

```


Input ........................................: /workspace/jmarkwelchlab/P_0003_Neisseriaceae/02_ASSEMBLIES/G_0421-RAW.fa
Output .......................................: /workspace/jmarkwelchlab/P_0003_Neisseriaceae/CONTIG_LENGTH_TEST/03_GENOMES_EDITED/G_0421.fa
Minimum length ...............................: 0
Max % gaps allowed ...........................: 100.00%
Total num contigs ............................: 229
Total num nucleotides ........................: 2,332,356
Contigs removed ..............................: 0 (0.00% of all)
Nucleotides removed ..........................: 0 (0.00% of all)
Nucleotides modified .........................: 0 (0.00% of all)
Deflines simplified ..........................: True

Input ........................................: /workspace/jmarkwelchlab/P_0003_Neisseriaceae/02_ASSEMBLIES/G_0400-RAW.fa
Output .......................................: /workspace/jmarkwelchlab/P_0003_Neisseriaceae/CONTIG_LENGTH_TEST/03_GENOMES_EDITED/G_0400.fa
Minimum length ...............................: 0
Max % gaps allowed ...........................: 100.00%
Total num contigs ............................: 406
Total num nucleotides ........................: 2,070,043
Contigs removed ..............................: 0 (0.00% of all)
Nucleotides removed ..........................: 0 (0.00% of all)
Nucleotides modified .........................: 0 (0.00% of all)
Deflines simplified ..........................: True

Input ........................................: /workspace/jmarkwelchlab/P_0003_Neisseriaceae/02_ASSEMBLIES/G_0271-RAW.fa
Output .......................................: /workspace/jmarkwelchlab/P_0003_Neisseriaceae/CONTIG_LENGTH_TEST/03_GENOMES_EDITED/G_0271.fa
Minimum length ...............................: 0
Max % gaps allowed ...........................: 100.00%
Total num contigs ............................: 1
Total num nucleotides ........................: 2,058,531
Contigs removed ..............................: 0 (0.00% of all)
Nucleotides removed ..........................: 0 (0.00% of all)
Nucleotides modified .........................: 0 (0.00% of all)
Deflines simplified ..........................: True

Input ........................................: /workspace/jmarkwelchlab/P_0003_Neisseriaceae/02_ASSEMBLIES/G_0250-RAW.fa
Output .......................................: /workspace/jmarkwelchlab/P_0003_Neisseriaceae/CONTIG_LENGTH_TEST/03_GENOMES_EDITED/G_0250.fa
Minimum length ...............................: 0
Max % gaps allowed ...........................: 100.00%
Total num contigs ............................: 1
Total num nucleotides ........................: 2,195,659
Contigs removed ..............................: 0 (0.00% of all)
Nucleotides removed ..........................: 0 (0.00% of all)
Nucleotides modified .........................: 0 (0.00% of all)
Deflines simplified ..........................: True

# 3. Build Anvio contigs databases
```{bash}

for genome in `cat $CLE_genomesID`
do
EDITED_GENOMES=$DIR_Contigs/${genome}.fa
contigsDB=$DIR_ContigsDB/${genome}-contigs.db
numThreads=10
anvi-gen-contigs-database -f $EDITED_GENOMES -n ${genome} -o $contigsDB -T $numThreads 
done
```

# 4. Concatenate fasta files for mapping (if needed) and reformat once again

The following code generated three types opf concatenated fasta files and their respective binning and docmposition files that aren need downstream for profiling with anvio. The three types are as follows:

1. Complete genomes only
2. Fragmented genomes only
3. Both complete and fragmented genomes together


Fragmented
G_0400 G_0421

Complete
G_0250 G_0271

Both
G_0400 G_0421 G_0250 G_0271

```{bash}

EDITED_GENOMES=$DIR_Contigs/${genome}.fa

RAW_CONCAT_GENOMES_COMPLETE=$DIR_Contigs/Raw_concatenated_complete.fa
RAW_CONCAT_GENOMES_FRAGED=$DIR_Contigs/Raw_concatenated_fragmented.fa
RAW_CONCAT_GENOMES_BOTH=$DIR_Contigs/Raw_concatenated_both.fa

EDITED_CONCAT_GENOMES_COMPLETE=$DIR_Contigs/Edited_Concatenated_complete.fa
EDITED_CONCAT_GENOMES_FRAGED=$DIR_Contigs/Edited_Concatenated_fragmented.fa
EDITED_CONCAT_GENOMES_BOTH=$DIR_Contigs/Edited_Concatenated_both.fa

PROJECT_REPORT_COMPLETE=$DIR_Contigs/complete_report.tsv
PROJECT_REPORT_FRAGED=$DIR_Contigs/fragmented_report.tsv
PROJECT_REPORT_BOTH=$DIR_Contigs/both_report.tsv

BINNING_COMPLETE=$DIR_Contigs/complete_binning.tsv
BINNING_FRAGED=$DIR_Contigs/fragmented_binning.tsv
BINNING_BOTH=$DIR_Contigs/both_binning.tsv

DECOMPOSE_COMPLETE=$DIR_Contigs/complete_decompose.tsv
DECOMPOSE_FRAGED=$DIR_Contigs/fragmented_decompose.tsv
DECOMPOSE_BOTH=$DIR_Contigs/both_decompose.tsv

CONTIG_LENGTH_TEST_name_conversions=$DIR_Data/CONTIG_LENGTH_TEST_name_conversions.txt 


# make concatenated genome file for complete genomes
for genome in G_0250 G_0271
do
EDITED_GENOMES=$DIR_Contigs/${genome}.fa
cat $EDITED_GENOMES >> $RAW_CONCAT_GENOMES_COMPLETE
done

# make concatenated genome file for fragmented genomes
for genome in G_0400 G_0421
do
EDITED_GENOMES=$DIR_Contigs/${genome}.fa
cat $EDITED_GENOMES >> $RAW_CONCAT_GENOMES_FRAGED
done

# make concatenated genome file for both fragmented and complete genomes
for genome in G_0250 G_0271 G_0400 G_0421
do
EDITED_GENOMES=$DIR_Contigs/${genome}.fa
cat $EDITED_GENOMES >> $RAW_CONCAT_GENOMES_BOTH
done



# Reformat concatenated-genomes using ANVIO (anvi-script-reformat-fasta)
anvi-script-reformat-fasta -o $EDITED_CONCAT_GENOMES_COMPLETE --simplify-names --prefix Complete_genomes -r $PROJECT_REPORT_COMPLETE $RAW_CONCAT_GENOMES_COMPLETE
anvi-script-reformat-fasta -o $EDITED_CONCAT_GENOMES_FRAGED --simplify-names --prefix Fragmented_genomes -r $PROJECT_REPORT_FRAGED $RAW_CONCAT_GENOMES_FRAGED
anvi-script-reformat-fasta -o $EDITED_CONCAT_GENOMES_BOTH --simplify-names --prefix Both_types -r $PROJECT_REPORT_BOTH $RAW_CONCAT_GENOMES_BOTH


# Binning and decompose-file 
cat $PROJECT_REPORT_COMPLETE | awk -F'_' -v OFS="_" 'NF{--NF};1' > $BINNING_COMPLETE
while IFS= read -r line
do
intGenomeID=$( echo "$line" | awk '{print $1}' )
newBinID=$( echo "$line" | awk '{print $2}' )
grep "$intGenomeID" $BINNING_COMPLETE | sed -e "s/$intGenomeID/$newBinID/" >> $DECOMPOSE_COMPLETE
done < $CONTIG_LENGTH_TEST_name_conversions


cat $PROJECT_REPORT_FRAGED | awk -F'_' -v OFS="_" 'NF{--NF};1' > $BINNING_FRAGED
while IFS= read -r line
do
intGenomeID=$( echo "$line" | awk '{print $1}' )
newBinID=$( echo "$line" | awk '{print $2}' )
grep "$intGenomeID" $BINNING_FRAGED | sed -e "s/$intGenomeID/$newBinID/" >> $DECOMPOSE_FRAGED
done < $CONTIG_LENGTH_TEST_name_conversions


cat $PROJECT_REPORT_BOTH | awk -F'_' -v OFS="_" 'NF{--NF};1' > $BINNING_BOTH
while IFS= read -r line
do
intGenomeID=$( echo "$line" | awk '{print $1}' )
newBinID=$( echo "$line" | awk '{print $2}' )
grep "$intGenomeID" $BINNING_BOTH | sed -e "s/$intGenomeID/$newBinID/" >> $DECOMPOSE_BOTH
done < $CONTIG_LENGTH_TEST_name_conversions

```

```{bash Build contigs DB for concatenated reference}
projectID=P_0003_Neisseriaceae
mainDIR=/workspace/jmarkwelchlab
DIR_CLE=$mainDIR/$projectID/CONTIG_LENGTH_TEST
DIR_Contigs=$DIR_CLE/03_GENOMES_EDITED
DIR_ContigsDB=$DIR_CLE/04_CONTIGS_DB
EDITED_CONCAT_GENOMES_BOTH=$DIR_Contigs/Edited_Concatenated_both.fa
CONTIGS_DB=$DIR_ContigsDB/Concatenated_both-contigs.db
TH=10
name=Edited_Concatenated_both

anvi-gen-contigs-database -f $EDITED_CONCAT_GENOMES_BOTH -n $name -o $CONTIGS_DB -T $TH

```

# 5. Mapping

Set up samples.txt file with list of metagenomes that will be mapped to the selected refrence genomes. We will focus only on TD samples in whoch N. subflava and K.kingae were detected in the metaphlan 4 results.

```{r create samples txt file}


# load packages
library(ggplot2)
library(reshape2)
library(tidyr)

# load metaphlan results
metaphlan_df <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/DATA/Metaphlan4_jan31db_Neisseriaceae.csv", header = TRUE)

# melt wide to long
metaphlan_df_melted <- reshape2::melt(metaphlan_df)

# make oral site variable for facet grid
metaphlan_df_melted <- metaphlan_df_melted %>% 
  mutate(site = case_when(grepl('_TD', variable) ~ 'TD',
                          grepl('_KG', variable) ~ 'KG', 
                          grepl('_SUPP', variable) ~ 'SUPP',
                          grepl('_SUBP', variable) ~ 'SUBP',
                          grepl('_SV', variable) ~ 'SV',
                          grepl('_TH', variable) ~ 'TH',
                          grepl('_PT', variable) ~ 'PT',
                          grepl('_HP', variable) ~ 'HP',
                          grepl('_BM', variable) ~ 'BM',
                          grepl('_PERIO', variable) ~ 'PERIO'))

metaphlan_df_melted_TD <- metaphlan_df_melted %>% 
  filter(site == "TD") %>% 
  filter(clade_name == "Neisseria_subflava" | clade_name == "Kingella_kingae") %>% 
  arrange(-value) %>% 
  droplevels()

# Assuming your data frame is called 'df'
metaphlan_df_melted_TD <- metaphlan_df_melted_TD %>%
  mutate(variable = gsub("_TD.profile", "", variable)) %>% 
  select(clade_name,variable, value)

# Assuming your data frame is called 'df'
metaphlan_df_melted_TD <- metaphlan_df_melted_TD %>%
  pivot_wider(names_from = clade_name, values_from = value) %>% # pivot wider so that there is one column of rel abs for each taxa
  arrange(-Kingella_kingae) # sort descending for Kingella_kingae
 
  
# load metadata
metadata <- read.csv("/Users/home/SPECIES_LEVEL_PANGENOMES/Haemophilus_and_Aggregatibacter/DATA/hmp_metadata_merged.csv", header = TRUE)

metadata_subset <- metadata %>% 
  select(BioSample, Internal_ID)

# merge the two
merged_df <- merge(metaphlan_df_melted_TD, metadata_subset, by.x = "variable", by.y = "BioSample")

# Assuming your data frame is called 'df'
samples <- merged_df %>%
  arrange(-Kingella_kingae) %>% # sort descending for Kingella_kingae
  distinct() %>% 
  slice(1:30) # get top 30 sample IDs

# convert to R data frame object
samples <- as.data.frame(samples)

# save list of sample IDs
write.table(samples, "/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/Top_30_TD_samples.txt", quote = FALSE, row.names = FALSE, sep = "\t")


```

```{bash}

scp -r /Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/Top_30_TD_samples.txt jgiacomini@evol5.mbl.edu:/workspace/jmarkwelchlab/P_0003_Neisseriaceae/CONTIG_LENGTH_TEST/Top_30_TD_samples.txt

```


```{bash Mapping}

clusterize -n 10 -m jgiacomini@forsyth.org -log LOGS/Contig_length_mapping_test.log /workspace/jmarkwelchlab/P_0003_Neisseriaceae/SCRIPTS/contig_length_test-indexing_and_mapping_parallel.sh

```

# 6. Profile mapped reads

```{bash Profiling - No contig length filter}

##### No contig length filter #####
clusterize -n 10 -m jgiacomini@forsyth.org -log LOGS/Contig_length_mapping_test_profiling_no_filter.log /workspace/jmarkwelchlab/P_0003_Neisseriaceae/SCRIPTS/contig_length_test-single_profiling_parallel.sh


# Merge single profiles - No contig length filter
projectID=P_0003_Neisseriaceae
mainDIR=/workspace/jmarkwelchlab
DIR_CLE=$mainDIR/$projectID/CONTIG_LENGTH_TEST
DIR_ContigsDB=$DIR_CLE/04_CONTIGS_DB
DIR_SinglePROF=$DIR_CLE/06_PROFILES
DIR_MergedPROF=$DIR_CLE/07_MERGED_PROFILE
SINGLE_PROFILES_DB=$DIR_SinglePROF/Concatenated_both_no_filter/*/PROFILE.db
CONTIGS_DB=$DIR_ContigsDB/Concatenated_both-contigs.db
MERGED_PROFILE=$DIR_MergedPROF/Concatenated_both_no_filter
SAMPLE_NAME=Concatenated_both_no_filter

anvi-merge -c $CONTIGS_DB --enforce-hierarchical-clustering -o $MERGED_PROFILE -S $SAMPLE_NAME $SINGLE_PROFILES_DB

```

```{bash Profiling - Short contig length filter}

##### Short contig length filter (300 nts) #####
clusterize -n 10 -m jgiacomini@forsyth.org -log LOGS/Contig_length_mapping_test_profiling_short_contig_filter.log /workspace/jmarkwelchlab/P_0003_Neisseriaceae/SCRIPTS/contig_length_test-single_profiling_short_contig_filter_parallel.sh

# Merge single profiles - Short contig length filter
projectID=P_0003_Neisseriaceae
mainDIR=/workspace/jmarkwelchlab
DIR_CLE=$mainDIR/$projectID/CONTIG_LENGTH_TEST
DIR_ContigsDB=$DIR_CLE/04_CONTIGS_DB
DIR_SinglePROF=$DIR_CLE/06_PROFILES
DIR_MergedPROF=$DIR_CLE/07_MERGED_PROFILE
SINGLE_PROFILES_DB=$DIR_SinglePROF/Concatenated_both_short_filter/*/PROFILE.db
CONTIGS_DB=$DIR_ContigsDB/Concatenated_both-contigs.db
MERGED_PROFILE=$DIR_MergedPROF/Concatenated_both_short_filter
SAMPLE_NAME=Concatenated_both_short_filter

anvi-merge -c $CONTIGS_DB --enforce-hierarchical-clustering -o $MERGED_PROFILE -S $SAMPLE_NAME $SINGLE_PROFILES_DB

```


# 7. Summarize profiles

```{bash Summarize results for no contig filter}

# decompose merged profile and summarize

projectID=P_0003_Neisseriaceae
mainDIR=/workspace/jmarkwelchlab
DIR_CLE=$mainDIR/$projectID/CONTIG_LENGTH_TEST
DIR_ContigsDB=$DIR_CLE/04_CONTIGS_DB
CONTIGS_DB=$DIR_ContigsDB/Concatenated_both-contigs.db
DIR_MergedPROF=$DIR_CLE/07_MERGED_PROFILE
MERGED_PROFILE_DB=$DIR_MergedPROF/Concatenated_both_no_filter/PROFILE.db
COLLECTION=Genomes
DECOMPOSE_BOTH=$DIR_Contigs/both_decompose.tsv
DIR_SummaryPROF=$DIR_CLE/08_PROFILE_SUMMARY
SUMMARY=$DIR_SummaryPROF/Concatenated_both_no_filter-profile 
                                                                                                                                     

anvi-import-collection -c $CONTIGS_DB -p $MERGED_PROFILE_DB -C $COLLECTION --contigs-mode $DECOMPOSE_BOTH

anvi-summarize -c $CONTIGS_DB -p $MERGED_PROFILE_DB --init-gene-coverages --report-aa-seqs-for-gene-calls -C $COLLECTION -o $SUMMARY

```

```{bash Summarize results for short contig filter}

# decompose merged profile and summarize
projectID=P_0003_Neisseriaceae
mainDIR=/workspace/jmarkwelchlab
DIR_CLE=$mainDIR/$projectID/CONTIG_LENGTH_TEST
DIR_ContigsDB=$DIR_CLE/04_CONTIGS_DB
CONTIGS_DB=$DIR_ContigsDB/Concatenated_both-contigs.db
DIR_MergedPROF=$DIR_CLE/07_MERGED_PROFILE
MERGED_PROFILE_DB=$DIR_MergedPROF/Concatenated_both_short_filter/PROFILE.db
COLLECTION=Genomes_short_filter
DECOMPOSE_BOTH=$DIR_Contigs/both_decompose.tsv
DIR_SummaryPROF=$DIR_CLE/08_PROFILE_SUMMARY
SUMMARY=$DIR_SummaryPROF/Concatenated_both_short_filter-profile 
                                                                                                                                     

anvi-import-collection -c $CONTIGS_DB -p $MERGED_PROFILE_DB -C $COLLECTION --contigs-mode $DECOMPOSE_BOTH

anvi-summarize -c $CONTIGS_DB -p $MERGED_PROFILE_DB --init-gene-coverages --report-aa-seqs-for-gene-calls -C $COLLECTION -o $SUMMARY

```



Profile reads for different thresholds of contig length filters - 300 (already complete), 500 and 1000

* Still need to edit the threshiolds in the single profile scripts *
```{bash}

# set up directories (Note that we do not make the merged profile directory ahead of running the anvi-merge function)
projectID=P_0003_Neisseriaceae
mainDIR=/workspace/jmarkwelchlab
DIR_CLE=$mainDIR/$projectID/CONTIG_LENGTH_TEST
DIR_ContigsDB=$DIR_CLE/04_CONTIGS_DB
DIR_SinglePROF=$DIR_CLE/06_PROFILES
DIR_MergedPROF=$DIR_CLE/07_MERGED_PROFILE
CONTIGS_DB=$DIR_ContigsDB/Concatenated_both-contigs.db
DECOMPOSE_BOTH=$DIR_Contigs/both_decompose.tsv
DIR_SummaryPROF=$DIR_CLE/08_PROFILE_SUMMARY


mkdir $DIR_SinglePROF/Concatenated_both_short_500_filter
mkdir $DIR_SinglePROF/Concatenated_both_short_1000_filter


clusterize -n 10 -m jgiacomini@forsyth.org -log LOGS/Contig_length_mapping_test_profiling_short_500_contig_filter.log /workspace/jmarkwelchlab/P_0003_Neisseriaceae/SCRIPTS/contig_length_test-single_profiling_short_500_contig_filter_parallel.sh

clusterize -n 10 -m jgiacomini@forsyth.org -log LOGS/Contig_length_mapping_test_profiling_short_1000_contig_filter.log /workspace/jmarkwelchlab/P_0003_Neisseriaceae/SCRIPTS/contig_length_test-single_profiling_short_1000_contig_filter_parallel.sh

SINGLE_PROFILES_DB=$DIR_SinglePROF/Concatenated_both_short_500_filter/*/PROFILE.db
MERGED_PROFILE=$DIR_MergedPROF/Concatenated_both_short_500_filter
MERGED_PROFILE_DB=$DIR_MergedPROF/Concatenated_both_short_500_filter/PROFILE.db
SAMPLE_NAME=Concatenated_both_short_500_filter
COLLECTION=Genomes_short_500_filter
SUMMARY=$DIR_SummaryPROF/Concatenated_both_short_500_filter-profile 

anvi-merge -c $CONTIGS_DB --enforce-hierarchical-clustering -o $MERGED_PROFILE -S $SAMPLE_NAME $SINGLE_PROFILES_DB
anvi-import-collection -c $CONTIGS_DB -p $MERGED_PROFILE_DB -C $COLLECTION --contigs-mode $DECOMPOSE_BOTH
anvi-summarize -c $CONTIGS_DB -p $MERGED_PROFILE_DB --init-gene-coverages --report-aa-seqs-for-gene-calls -C $COLLECTION -o $SUMMARY


# TD_HC_HMP_S072_01 failed for no good reason; seems like it was cut short. Maybe issue with memory
anvi-profile -i CONTIG_LENGTH_TEST/05_MAPPING/TD_HC_HMP_S072_01.bam -c $CONTIGS_DB -o $DIR_SinglePROF/Concatenated_both_short_1000_filter/TD_HC_HMP_S072_01 --min-contig-length 1000 -S TD_HC_HMP_S072_01 --num-threads 10 --write-buffer-size-per-thread 30000

SINGLE_PROFILES_DB=$DIR_SinglePROF/Concatenated_both_short_1000_filter/*/PROFILE.db
MERGED_PROFILE=$DIR_MergedPROF/Concatenated_both_short_1000_filter
MERGED_PROFILE_DB=$DIR_MergedPROF/Concatenated_both_short_1000_filter/PROFILE.db
SAMPLE_NAME=Concatenated_both_short_1000_filter
COLLECTION=Genomes_short_1000_filter
SUMMARY=$DIR_SummaryPROF/Concatenated_both_short_1000_filter-profile 

anvi-merge -c $CONTIGS_DB --enforce-hierarchical-clustering -o $MERGED_PROFILE -S $SAMPLE_NAME $SINGLE_PROFILES_DB
anvi-import-collection -c $CONTIGS_DB -p $MERGED_PROFILE_DB -C $COLLECTION --contigs-mode $DECOMPOSE_BOTH
anvi-summarize -c $CONTIGS_DB -p $MERGED_PROFILE_DB --init-gene-coverages --report-aa-seqs-for-gene-calls -C $COLLECTION -o $SUMMARY





```

# 8. Plot results

Send to local
```{bash}
scp -r jgiacomini@evol5.mbl.edu:/workspace/jmarkwelchlab/P_0003_Neisseriaceae/CONTIG_LENGTH_TEST/08_PROFILE_SUMMARY /Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/

```


mean coverage plots

```{r mean coverage plots}

library(reshape2)
library(dplyr)
library(ggplot2)

No_filter_mean_coverage <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/08_PROFILE_SUMMARY/Concatenated_both_no_filter-profile/bins_across_samples/mean_coverage.txt", header = TRUE)

Short_300_filter_mean_coverage <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/08_PROFILE_SUMMARY/Concatenated_both_short_300_filter-profile/bins_across_samples/mean_coverage.txt", header = TRUE)

Short_500_filter_mean_coverage <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/08_PROFILE_SUMMARY/Concatenated_both_short_500_filter-profile/bins_across_samples/mean_coverage.txt", header = TRUE)

Short_1000_filter_mean_coverage <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/08_PROFILE_SUMMARY/Concatenated_both_short_1000_filter-profile/bins_across_samples/mean_coverage.txt", header = TRUE)


# melt each & merge into one df
No_filter_mean_coverage_melted <- melt(No_filter_mean_coverage) %>% 
  rename(no_filter_value = value)

Short_300_filter_mean_coverage_melted <- melt(Short_300_filter_mean_coverage) %>% 
  rename(short_300_filter_value = value)

Short_500_filter_mean_coverage_melted <- melt(Short_500_filter_mean_coverage) %>% 
  rename(short_500_filter_value = value)

Short_1000_filter_mean_coverage_melted <- melt(Short_1000_filter_mean_coverage) %>% 
  rename(short_1000_filter_value = value)

merged_df <- data.frame(No_filter_mean_coverage_melted, 
                        Short_300_filter_mean_coverage_melted$short_300_filter_value,
                        Short_500_filter_mean_coverage_melted$short_500_filter_value,
                        Short_1000_filter_mean_coverage_melted$short_1000_filter_value)


# rename genomes
merged_df <- merged_df %>% 
  mutate(Genome = case_when(grepl('K_kingae_str_ATCC_23332_id_GCA_011612385_1', bins) ~ 'Complete K. kingae',
                            grepl('K_kingae_str_KK100_id_GCA_030182025_1', bins) ~ 'Fragmented K. kingae',
                            grepl('N_subflava_str_ATCC_49275_id_GCA_005221305_1', bins) ~ 'Complete N. subflava',
                            grepl('N_subflava_str_UMB1050_id_GCA_030218065_1', bins) ~ 'Fragmented N. subflava')) %>% 
  rename(no_filter = no_filter_value,
         filter_300 = Short_300_filter_mean_coverage_melted.short_300_filter_value,
         filter_500 = Short_500_filter_mean_coverage_melted.short_500_filter_value,
         filter_1000 = Short_1000_filter_mean_coverage_melted.short_1000_filter_value)



merged_df <- merged_df %>% 
  select(bins, Genome, variable, no_filter, filter_300, filter_500, filter_1000)


library(tidyr)

# Assuming your data frame is called 'df'
merged_df_melted <- merged_df %>%
  pivot_longer(cols = -c(bins, Genome, variable), 
               names_to = "ID", 
               values_to = "values")

merged_df_melted <- as.data.frame(merged_df_melted)

merged_df_melted$ID <- factor(merged_df_melted$ID, levels = c("no_filter", "filter_300", "filter_500", "filter_1000"))

# Check the levels of the 'ID' column
levels(merged_df_melted$ID)

# Estimate average detection values for each genome
merged_df_melted %>% 
  group_by(bins) %>% 
  summarize(mean = mean(values),
            sd = sd(values))

##### N. subflava #####
merged_df_melted_N_subflava <- merged_df_melted %>% 
  filter(Genome %in% c("Complete N. subflava", "Fragmented N. subflava"))

legend_labels <- c("None", "300", "500", "1000")

N_subflava_mean_covergae <- ggplot(merged_df_melted_N_subflava, aes(x = Genome, y = values, fill = ID)) +
  geom_boxplot(position = position_dodge(width = 0.8)) +
  geom_jitter(aes(y = values, group = ID), position = position_dodge(width = 0.8)) +
  labs(x = NULL, y = "Mean genome coverage (depth)", fill = "Contig length filter (nts)") +  
  scale_fill_manual(values = c("no_filter" = "gray","filter_300" = "brown1", "filter_500" = "darkslategray1", "filter_1000" = "chartreuse3"),
                    labels = legend_labels) +
  theme_classic() +
  theme(axis.text = element_text(color = "black", size = 12),
        axis.title = element_text(color = "black", size = 12))
  theme_classic()

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/Figures/N_subflava_mean_covergae_box_whisker.pdf", N_subflava_mean_covergae, width = 7, height = 5)


##### K. kingae #####
merged_df_melted_K_kingae <- merged_df_melted %>% 
  filter(Genome %in% c("Complete K. kingae", "Fragmented K. kingae"))

legend_labels <- c("None", "300", "500", "1000")

K_kingae_mean_covergae <- ggplot(merged_df_melted_K_kingae, aes(x = Genome, y = values, fill = ID)) +
  geom_boxplot(position = position_dodge(width = 0.8)) +
  geom_jitter(aes(y = values, group = ID), position = position_dodge(width = 0.8)) +
  labs(x = NULL, y = "Mean genome coverage (depth)", fill = "Contig length filter (nts)") +  
  scale_fill_manual(values = c("no_filter" = "gray","filter_300" = "brown1", "filter_500" = "darkslategray1", "filter_1000" = "chartreuse3"),
                    labels = legend_labels) +
  theme_classic() +
  theme(axis.text = element_text(color = "black", size = 12),
        axis.title = element_text(color = "black", size = 12))
  theme_classic()

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/Figures/K_kingae_mean_covergae_box_whisker.pdf", K_kingae_mean_covergae, width = 7, height = 5)



##### Fragmented - differences #####
merged_df_melted_fragmented <- merged_df_melted %>% 
  filter(Genome %in% c("Fragmented K. kingae", "Fragmented N. subflava")) %>% 
  pivot_wider(names_from = ID, values_from = values) %>%
  mutate(
    diff_300 = filter_300 - no_filter,
    diff_500 = filter_500 - no_filter,
    diff_1000 = filter_1000 - no_filter) %>% 
  select(bins, Genome, variable, diff_300, diff_500, diff_1000) %>% 
  pivot_longer(cols = -c(bins, Genome, variable), 
               names_to = "ID", 
               values_to = "values")

merged_df_melted_fragmented <- as.data.frame(merged_df_melted_fragmented)
merged_df_melted_fragmented$ID <- factor(merged_df_melted_fragmented$ID, levels = c("diff_300", "diff_500", "diff_1000"))
merged_df_melted_fragmented$Genome <- factor(merged_df_melted_fragmented$Genome, levels = c("Fragmented N. subflava", "Fragmented K. kingae"))

legend_labels <- c("300", "500", "1000")

frag_diff_mean_coverage_box_whisker <- ggplot(merged_df_melted_fragmented, aes(x = Genome, y = values, fill = ID)) +
  geom_boxplot(position = position_dodge(width = 0.8)) +
  geom_jitter(aes(y = values, group = ID), position = position_dodge(width = 0.8)) +
  labs(x = NULL, y = "Difference in mean genome coverage (depth)", fill = "Contig length filter (nts)") +  
  scale_fill_manual(values = c("diff_300" = "brown1", "diff_500" = "darkslategray1", "diff_1000" = "chartreuse3"),
                    labels = legend_labels) +
  theme_classic() +
  theme(axis.text = element_text(color = "black", size = 12),
        axis.title = element_text(color = "black", size = 12))

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/Figures/frag_diff_mean_coverage_box_whisker.pdf", frag_diff_mean_coverage_box_whisker, width = 7, height = 5)

```



Q2Q3 mean coverage plots

```{r Q2Q3 mean coverage plots}

library(reshape2)
library(dplyr)
library(ggplot2)

No_filter_mean_coverage <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/08_PROFILE_SUMMARY/Concatenated_both_no_filter-profile/bins_across_samples/mean_coverage_Q2Q3.txt", header = TRUE)

Short_300_filter_mean_coverage <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/08_PROFILE_SUMMARY/Concatenated_both_short_300_filter-profile/bins_across_samples/mean_coverage_Q2Q3.txt", header = TRUE)

Short_500_filter_mean_coverage <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/08_PROFILE_SUMMARY/Concatenated_both_short_500_filter-profile/bins_across_samples/mean_coverage_Q2Q3.txt", header = TRUE)

Short_1000_filter_mean_coverage <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/08_PROFILE_SUMMARY/Concatenated_both_short_1000_filter-profile/bins_across_samples/mean_coverage_Q2Q3.txt", header = TRUE)


# melt each & merge into one df
No_filter_mean_coverage_melted <- melt(No_filter_mean_coverage) %>% 
  rename(no_filter_value = value)

Short_300_filter_mean_coverage_melted <- melt(Short_300_filter_mean_coverage) %>% 
  rename(short_300_filter_value = value)

Short_500_filter_mean_coverage_melted <- melt(Short_500_filter_mean_coverage) %>% 
  rename(short_500_filter_value = value)

Short_1000_filter_mean_coverage_melted <- melt(Short_1000_filter_mean_coverage) %>% 
  rename(short_1000_filter_value = value)

merged_df <- data.frame(No_filter_mean_coverage_melted, 
                        Short_300_filter_mean_coverage_melted$short_300_filter_value,
                        Short_500_filter_mean_coverage_melted$short_500_filter_value,
                        Short_1000_filter_mean_coverage_melted$short_1000_filter_value)


# rename genomes
merged_df <- merged_df %>% 
  mutate(Genome = case_when(grepl('K_kingae_str_ATCC_23332_id_GCA_011612385_1', bins) ~ 'Complete K. kingae',
                            grepl('K_kingae_str_KK100_id_GCA_030182025_1', bins) ~ 'Fragmented K. kingae',
                            grepl('N_subflava_str_ATCC_49275_id_GCA_005221305_1', bins) ~ 'Complete N. subflava',
                            grepl('N_subflava_str_UMB1050_id_GCA_030218065_1', bins) ~ 'Fragmented N. subflava')) %>% 
  rename(no_filter = no_filter_value,
         filter_300 = Short_300_filter_mean_coverage_melted.short_300_filter_value,
         filter_500 = Short_500_filter_mean_coverage_melted.short_500_filter_value,
         filter_1000 = Short_1000_filter_mean_coverage_melted.short_1000_filter_value)



merged_df <- merged_df %>% 
  select(bins, Genome, variable, no_filter, filter_300, filter_500, filter_1000)




# Assuming your data frame is called 'df'
merged_df_melted <- merged_df %>%
  pivot_longer(cols = -c(bins, Genome, variable), 
               names_to = "ID", 
               values_to = "values")

merged_df_melted <- as.data.frame(merged_df_melted)

merged_df_melted$ID <- factor(merged_df_melted$ID, levels = c("no_filter", "filter_300", "filter_500", "filter_1000"))

# Estimate average detection values for each genome
merged_df_melted %>% 
  group_by(bins) %>% 
  summarize(mean = mean(values),
            sd = sd(values))

##### N. subflava #####
merged_df_melted_N_subflava <- merged_df_melted %>% 
  filter(Genome %in% c("Complete N. subflava", "Fragmented N. subflava"))

legend_labels <- c("None", "300", "500", "1000")

N_subflava_mean_covergae <- ggplot(merged_df_melted_N_subflava, aes(x = Genome, y = values, fill = ID)) +
  geom_boxplot(position = position_dodge(width = 0.8)) +
  geom_jitter(aes(y = values, group = ID), position = position_dodge(width = 0.8)) +
  labs(x = NULL, y = "Mean genome coverage (depth)", fill = "Contig length filter (nts)") +  
  scale_fill_manual(values = c("no_filter" = "gray","filter_300" = "brown1", "filter_500" = "darkslategray1", "filter_1000" = "chartreuse3"),
                    labels = legend_labels) +
  theme_classic() +
  theme(axis.text = element_text(color = "black", size = 12),
        axis.title = element_text(color = "black", size = 12)) 

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/Figures/N_subflava_Q2Q3_mean_covergae_box_whisker.pdf", N_subflava_mean_covergae, width = 7, height = 5)


##### K. kingae #####
merged_df_melted_K_kingae <- merged_df_melted %>% 
  filter(Genome %in% c("Complete K. kingae", "Fragmented K. kingae"))

legend_labels <- c("None", "300", "500", "1000")

K_kingae_mean_covergae <- ggplot(merged_df_melted_K_kingae, aes(x = Genome, y = values, fill = ID)) +
  geom_boxplot(position = position_dodge(width = 0.8)) +
  geom_jitter(aes(y = values, group = ID), position = position_dodge(width = 0.8)) +
  labs(x = NULL, y = "Mean genome coverage (depth)", fill = "Contig length filter (nts)") +  
  scale_fill_manual(values = c("no_filter" = "gray","filter_300" = "brown1", "filter_500" = "darkslategray1", "filter_1000" = "chartreuse3"),
                    labels = legend_labels) +
  theme_classic() +
  theme(axis.text = element_text(color = "black", size = 12),
        axis.title = element_text(color = "black", size = 12)) 

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/Figures/K_kingae_Q2Q3_mean_covergae_box_whisker.pdf", K_kingae_mean_covergae, width = 7, height = 5)



##### Fragmented - differences #####
merged_df_melted_fragmented <- merged_df_melted %>% 
  filter(Genome %in% c("Fragmented K. kingae", "Fragmented N. subflava")) %>% 
  pivot_wider(names_from = ID, values_from = values) %>%
  mutate(
    diff_300 = filter_300 - no_filter,
    diff_500 = filter_500 - no_filter,
    diff_1000 = filter_1000 - no_filter) %>% 
  select(bins, Genome, variable, diff_300, diff_500, diff_1000) %>% 
  pivot_longer(cols = -c(bins, Genome, variable), 
               names_to = "ID", 
               values_to = "values")

merged_df_melted_fragmented <- as.data.frame(merged_df_melted_fragmented)
merged_df_melted_fragmented$ID <- factor(merged_df_melted_fragmented$ID, levels = c("diff_300", "diff_500", "diff_1000"))
merged_df_melted_fragmented$Genome <- factor(merged_df_melted_fragmented$Genome, levels = c("Fragmented N. subflava", "Fragmented K. kingae"))

legend_labels <- c("300", "500", "1000")

frag_diff_mean_coverage_box_whisker <- ggplot(merged_df_melted_fragmented, aes(x = Genome, y = values, fill = ID)) +
  geom_boxplot(position = position_dodge(width = 0.8)) +
  geom_jitter(aes(y = values, group = ID), position = position_dodge(width = 0.8)) +
  labs(x = NULL, y = "Difference in mean genome coverage (depth)", fill = "Contig length filter (nts)") +  
  scale_fill_manual(values = c("diff_300" = "brown1", "diff_500" = "darkslategray1", "diff_1000" = "chartreuse3"),
                    labels = legend_labels) +
  theme_classic() +
  theme(axis.text = element_text(color = "black", size = 12),
        axis.title = element_text(color = "black", size = 12))

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/Figures/frag_diff_Q2Q3_mean_coverage_box_whisker.pdf", frag_diff_mean_coverage_box_whisker, width = 7, height = 5)

merged_df_melted_fragmented %>% 
  filter(Genome == "Fragmented K. kingae") %>% 
  summarise(mean = mean(values))

merged_df_melted_fragmented %>% 
  filter(Genome == "Fragmented N. subflava") %>% 
  summarise(mean = mean(values))

 log(2.435786) - log(0.0005188609) = 8.454144

```


Q2Q3 mean coverage plots

```{r Q2Q3 mean coverage}

library(reshape2)
library(dplyr)
library(ggplot2)

No_filter_mean_coverage <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/08_PROFILE_SUMMARY/Concatenated_both_no_filter-profile/bins_across_samples/mean_coverage_Q2Q3.txt", header = TRUE)

Short_300_filter_mean_coverage <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/08_PROFILE_SUMMARY/Concatenated_both_short_300_filter-profile/bins_across_samples/mean_coverage_Q2Q3.txt", header = TRUE)

Short_500_filter_mean_coverage <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/08_PROFILE_SUMMARY/Concatenated_both_short_500_filter-profile/bins_across_samples/mean_coverage_Q2Q3.txt", header = TRUE)

Short_1000_filter_mean_coverage <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/08_PROFILE_SUMMARY/Concatenated_both_short_1000_filter-profile/bins_across_samples/mean_coverage_Q2Q3.txt", header = TRUE)


# melt each & merge into one df
No_filter_mean_coverage_melted <- melt(No_filter_mean_coverage) %>% 
  rename(no_filter_value = value)

Short_300_filter_mean_coverage_melted <- melt(Short_300_filter_mean_coverage) %>% 
  rename(short_300_filter_value = value)

Short_500_filter_mean_coverage_melted <- melt(Short_500_filter_mean_coverage) %>% 
  rename(short_500_filter_value = value)

Short_1000_filter_mean_coverage_melted <- melt(Short_1000_filter_mean_coverage) %>% 
  rename(short_1000_filter_value = value)

merged_df <- data.frame(No_filter_mean_coverage_melted, 
                        Short_300_filter_mean_coverage_melted$short_300_filter_value,
                        Short_500_filter_mean_coverage_melted$short_500_filter_value,
                        Short_1000_filter_mean_coverage_melted$short_1000_filter_value)


# rename genomes
merged_df <- merged_df %>% 
  mutate(Genome = case_when(grepl('K_kingae_str_ATCC_23332_id_GCA_011612385_1', bins) ~ 'Complete K. kingae',
                            grepl('K_kingae_str_KK100_id_GCA_030182025_1', bins) ~ 'Fragmented K. kingae',
                            grepl('N_subflava_str_ATCC_49275_id_GCA_005221305_1', bins) ~ 'Complete N. subflava',
                            grepl('N_subflava_str_UMB1050_id_GCA_030218065_1', bins) ~ 'Fragmented N. subflava')) %>% 
  rename(no_filter = no_filter_value,
         filter_300 = Short_300_filter_mean_coverage_melted.short_300_filter_value,
         filter_500 = Short_500_filter_mean_coverage_melted.short_500_filter_value,
         filter_1000 = Short_1000_filter_mean_coverage_melted.short_1000_filter_value)



merged_df <- merged_df %>% 
  select(bins, Genome, variable, no_filter, filter_300, filter_500, filter_1000)




# Assuming your data frame is called 'df'
merged_df_melted <- merged_df %>%
  pivot_longer(cols = -c(bins, Genome, variable), 
               names_to = "ID", 
               values_to = "values")

merged_df_melted <- as.data.frame(merged_df_melted)

merged_df_melted$ID <- factor(merged_df_melted$ID, levels = c("no_filter", "filter_300", "filter_500", "filter_1000"))

# Estimate average detection values for each genome
merged_df_melted %>% 
  group_by(bins) %>% 
  summarize(mean = mean(values),
            sd = sd(values))

##### N. subflava #####
merged_df_melted_N_subflava <- merged_df_melted %>% 
  filter(Genome %in% c("Complete N. subflava", "Fragmented N. subflava"))

legend_labels <- c("None", "300", "500", "1000")

N_subflava_mean_covergae <- ggplot(merged_df_melted_N_subflava, aes(x = Genome, y = values, fill = ID)) +
  geom_boxplot(position = position_dodge(width = 0.8)) +
  geom_jitter(aes(y = values, group = ID), position = position_dodge(width = 0.8)) +
  labs(x = NULL, y = "Mean genome coverage (depth)", fill = "Contig length filter (nts)") +  
  scale_fill_manual(values = c("no_filter" = "gray","filter_300" = "brown1", "filter_500" = "darkslategray1", "filter_1000" = "chartreuse3"),
                    labels = legend_labels) +
  theme_classic() +
  theme(axis.text = element_text(color = "black", size = 12),
        axis.title = element_text(color = "black", size = 12)) 

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/Figures/N_subflava_Q2Q3_mean_covergae_box_whisker.pdf", N_subflava_mean_covergae, width = 7, height = 5)


##### K. kingae #####
merged_df_melted_K_kingae <- merged_df_melted %>% 
  filter(Genome %in% c("Complete K. kingae", "Fragmented K. kingae"))

legend_labels <- c("None", "300", "500", "1000")

K_kingae_mean_covergae <- ggplot(merged_df_melted_K_kingae, aes(x = Genome, y = values, fill = ID)) +
  geom_boxplot(position = position_dodge(width = 0.8)) +
  geom_jitter(aes(y = values, group = ID), position = position_dodge(width = 0.8)) +
  labs(x = NULL, y = "Mean genome coverage (depth)", fill = "Contig length filter (nts)") +  
  scale_fill_manual(values = c("no_filter" = "gray","filter_300" = "brown1", "filter_500" = "darkslategray1", "filter_1000" = "chartreuse3"),
                    labels = legend_labels) +
  theme_classic() +
  theme(axis.text = element_text(color = "black", size = 12),
        axis.title = element_text(color = "black", size = 12)) 

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/Figures/K_kingae_Q2Q3_mean_covergae_box_whisker.pdf", K_kingae_mean_covergae, width = 7, height = 5)



##### Fragmented - differences #####
merged_df_melted_fragmented <- merged_df_melted %>% 
  filter(Genome %in% c("Fragmented K. kingae", "Fragmented N. subflava")) %>% 
  pivot_wider(names_from = ID, values_from = values) %>%
  mutate(
    diff_300 = filter_300 - no_filter,
    diff_500 = filter_500 - no_filter,
    diff_1000 = filter_1000 - no_filter) %>% 
  select(bins, Genome, variable, diff_300, diff_500, diff_1000) %>% 
  pivot_longer(cols = -c(bins, Genome, variable), 
               names_to = "ID", 
               values_to = "values")

merged_df_melted_fragmented <- as.data.frame(merged_df_melted_fragmented)
merged_df_melted_fragmented$ID <- factor(merged_df_melted_fragmented$ID, levels = c("diff_300", "diff_500", "diff_1000"))
merged_df_melted_fragmented$Genome <- factor(merged_df_melted_fragmented$Genome, levels = c("Fragmented N. subflava", "Fragmented K. kingae"))

legend_labels <- c("300", "500", "1000")

frag_diff_mean_coverage_box_whisker <- ggplot(merged_df_melted_fragmented, aes(x = Genome, y = values, fill = ID)) +
  geom_boxplot(position = position_dodge(width = 0.8)) +
  geom_jitter(aes(y = values, group = ID), position = position_dodge(width = 0.8)) +
  labs(x = NULL, y = "Difference in mean genome coverage (depth)", fill = "Contig length filter (nts)") +  
  scale_fill_manual(values = c("diff_300" = "brown1", "diff_500" = "darkslategray1", "diff_1000" = "chartreuse3"),
                    labels = legend_labels) +
  theme_classic() +
  theme(axis.text = element_text(color = "black", size = 12),
        axis.title = element_text(color = "black", size = 12))

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/Figures/frag_diff_Q2Q3_mean_coverage_box_whisker.pdf", frag_diff_mean_coverage_box_whisker, width = 7, height = 5)


merged_df_melted_fragmented %>% 
  filter(Genome == "Fragmented K. kingae") %>% 
  summarise(mean = mean(values))

merged_df_melted_fragmented %>% 
  filter(Genome == "Fragmented N. subflava") %>% 
  summarise(mean = mean(values))

 log(2.435786) - log(0.0005188609) = 8.454144
 
 kruskal_Q2Q3_mean_covergae <- kruskal.test(values ~ ID, data =  merged_df_melted_fragmented)
# 	Kruskal-Wallis rank sum test
# 
# data:  values by ID
# Kruskal-Wallis chi-squared = 1.3795, df = 2, p-value = 0.5017

library(FSA)

post_hoc_dunn<- dunnTest(values ~ ID,
  data = merged_df_melted_fragmented,
  method = "holm")

post_hoc_dunn_results <- as.data.frame(post_hoc_dunn$res)
#             Comparison          Z   P.unadj     P.adj
# 1 diff_1000 - diff_300  1.1685581 0.2425817 0.7277452
# 2 diff_1000 - diff_500  0.4817893 0.6299556 0.6299556
# 3  diff_300 - diff_500 -0.6867688 0.4922285 0.9844569



kruskal_Q2Q3_mean_covergae_Genome <- kruskal.test(values ~ Genome, data =  merged_df_melted_fragmented)
#Kruskal-Wallis chi-squared = 116.95, df = 1, p-value < 2.2e-16

```


Detection plots

```{r detection}

library(reshape2)
library(dplyr)
library(ggplot2)

No_filter_detection <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/08_PROFILE_SUMMARY/Concatenated_both_no_filter-profile/bins_across_samples/detection.txt", header = TRUE)

Short_300_filter_detection <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/08_PROFILE_SUMMARY/Concatenated_both_short_300_filter-profile/bins_across_samples/detection.txt", header = TRUE)

Short_500_filter_detection <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/08_PROFILE_SUMMARY/Concatenated_both_short_500_filter-profile/bins_across_samples/detection.txt", header = TRUE)

Short_1000_filter_detection <- read.table("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/08_PROFILE_SUMMARY/Concatenated_both_short_1000_filter-profile/bins_across_samples/detection.txt", header = TRUE)


# melt each & merge into one df
No_filter_detection_melted <- melt(No_filter_detection) %>% 
  rename(no_filter_value = value)

Short_300_filter_detection_melted <- melt(Short_300_filter_detection) %>% 
  rename(short_300_filter_value = value)

Short_500_filter_detection_melted <- melt(Short_500_filter_detection) %>% 
  rename(short_500_filter_value = value)

Short_1000_filter_detection_melted <- melt(Short_1000_filter_detection) %>% 
  rename(short_1000_filter_value = value)

merged_detection_df <- data.frame(No_filter_detection_melted, 
                        Short_300_filter_detection_melted$short_300_filter_value,
                        Short_500_filter_detection_melted$short_500_filter_value,
                        Short_1000_filter_detection_melted$short_1000_filter_value)


# rename genomes
merged_detection_df <- merged_detection_df %>% 
  mutate(Genome = case_when(grepl('K_kingae_str_ATCC_23332_id_GCA_011612385_1', bins) ~ 'Complete K. kingae',
                            grepl('K_kingae_str_KK100_id_GCA_030182025_1', bins) ~ 'Fragmented K. kingae',
                            grepl('N_subflava_str_ATCC_49275_id_GCA_005221305_1', bins) ~ 'Complete N. subflava',
                            grepl('N_subflava_str_UMB1050_id_GCA_030218065_1', bins) ~ 'Fragmented N. subflava')) %>% 
  rename(no_filter = no_filter_value,
         filter_300 = Short_300_filter_detection_melted.short_300_filter_value,
         filter_500 = Short_500_filter_detection_melted.short_500_filter_value,
         filter_1000 = Short_1000_filter_detection_melted.short_1000_filter_value)


merged_detection_df <- merged_detection_df %>% 
  select(bins, Genome, variable, no_filter, filter_300, filter_500, filter_1000)


# Assuming your data frame is called 'df'
merged_detection_df_melted <- merged_detection_df %>%
  pivot_longer(cols = -c(bins, Genome, variable), 
               names_to = "ID", 
               values_to = "values")

merged_detection_df_melted <- as.data.frame(merged_detection_df_melted)

merged_detection_df_melted$ID <- factor(merged_detection_df_melted$ID, levels = c("no_filter", "filter_300", "filter_500", "filter_1000"))

# Estimate average detection values for each genome
merged_detection_df_melted %>% 
  group_by(bins) %>% 
  summarize(mean = mean(values),
            sd = sd(values))

##### N. subflava #####
merged_detection_df_melted_N_subflava <- merged_detection_df_melted %>% 
  filter(Genome %in% c("Complete N. subflava", "Fragmented N. subflava"))

legend_labels <- c("None", "300", "500", "1000")

N_subflava_detection <- ggplot(merged_detection_df_melted_N_subflava, aes(x = Genome, y = values, fill = ID)) +
  geom_boxplot(position = position_dodge(width = 0.8)) +
  geom_jitter(aes(y = values, group = ID), position = position_dodge(width = 0.8)) +
  labs(x = NULL, y = "Genome breadth of coverage", fill = "Contig length filter (nts)") +  
  scale_fill_manual(values = c("no_filter" = "gray","filter_300" = "brown1", "filter_500" = "darkslategray1", "filter_1000" = "chartreuse3"),
                    labels = legend_labels) +
  theme_classic() +
  theme(axis.text = element_text(color = "black", size = 12),
        axis.title = element_text(color = "black", size = 12)) 

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/Figures/N_subflava_detection_box_whisker.pdf", N_subflava_detection, width = 7, height = 5)


##### K. kingae #####
merged_detection_df_melted_K_kingae <- merged_detection_df_melted %>% 
  filter(Genome %in% c("Complete K. kingae", "Fragmented K. kingae"))

legend_labels <- c("None", "300", "500", "1000")

K_kingae_detection <- ggplot(merged_detection_df_melted_K_kingae, aes(x = Genome, y = values, fill = ID)) +
  geom_boxplot(position = position_dodge(width = 0.8)) +
  geom_jitter(aes(y = values, group = ID), position = position_dodge(width = 0.8)) +
  labs(x = NULL, y = "Genome breadth of coverage", fill = "Contig length filter (nts)") +  
  scale_fill_manual(values = c("no_filter" = "gray","filter_300" = "brown1", "filter_500" = "darkslategray1", "filter_1000" = "chartreuse3"),
                    labels = legend_labels) +
  theme_classic() +
  theme(axis.text = element_text(color = "black", size = 12),
        axis.title = element_text(color = "black", size = 12)) 

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/Figures/K_kingae_detection_box_whisker.pdf", K_kingae_detection, width = 7, height = 5)



##### Fragmented - differences #####
merged_detection_df_melted_fragmented <- merged_detection_df_melted %>% 
  filter(Genome %in% c("Fragmented K. kingae", "Fragmented N. subflava")) %>% 
  pivot_wider(names_from = ID, values_from = values) %>%
  mutate(
    diff_300 = filter_300 - no_filter,
    diff_500 = filter_500 - no_filter,
    diff_1000 = filter_1000 - no_filter) %>% 
  select(bins, Genome, variable, diff_300, diff_500, diff_1000) %>% 
  pivot_longer(cols = -c(bins, Genome, variable), 
               names_to = "ID", 
               values_to = "values")

merged_detection_df_melted_fragmented <- as.data.frame(merged_detection_df_melted_fragmented)
merged_detection_df_melted_fragmented$ID <- factor(merged_detection_df_melted_fragmented$ID, levels = c("diff_300", "diff_500", "diff_1000"))
merged_detection_df_melted_fragmented$Genome <- factor(merged_detection_df_melted_fragmented$Genome, levels = c("Fragmented N. subflava", "Fragmented K. kingae"))

legend_labels <- c("300", "500", "1000")

frag_diff_detection_box_whisker <- ggplot(merged_detection_df_melted_fragmented, aes(x = Genome, y = values*100, fill = ID)) +
  geom_boxplot(position = position_dodge(width = 0.8)) +
  geom_jitter(aes(y = values*100, group = ID), position = position_dodge(width = 0.8)) +
  labs(x = NULL, y = "Difference in genome breadth of coverage (%)", fill = "Contig length filter (nts)") +  
  scale_fill_manual(values = c("diff_300" = "brown1", "diff_500" = "darkslategray1", "diff_1000" = "chartreuse3"),
                    labels = legend_labels) +
  theme_classic() +
  theme(axis.text = element_text(color = "black", size = 12),
        axis.title = element_text(color = "black", size = 12))

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/Figures/frag_diff_detection_box_whisker.pdf", frag_diff_detection_box_whisker, width = 7, height = 5)



########################
merged_detection_df_melted_fragmented %>% 
  filter(Genome == "Fragmented K. kingae") %>% 
  summarise(mean = mean(values))

merged_detection_df_melted_fragmented %>% 
  filter(Genome == "Fragmented N. subflava") %>% 
  summarise(mean = mean(values))

 log(0.01688368) - log(0.0009634714) = 2.86356
 
 
 anova_detection <- aov(log(values) ~ ID*Genome, data =  merged_detection_df_melted_fragmented)
 summary(anova_detection)
#               Df Sum Sq Mean Sq F value   Pr(>F)    
# ID            2   14.3     7.2  12.959 5.78e-06 ***
# Genome        1  540.6   540.6 976.506  < 2e-16 ***
# ID:Genome     2    8.0     4.0   7.246 0.000955 ***
# Residuals   170   94.1     0.6                     
# ---
# Signif. codes:  0 â€˜***â€™ 0.001 â€˜**â€™ 0.01 â€˜*â€™ 0.05 â€˜.â€™ 0.1 â€˜ â€™ 1
 
#4 observations deleted due to missingness
 TukeyHSD(anova_detection, "ID")
# 
#                         diff         lwr       upr     p adj
# diff_500-diff_300  0.2612850 -0.06401492 0.5865849 0.1420456
# diff_1000-diff_300 0.6929411  0.36764122 1.0182411 0.0000036
# diff_1000-diff_500 0.4316561  0.10774936 0.7555629 0.0054357
 
 par(mfrow = c(1, 2)) # combine plots

# histogram
hist(anova_detection$residuals)

# QQ-plot
library(car)
qqPlot(anova_detection$residuals,
  id = FALSE # id = FALSE to remove point identification
)

shapiro.test(anova_detection$residuals)

# 	Shapiro-Wilk normality test
# 
# data:  anova_detection$residuals
#W = 0.81122, p-value = 8.103e-14


kruskal_detection <- kruskal.test(values ~ ID, data =  merged_detection_df_melted_fragmented)
# 	Kruskal-Wallis rank sum test
# 
# data:  values by ID
# Kruskal-Wallis chi-squared = 16.514, df = 2, p-value = 0.0002595


library(FSA)

post_hoc_dunn<- dunnTest(values ~ ID,
  data = merged_detection_df_melted_fragmented,
  method = "holm")

post_hoc_dunn_results <- as.data.frame(post_hoc_dunn$res)
#             Comparison         Z      P.unadj        P.adj
# 1 diff_1000 - diff_300  3.996223 6.436103e-05 0.0001930831
# 2 diff_1000 - diff_500  1.359522 1.739813e-01 0.1739812940
# 3  diff_300 - diff_500 -2.636701 8.371647e-03 0.0167432945


merged_detection_df_melted_fragmented %>% 
  group_by(ID, Genome) %>% 
  summarise(mean = mean(values))
```

bins    mean    sd
K_kingae_str_ATCC_23332_id_GCA_011612385_1	    0.08021728	    0.06916203		
K_kingae_str_KK100_id_GCA_030182025_1	    0.07085268	    0.06702211		
N_subflava_str_ATCC_49275_id_GCA_005221305_1	    0.90054143	    0.02503405		
N_subflava_str_UMB1050_id_GCA_030218065_1	    0.92963516	    0.02554472	

Does the difference in mean coverage and detection have an impact on relative abundance?

```{r}


merged_detection_df_melted <- merged_detection_df_melted %>% 
  mutate(detected = ifelse(values >= 0.5, "detected", "un-detected"))

merged_detection_df_melted$detected <- as.factor(merged_detection_df_melted$detected)

melted_df_detect_filtered <- merged_detection_df_melted %>% 
  filter(detected == "detected")

#Filter mean_coverage_Q2Q3 data frame by detected genomes
filtered_melted_df <-merge(merged_df_melted, melted_df_detect_filtered, by=c("variable","bins", "ID")) 

filtered_melted_df <- filtered_melted_df %>% dplyr::rename(coverage=values.x,
                                                           breadth=values.y)


# select columns and transpose from long to wide
meanCov <- filtered_melted_df %>% 
  dplyr::select(variable, bins,ID, coverage) %>% 
  pivot_wider(id_cols=c(bins, ID), names_from=variable, values_from=coverage)
# chnage NA values to zeros
meanCov[is.na(meanCov)] <- 0



meanCov_no_filter <- meanCov %>% 
  filter(ID %in% "no_filter")
rownames_no_filter <- meanCov_no_filter$bins
meanCovT_no_filter <- as.data.frame(t(meanCov_no_filter[-1:-2])) # transpose data frame
colnames(meanCovT_no_filter) <- rownames_no_filter
relAbsT_no_filter <- decostand(meanCovT_no_filter, method = "total") # mean coverage of a genome / sum of mean coverages all genomes in a sample
relAbs_no_filter <- t(relAbsT_no_filter) # transpose relative abundance matrix
relAbs_no_filter <- 100*relAbs_no_filter  # multiply by 100
relAbs_no_filter <- as.data.frame(relAbs_no_filter) # add filter ID column back 
relAbs_no_filter <- relAbs_no_filter %>% 
  mutate(ID = meanCov_no_filter$ID)

meanCov_300_filter <- meanCov %>% 
  filter(ID %in% "filter_300")
rownames_300_filter <- meanCov_300_filter$bins
meanCovT_300_filter <- as.data.frame(t(meanCov_300_filter[-1:-2])) # transpose data frame
colnames(meanCovT_300_filter) <- rownames_300_filter
relAbsT_300_filter <- decostand(meanCovT_300_filter, method = "total") 
relAbs_300_filter <- t(relAbsT_300_filter) # transpose relative abundance matrix
relAbs_300_filter <- 100*relAbs_300_filter  # multiply by 100
relAbs_300_filter <- as.data.frame(relAbs_300_filter) # add filter ID column back 
relAbs_300_filter <- relAbs_300_filter %>% 
  mutate(ID = meanCov_300_filter$ID)

meanCov_500_filter <- meanCov %>% 
  filter(ID %in% "filter_500")
rownames_500_filter <- meanCov_500_filter$bins
meanCovT_500_filter <- as.data.frame(t(meanCov_500_filter[-1:-2])) # transpose data frame
colnames(meanCovT_500_filter) <- rownames_500_filter
relAbsT_500_filter <- decostand(meanCovT_500_filter, method = "total") 
relAbs_500_filter <- t(relAbsT_500_filter) # transpose relative abundance matrix
relAbs_500_filter <- 100*relAbs_500_filter  # multiply by 100
relAbs_500_filter <- as.data.frame(relAbs_500_filter) # add filter ID column back 
relAbs_500_filter <- relAbs_500_filter %>% 
  mutate(ID = meanCov_500_filter$ID)


meanCov_1000_filter <- meanCov %>% 
  filter(ID %in% "filter_1000")
rownames_1000_filter <- meanCov_1000_filter$bins
meanCovT_1000_filter <- as.data.frame(t(meanCov_1000_filter[-1:-2])) # transpose data frame
colnames(meanCovT_1000_filter) <- rownames_1000_filter
relAbsT_1000_filter <- decostand(meanCovT_1000_filter, method = "total") 
relAbs_1000_filter <- t(relAbsT_1000_filter) # transpose relative abundance matrix
relAbs_1000_filter <- 100*relAbs_1000_filter  # multiply by 100
relAbs_1000_filter <- as.data.frame(relAbs_1000_filter) # add filter ID column back 
relAbs_1000_filter <- relAbs_1000_filter %>% 
  mutate(ID = meanCov_1000_filter$ID)


merged_relAbs_wide <- rbind(relAbs_no_filter,relAbs_300_filter,relAbs_500_filter,relAbs_1000_filter)

library(tibble)
merged_relAbs_wide <- tibble::rownames_to_column(merged_relAbs_wide, "Genome")

# change covergae data frame from wide to long format
merged_relAbs_melted <- reshape2::melt(merged_relAbs_wide, id.vars = c("Genome", "ID")) 


merged_relAbs_melted <- merged_relAbs_melted %>% 
  mutate(Genome2 = case_when(grepl('N_subflava_str_ATCC_49275_id_GCA_005221305_1', Genome) ~ 'Complete N. subflava',
                            grepl('N_subflava_str_UMB1050_id_GCA_030218065_1', Genome) ~ 'Fragmented N. subflava')) 

relabs_legend_labels <- c("None", "300", "500", "1000")

relative_abundance_box_whisker <- ggplot(merged_relAbs_melted, aes(x = Genome2, y = value, fill = ID)) +
  geom_boxplot(position = position_dodge(width = 0.8)) +
  geom_jitter(aes(y = value, group = ID), position = position_dodge(width = 0.8)) +
  labs(x = NULL, y = "Relative abundance (%)", fill = "Contig length filter (nts)") +  
  scale_fill_manual(values = c("no_filter" = "gray","filter_300" = "brown1", "filter_500" = "darkslategray1", "filter_1000" = "chartreuse3"),
                    labels = relabs_legend_labels) +
  theme_classic() +
  theme(axis.text = element_text(color = "black", size = 12),
        axis.title = element_text(color = "black", size = 12))

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/Figures/relative_abundance_box_whisker.pdf", relative_abundance_box_whisker, width = 7, height = 5)



### Differences in relative abundance

merged_relAbs_melted_differences <- merged_relAbs_melted %>% 
  filter(Genome2 == "Fragmented N. subflava") %>% 
  select(variable, ID, value) %>% 
  pivot_wider(names_from = ID, values_from = value) %>%
  mutate(
    diff_300 = filter_300 - no_filter,
    diff_500 = filter_500 - no_filter,
    diff_1000 = filter_1000 - no_filter) %>% 
  select(variable, diff_300, diff_500, diff_1000) %>% 
  pivot_longer(cols = -c(variable), 
               names_to = "ID", 
               values_to = "values")

merged_relAbs_melted_differences <- as.data.frame(merged_relAbs_melted_differences)
merged_relAbs_melted_differences$ID <- factor(merged_relAbs_melted_differences$ID, levels = c("diff_300", "diff_500", "diff_1000"))


legend_labels <- c("300", "500", "1000")

merged_relAbs_melted_differences_box_whisker <- ggplot(merged_relAbs_melted_differences, aes(x = ID, y = values, fill = ID)) +
  geom_boxplot(position = position_dodge(width = 0.8)) +
  geom_jitter(aes(y = values, group = ID), position = position_dodge(width = 0.8)) +
  labs(x = "Contig length filter (nts)", y = "Difference in relative abundance (%)", fill = NULL) +  
  scale_fill_manual(values = c("diff_300" = "brown1", "diff_500" = "darkslategray1", "diff_1000" = "chartreuse3"),
                    labels = legend_labels) +
  theme_classic() +
  theme(axis.text = element_text(color = "black", size = 12),
        axis.title = element_text(color = "black", size = 12))

ggsave("/Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/Figures/Fragmented_N_subflava_relAbs_differences_box_whisker.pdf", merged_relAbs_melted_differences_box_whisker, width = 7, height = 5)

summary(aov(value ~ ID*Genome2, data =  merged_relAbs_melted))
#              Df Sum Sq Mean Sq F value Pr(>F)    
# ID            3      0       0   0.000  1.000    
# Genome2       1  20089   20089 641.773 <2e-16 ***
# ID:Genome2    3      6       2   0.067  0.977    
# Residuals   232   7262      31         

kruskal_relative_abundance <- kruskal.test(values ~ ID, data =  merged_relAbs_melted_differences)
# 	Kruskal-Wallis rank sum test
# 
# data:  values by ID
#Kruskal-Wallis chi-squared = 27.158, df = 2, p-value = 1.267e-06

post_hoc_dunn_relative_abundance <- dunnTest(values ~ ID,
  data = merged_relAbs_melted_differences,
  method = "holm")

post_hoc_dunn_relative_abundance_results <- as.data.frame(post_hoc_dunn_relative_abundance$res)
#            Comparison         Z      P.unadj        P.adj
# 1 diff_1000 - diff_300  1.942073 5.212824e-02 5.212824e-02
# 2 diff_1000 - diff_500 -3.217022 1.295287e-03 2.590573e-03
# 3  diff_300 - diff_500 -5.159095 2.481461e-07 7.444382e-07


merged_relAbs_melted_differences %>% 
  group_by(ID) %>% 
  summarise(mean = mean(values),
            sd = sd(values),
            n = n(),
            se = sd/sqrt(n))

```



```{bash}
scp -r jgiacomini@evol5.mbl.edu:/workspace/jmarkwelchlab/P_0003_Neisseriaceae/CONTIG_LENGTH_TEST/05_MAPPING /Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/
 
scp -r jgiacomini@evol5.mbl.edu:/workspace/jmarkwelchlab/P_0003_Neisseriaceae/CONTIG_LENGTH_TEST/03_GENOMES_EDITED /Users/home/SPECIES_LEVEL_PANGENOMES/Neisseriaceae/CONTIG_LENGTH_TEST/

DIR=/workspace/jmarkwelchlab/P_0003_Neisseriaceae/CONTIG_LENGTH_TEST/
input=$DIR/03_GENOMES_EDITED/Edited_Concatenated_both.fa
output=$DIR/03_GENOMES_EDITED/Edited_Concatenated_both_seq_less_than_300.fa
grep -A 1 '^>.*$' $input | awk '{if (length($2) < 300) print $0}' > $output
grep -A 1 '^>.*$' $input | grep -v "^--$" | awk '{if (length($2) < 300) print $0}' > $output




echo "Both_types_000000000627" > $DIR/Both_types_000000000627.txt

DIR=/workspace/jmarkwelchlab/P_0003_Neisseriaceae/CONTIG_LENGTH_TEST/

anvi-export-contigs -c $DIR/04_CONTIGS_DB/Concatenated_both-contigs.db \
                    -o $DIR/04_CONTIGS_DB/N_subflava_str_UMB1050_id_GCA_030218065_1_contig_0627-fasta \
                    --contigs-of-interest $DIR/Both_types_000000000627.txt
                    
                    
echo "Both_types_000000000636" > $DIR/Both_types_000000000636.txt   
anvi-export-contigs -c $DIR/04_CONTIGS_DB/Concatenated_both-contigs.db \
                    -o $DIR/04_CONTIGS_DB/N_subflava_str_UMB1050_id_GCA_030218065_1_contig_0636-fasta \
                    --contigs-of-interest $DIR/Both_types_000000000636.txt
                    
                    
                    
```



The following code extracts contigs in the original concatenated fasta file that are less than 300 bases long and outputs them into a new fasta file that can be used in a blast search. 


```{bash}
python SCRIPTS/filter_short_sequences.py
```


We then need to update the deflines of the new fasta file with IDs that will allow us to identify which contigs came from which genomes. This is crucial since the original concatentaed fasta file that we are working with contains 4 different genomes - 2 complete and 2 fragmentted. 

```{bash}

python SCRIPTS/update_deflines.py 
```

We can now blast the short contigs (NEED TO UPDATE CODE IF RUN ON MBL - 10/11/2023)

To use the new blast databases,
first log into barhal, (won't work anywhere else yet), and load the
module for it:

    module load blast+/2.14.1-testing-barhal

At which point your BLASTDB environment variable should point to the new
databases at /testpool/data/blastdb
```{bash NOT RUN YET}


main_DIR=/workspace/jmarkwelchlab/P_0003_Neisseriaceae
IN=$main_DIR/CONTIG_LENGTH_TEST/03_GENOMES_EDITED/Edited_Concatenated_both_seq_less_than_300_renamed.fa
OUT=$main_DIR/CONTIG_LENGTH_TEST/09_BLAST/seq_less_than_300_blastx_results.txt
clusterize -n 5 -l $main_DIR/LOGS/CONTIG_LENGTH_TEST_blastx_on_short_contigs.log -m jgiacomini@forsyth.org blastx -query $IN -db /testpool/data/blastdb/nr -out $OUT -evalue 1e-5 -outfmt 6

```
